// molstar/lib/extensions/pdbe/preferred-assembly.js
/**
 * Copyright (c) 2018 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
import { __awaiter, __generator } from "tslib";
import { Column, Table } from '../../mol-data/db';
import { toTable } from '../../mol-io/reader/cif/schema';
import { CifWriter } from '../../mol-io/writer/cif';
import { ModelSymmetry } from '../../mol-model-formats/structure/property/symmetry';
import { MmcifFormat } from '../../mol-model-formats/structure/mmcif';
import { CustomPropertyDescriptor } from '../../mol-model/custom-property';
export var PDBePreferredAssembly;
(function (PDBePreferredAssembly) {
    function getFirstFromModel(model) {
        var symmetry = ModelSymmetry.Provider.get(model);
        return (symmetry === null || symmetry === void 0 ? void 0 : symmetry.assemblies.length) ? symmetry.assemblies[0].id : '';
    }
    PDBePreferredAssembly.getFirstFromModel = getFirstFromModel;
    function get(model) {
        return model._staticPropertyData.__PDBePreferredAssebly__ || getFirstFromModel(model);
    }
    PDBePreferredAssembly.get = get;
    function set(model, prop) {
        model._staticPropertyData.__PDBePreferredAssebly__ = prop;
    }
    PDBePreferredAssembly.Schema = {
        pdbe_preferred_assembly: {
            assembly_id: Column.Schema.str
        }
    };
    PDBePreferredAssembly.Descriptor = CustomPropertyDescriptor({
        name: 'pdbe_preferred_assembly',
        cifExport: {
            prefix: 'pdbe',
            context: function (ctx) { return get(ctx.firstModel); },
            categories: [{
                    name: 'pdbe_preferred_assembly',
                    instance: function (ctx) {
                        return CifWriter.Category.ofTable(Table.ofArrays(PDBePreferredAssembly.Schema.pdbe_preferred_assembly, { assembly_id: [ctx] }));
                    }
                }]
        }
    });
    function fromCifData(model) {
        if (!MmcifFormat.is(model.sourceData))
            return void 0;
        var cat = model.sourceData.data.frame.categories.pdbe_preferred_assembly;
        if (!cat)
            return void 0;
        return toTable(PDBePreferredAssembly.Schema.pdbe_preferred_assembly, cat).assembly_id.value(0) || getFirstFromModel(model);
    }
    function attachFromCifOrApi(model, params) {
        return __awaiter(this, void 0, void 0, function () {
            var asmName, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (model.customProperties.has(PDBePreferredAssembly.Descriptor))
                            return [2 /*return*/, true];
                        asmName = fromCifData(model);
                        if (!(asmName === void 0 && params.PDBe_apiSourceJson)) return [3 /*break*/, 2];
                        return [4 /*yield*/, params.PDBe_apiSourceJson(model)];
                    case 1:
                        data = _a.sent();
                        if (!data)
                            return [2 /*return*/, false];
                        asmName = asmNameFromJson(model, data);
                        return [3 /*break*/, 3];
                    case 2: return [2 /*return*/, false];
                    case 3:
                        if (!asmName)
                            return [2 /*return*/, false];
                        model.customProperties.add(PDBePreferredAssembly.Descriptor);
                        set(model, asmName);
                        return [2 /*return*/, true];
                }
            });
        });
    }
    PDBePreferredAssembly.attachFromCifOrApi = attachFromCifOrApi;
})(PDBePreferredAssembly || (PDBePreferredAssembly = {}));
function asmNameFromJson(modelData, data) {
    var assemblies = data[0] && data[0].assemblies;
    if (!assemblies || !assemblies.length)
        return PDBePreferredAssembly.getFirstFromModel(modelData);
    for (var _i = 0, assemblies_1 = assemblies; _i < assemblies_1.length; _i++) {
        var asm = assemblies_1[_i];
        if (asm.preferred) {
            return asm.assembly_id;
        }
    }
    return assemblies[0].assembly_id;
}
