// molstar/lib/commonjs/mol-repr/structure/visual/element-point.js
"use strict";
/**
 * Copyright (c) 2018-2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElementPointParams = void 0;
exports.createElementPoint = createElementPoint;
exports.ElementPointVisual = ElementPointVisual;
const param_definition_1 = require("../../../mol-util/param-definition");
const units_visual_1 = require("../units-visual");
const points_1 = require("../../../mol-geo/geometry/points/points");
const points_builder_1 = require("../../../mol-geo/geometry/points/points-builder");
const linear_algebra_1 = require("../../../mol-math/linear-algebra");
const element_1 = require("./util/element");
const geometry_1 = require("../../../mol-math/geometry");
// avoiding namespace lookup improved performance in Chrome (Aug 2020)
const v3add = linear_algebra_1.Vec3.add;
exports.ElementPointParams = {
    ...units_visual_1.UnitsPointsParams,
    pointSizeAttenuation: param_definition_1.ParamDefinition.Boolean(false),
    ignoreHydrogens: param_definition_1.ParamDefinition.Boolean(false),
    ignoreHydrogensVariant: param_definition_1.ParamDefinition.Select('all', param_definition_1.ParamDefinition.arrayToOptions(['all', 'non-polar'])),
    traceOnly: param_definition_1.ParamDefinition.Boolean(false),
    stride: param_definition_1.ParamDefinition.Numeric(1, { min: 1, max: 100, step: 1 }),
};
// TODO size
function createElementPoint(ctx, unit, structure, theme, props, points) {
    // TODO sizeFactor
    const { child } = structure;
    if (child && !child.unitMap.get(unit.id))
        return points_1.Points.createEmpty(points);
    const elements = unit.elements;
    const n = elements.length;
    const builder = points_builder_1.PointsBuilder.create(n, n / 10, points);
    const p = (0, linear_algebra_1.Vec3)();
    const c = unit.conformation;
    const ignore = (0, element_1.makeElementIgnoreTest)(structure, unit, props);
    const center = (0, linear_algebra_1.Vec3)();
    let count = 0;
    if (ignore) {
        for (let i = 0; i < n; ++i) {
            if (ignore(elements[i]))
                continue;
            c.invariantPosition(elements[i], p);
            v3add(center, center, p);
            count += 1;
            builder.add(p[0], p[1], p[2], i);
        }
    }
    else {
        for (let i = 0; i < n; ++i) {
            c.invariantPosition(elements[i], p);
            v3add(center, center, p);
            count += 1;
            builder.add(p[0], p[1], p[2], i);
        }
    }
    const pt = builder.getPoints();
    if (count === 0)
        return pt;
    // re-use boundingSphere if it has not changed much
    let boundingSphere;
    linear_algebra_1.Vec3.scale(center, center, 1 / count);
    const oldBoundingSphere = points ? geometry_1.Sphere3D.clone(points.boundingSphere) : undefined;
    if (oldBoundingSphere && linear_algebra_1.Vec3.distance(center, oldBoundingSphere.center) / oldBoundingSphere.radius < 0.1) {
        boundingSphere = oldBoundingSphere;
    }
    else {
        boundingSphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, 1 * props.sizeFactor);
    }
    pt.setBoundingSphere(boundingSphere);
    return pt;
}
function ElementPointVisual(materialId) {
    return (0, units_visual_1.UnitsPointsVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.ElementPointParams),
        createGeometry: createElementPoint,
        createLocationIterator: element_1.ElementIterator.fromGroup,
        getLoci: element_1.getElementLoci,
        eachLocation: element_1.eachElement,
        setUpdateState: (state, newProps, currentProps) => {
            state.createGeometry = (newProps.ignoreHydrogens !== currentProps.ignoreHydrogens ||
                newProps.ignoreHydrogensVariant !== currentProps.ignoreHydrogensVariant ||
                newProps.traceOnly !== currentProps.traceOnly ||
                newProps.stride !== currentProps.stride);
        }
    }, materialId);
}
