// molstar/lib/commonjs/mol-repr/structure/visual/ellipsoid-mesh.js
"use strict";
/**
 * Copyright (c) 2019-2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EllipsoidMeshParams = void 0;
exports.EllipsoidMeshVisual = EllipsoidMeshVisual;
exports.createEllipsoidMesh = createEllipsoidMesh;
const param_definition_1 = require("../../../mol-util/param-definition");
const units_visual_1 = require("../../../mol-repr/structure/units-visual");
const element_1 = require("../../../mol-repr/structure/visual/util/element");
const structure_1 = require("../../../mol-model/structure");
const mesh_1 = require("../../../mol-geo/geometry/mesh/mesh");
const sphere_1 = require("../../../mol-geo/primitive/sphere");
const mesh_builder_1 = require("../../../mol-geo/geometry/mesh/mesh-builder");
const linear_algebra_1 = require("../../../mol-math/linear-algebra");
const common_1 = require("../../../mol-repr/structure/visual/util/common");
const ellipsoid_1 = require("../../../mol-geo/geometry/mesh/builder/ellipsoid");
const anisotropic_1 = require("../../../mol-model-formats/structure/property/anisotropic");
const common_2 = require("../../../mol-math/linear-algebra/3d/common");
const sphere_2 = require("../../../mol-geo/geometry/mesh/builder/sphere");
const geometry_1 = require("../../../mol-math/geometry");
const base_1 = require("../../../mol-geo/geometry/base");
const sorted_array_1 = require("../../../mol-data/int/sorted-array");
exports.EllipsoidMeshParams = {
    ...units_visual_1.UnitsMeshParams,
    sizeFactor: param_definition_1.ParamDefinition.Numeric(1, { min: 0, max: 10, step: 0.1 }),
    detail: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 3, step: 1 }, base_1.BaseGeometry.CustomQualityParamInfo),
    ignoreHydrogens: param_definition_1.ParamDefinition.Boolean(false),
};
function EllipsoidMeshVisual(materialId) {
    return (0, units_visual_1.UnitsMeshVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.EllipsoidMeshParams),
        createGeometry: createEllipsoidMesh,
        createLocationIterator: element_1.ElementIterator.fromGroup,
        getLoci: element_1.getElementLoci,
        eachLocation: element_1.eachElement,
        setUpdateState: (state, newProps, currentProps) => {
            state.createGeometry = (newProps.sizeFactor !== currentProps.sizeFactor ||
                newProps.detail !== currentProps.detail ||
                newProps.ignoreHydrogens !== currentProps.ignoreHydrogens);
        }
    }, materialId);
}
function createEllipsoidMesh(ctx, unit, structure, theme, props, mesh) {
    const { child } = structure;
    const childUnit = child === null || child === void 0 ? void 0 : child.unitMap.get(unit.id);
    if (child && !childUnit)
        return mesh_1.Mesh.createEmpty(mesh);
    const { detail, sizeFactor, ignoreHydrogens } = props;
    const { elements, model } = unit;
    const { atomicNumber } = unit.model.atomicHierarchy.derived.atom;
    const elementCount = elements.length;
    const vertexCount = elementCount * (0, sphere_1.sphereVertexCount)(detail);
    const builderState = mesh_builder_1.MeshBuilder.createState(vertexCount, vertexCount / 2, mesh);
    const atomSiteAnisotrop = anisotropic_1.AtomSiteAnisotrop.Provider.get(model);
    if (!atomSiteAnisotrop)
        return mesh_1.Mesh.createEmpty(mesh);
    const v = (0, linear_algebra_1.Vec3)();
    const mat = (0, linear_algebra_1.Mat3)();
    const eigvals = (0, linear_algebra_1.Vec3)();
    const eigvec1 = (0, linear_algebra_1.Vec3)();
    const eigvec2 = (0, linear_algebra_1.Vec3)();
    const { elementToAnsiotrop, data } = atomSiteAnisotrop;
    const { U } = data;
    const space = data._schema.U.space;
    const c = unit.conformation;
    const l = structure_1.StructureElement.Location.create(structure);
    l.unit = unit;
    for (let i = 0; i < elementCount; i++) {
        const ei = elements[i];
        const ai = elementToAnsiotrop[ei];
        if (ai === -1)
            continue;
        if (((!!childUnit && !sorted_array_1.SortedArray.has(childUnit.elements, ei))) ||
            (ignoreHydrogens && (0, common_1.isH)(atomicNumber, ei)))
            continue;
        l.element = ei;
        c.invariantPosition(ei, v);
        builderState.currentGroup = i;
        linear_algebra_1.Tensor.toMat3(mat, space, U.value(ai));
        linear_algebra_1.Mat3.symmtricFromLower(mat, mat);
        linear_algebra_1.Mat3.symmetricEigenvalues(eigvals, mat);
        linear_algebra_1.Mat3.eigenvector(eigvec1, mat, eigvals[1]);
        linear_algebra_1.Mat3.eigenvector(eigvec2, mat, eigvals[2]);
        for (let j = 0; j < 3; ++j) {
            // show 50% probability surface, needs sqrt as U matrix is in angstrom-squared
            // take abs of eigenvalue to avoid reflection
            // multiply by given size-factor
            eigvals[j] = sizeFactor * 1.5958 * Math.sqrt(Math.abs(eigvals[j]));
        }
        if ((0, common_2.equalEps)(eigvals[0], eigvals[1], linear_algebra_1.EPSILON) && (0, common_2.equalEps)(eigvals[1], eigvals[2], linear_algebra_1.EPSILON)) {
            (0, sphere_2.addSphere)(builderState, v, eigvals[0], detail);
        }
        else {
            (0, ellipsoid_1.addEllipsoid)(builderState, v, eigvec2, eigvec1, eigvals, detail);
        }
    }
    const m = mesh_builder_1.MeshBuilder.getMesh(builderState);
    const sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), (childUnit || unit).boundary.sphere, 1 * sizeFactor);
    m.setBoundingSphere(sphere);
    return m;
}
