// molstar/lib/commonjs/mol-repr/structure/visual/label-text.js
"use strict";
/**
 * Copyright (c) 2019-2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LabelTextVisual = exports.LabelTextParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var structure_1 = require("../../../mol-model/structure");
var text_builder_1 = require("../../../mol-geo/geometry/text/text-builder");
var complex_visual_1 = require("../complex-visual");
var element_1 = require("./util/element");
var names_1 = require("../../../mol-util/color/names");
var linear_algebra_1 = require("../../../mol-math/linear-algebra");
var boundary_helper_1 = require("../../../mol-math/geometry/boundary-helper");
exports.LabelTextParams = tslib_1.__assign(tslib_1.__assign({}, complex_visual_1.ComplexTextParams), { background: param_definition_1.ParamDefinition.Boolean(false), backgroundMargin: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 1, step: 0.01 }), backgroundColor: param_definition_1.ParamDefinition.Color(names_1.ColorNames.black), backgroundOpacity: param_definition_1.ParamDefinition.Numeric(0.5, { min: 0, max: 1, step: 0.01 }), borderWidth: param_definition_1.ParamDefinition.Numeric(0.25, { min: 0, max: 0.5, step: 0.01 }), level: param_definition_1.ParamDefinition.Select('residue', [['chain', 'Chain'], ['residue', 'Residue'], ['element', 'Element']], { isEssential: true }), chainScale: param_definition_1.ParamDefinition.Numeric(10, { min: 0, max: 20, step: 0.1 }), residueScale: param_definition_1.ParamDefinition.Numeric(1, { min: 0, max: 20, step: 0.1 }), elementScale: param_definition_1.ParamDefinition.Numeric(0.5, { min: 0, max: 20, step: 0.1 }) });
function LabelTextVisual(materialId) {
    return (0, complex_visual_1.ComplexTextVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.LabelTextParams),
        createGeometry: createLabelText,
        createLocationIterator: element_1.ElementIterator.fromStructure,
        getLoci: element_1.getSerialElementLoci,
        eachLocation: element_1.eachSerialElement,
        setUpdateState: function (state, newProps, currentProps) {
            state.createGeometry = (newProps.level !== currentProps.level ||
                (newProps.level === 'chain' && newProps.chainScale !== currentProps.chainScale) ||
                (newProps.level === 'residue' && newProps.residueScale !== currentProps.residueScale) ||
                (newProps.level === 'element' && newProps.elementScale !== currentProps.elementScale));
        }
    }, materialId);
}
exports.LabelTextVisual = LabelTextVisual;
function createLabelText(ctx, structure, theme, props, text) {
    switch (props.level) {
        case 'chain': return createChainText(ctx, structure, theme, props, text);
        case 'residue': return createResidueText(ctx, structure, theme, props, text);
        case 'element': return createElementText(ctx, structure, theme, props, text);
    }
}
//
var tmpVec = (0, linear_algebra_1.Vec3)();
var boundaryHelper = new boundary_helper_1.BoundaryHelper('98');
function createChainText(ctx, structure, theme, props, text) {
    var l = structure_1.StructureElement.Location.create(structure);
    var units = structure.units, serialMapping = structure.serialMapping;
    var _a = structure_1.StructureProperties.chain, auth_asym_id = _a.auth_asym_id, label_asym_id = _a.label_asym_id;
    var cumulativeUnitElementCount = serialMapping.cumulativeUnitElementCount;
    var count = units.length;
    var chainScale = props.chainScale;
    var builder = text_builder_1.TextBuilder.create(props, count, count / 2, text);
    for (var i = 0, il = units.length; i < il; ++i) {
        var unit = units[i];
        l.unit = unit;
        l.element = unit.elements[0];
        var _b = unit.lookup3d.boundary.sphere, center = _b.center, radius = _b.radius;
        linear_algebra_1.Vec3.transformMat4(tmpVec, center, unit.conformation.operator.matrix);
        var authId = auth_asym_id(l);
        var labelId = label_asym_id(l);
        var text_1 = authId === labelId ? labelId : "".concat(labelId, " [").concat(authId, "]");
        builder.add(text_1, tmpVec[0], tmpVec[1], tmpVec[2], radius, chainScale, cumulativeUnitElementCount[i]);
    }
    return builder.getText();
}
function createResidueText(ctx, structure, theme, props, text) {
    var l = structure_1.StructureElement.Location.create(structure);
    var units = structure.units, serialMapping = structure.serialMapping;
    var label_comp_id = structure_1.StructureProperties.atom.label_comp_id;
    var auth_seq_id = structure_1.StructureProperties.residue.auth_seq_id;
    var cumulativeUnitElementCount = serialMapping.cumulativeUnitElementCount;
    var count = structure.polymerResidueCount * 2;
    var residueScale = props.residueScale;
    var builder = text_builder_1.TextBuilder.create(props, count, count / 2, text);
    for (var i = 0, il = units.length; i < il; ++i) {
        var unit = units[i];
        var pos = unit.conformation.position;
        var elements = unit.elements;
        l.unit = unit;
        l.element = unit.elements[0];
        var residueIndex = unit.model.atomicHierarchy.residueAtomSegments.index;
        var groupOffset = cumulativeUnitElementCount[i];
        var j = 0;
        var jl = elements.length;
        while (j < jl) {
            var start = j, rI = residueIndex[elements[j]];
            j++;
            while (j < jl && residueIndex[elements[j]] === rI)
                j++;
            boundaryHelper.reset();
            for (var eI = start; eI < j; eI++) {
                pos(elements[eI], tmpVec);
                boundaryHelper.includePosition(tmpVec);
            }
            boundaryHelper.finishedIncludeStep();
            for (var eI = start; eI < j; eI++) {
                pos(elements[eI], tmpVec);
                boundaryHelper.radiusPosition(tmpVec);
            }
            l.element = elements[start];
            var _a = boundaryHelper.getSphere(), center = _a.center, radius = _a.radius;
            var authSeqId = auth_seq_id(l);
            var compId = label_comp_id(l);
            var text_2 = "".concat(compId, " ").concat(authSeqId);
            builder.add(text_2, center[0], center[1], center[2], radius, residueScale, groupOffset + start);
        }
    }
    return builder.getText();
}
function createElementText(ctx, structure, theme, props, text) {
    var l = structure_1.StructureElement.Location.create(structure);
    var units = structure.units, serialMapping = structure.serialMapping;
    var _a = structure_1.StructureProperties.atom, label_atom_id = _a.label_atom_id, label_alt_id = _a.label_alt_id;
    var cumulativeUnitElementCount = serialMapping.cumulativeUnitElementCount;
    var sizeTheme = theme.size;
    var count = structure.elementCount;
    var elementScale = props.elementScale;
    var builder = text_builder_1.TextBuilder.create(props, count, count / 2, text);
    for (var i = 0, il = units.length; i < il; ++i) {
        var unit = units[i];
        var pos = unit.conformation.position;
        var elements = unit.elements;
        l.unit = unit;
        var groupOffset = cumulativeUnitElementCount[i];
        for (var j = 0, _j = elements.length; j < _j; j++) {
            l.element = elements[j];
            pos(l.element, tmpVec);
            var atomId = label_atom_id(l);
            var altId = label_alt_id(l);
            var text_3 = altId ? "".concat(atomId, "%").concat(altId) : atomId;
            builder.add(text_3, tmpVec[0], tmpVec[1], tmpVec[2], sizeTheme.size(l), elementScale, groupOffset + j);
        }
    }
    return builder.getText();
}
