// molstar/lib/commonjs/mol-repr/structure/visual/molecular-surface-mesh.js
"use strict";
/**
 * Copyright (c) 2019-2022 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MolecularSurfaceMeshVisual = exports.MolecularSurfaceMeshParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var units_visual_1 = require("../units-visual");
var molecular_surface_1 = require("../../../mol-math/geometry/molecular-surface");
var mesh_1 = require("../../../mol-geo/geometry/mesh/mesh");
var molecular_surface_2 = require("./util/molecular-surface");
var algorithm_1 = require("../../../mol-geo/util/marching-cubes/algorithm");
var element_1 = require("./util/element");
var common_1 = require("./util/common");
var geometry_1 = require("../../../mol-math/geometry");
var color_smoothing_1 = require("../../../mol-geo/geometry/mesh/color-smoothing");
var base_1 = require("../../../mol-geo/geometry/base");
var mol_util_1 = require("../../../mol-util");
exports.MolecularSurfaceMeshParams = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, units_visual_1.UnitsMeshParams), molecular_surface_1.MolecularSurfaceCalculationParams), common_1.CommonSurfaceParams), base_1.ColorSmoothingParams);
//
function createMolecularSurfaceMesh(ctx, unit, structure, theme, props, mesh) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var _a, transform, field, idField, resolution, maxRadius, params, surface, iterations, sphere;
        return tslib_1.__generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, (0, molecular_surface_2.computeUnitMolecularSurface)(structure, unit, theme.size, props).runInContext(ctx.runtime)];
                case 1:
                    _a = _b.sent(), transform = _a.transform, field = _a.field, idField = _a.idField, resolution = _a.resolution, maxRadius = _a.maxRadius;
                    params = {
                        isoLevel: props.probeRadius,
                        scalarField: field,
                        idField: idField
                    };
                    return [4 /*yield*/, (0, algorithm_1.computeMarchingCubesMesh)(params, mesh).runAsChild(ctx.runtime)];
                case 2:
                    surface = _b.sent();
                    if (props.includeParent) {
                        iterations = Math.ceil(2 / props.resolution);
                        mesh_1.Mesh.smoothEdges(surface, { iterations: iterations, maxNewEdgeLength: Math.sqrt(2) });
                    }
                    mesh_1.Mesh.transform(surface, transform);
                    if (ctx.webgl && !ctx.webgl.isWebGL2) {
                        mesh_1.Mesh.uniformTriangleGroup(surface);
                        mol_util_1.ValueCell.updateIfChanged(surface.varyingGroup, false);
                    }
                    else {
                        mol_util_1.ValueCell.updateIfChanged(surface.varyingGroup, true);
                    }
                    sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, maxRadius);
                    surface.setBoundingSphere(sphere);
                    surface.meta.resolution = resolution;
                    return [2 /*return*/, surface];
            }
        });
    });
}
function MolecularSurfaceMeshVisual(materialId) {
    return (0, units_visual_1.UnitsMeshVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.MolecularSurfaceMeshParams),
        createGeometry: createMolecularSurfaceMesh,
        createLocationIterator: element_1.ElementIterator.fromGroup,
        getLoci: element_1.getElementLoci,
        eachLocation: element_1.eachElement,
        setUpdateState: function (state, newProps, currentProps) {
            if (newProps.resolution !== currentProps.resolution)
                state.createGeometry = true;
            if (newProps.probeRadius !== currentProps.probeRadius)
                state.createGeometry = true;
            if (newProps.probePositions !== currentProps.probePositions)
                state.createGeometry = true;
            if (newProps.ignoreHydrogens !== currentProps.ignoreHydrogens)
                state.createGeometry = true;
            if (newProps.traceOnly !== currentProps.traceOnly)
                state.createGeometry = true;
            if (newProps.includeParent !== currentProps.includeParent)
                state.createGeometry = true;
            if (newProps.smoothColors.name !== currentProps.smoothColors.name) {
                state.updateColor = true;
            }
            else if (newProps.smoothColors.name === 'on' && currentProps.smoothColors.name === 'on') {
                if (newProps.smoothColors.params.resolutionFactor !== currentProps.smoothColors.params.resolutionFactor)
                    state.updateColor = true;
                if (newProps.smoothColors.params.sampleStride !== currentProps.smoothColors.params.sampleStride)
                    state.updateColor = true;
            }
        },
        processValues: function (values, geometry, props, theme, webgl) {
            var _a = geometry.meta, resolution = _a.resolution, colorTexture = _a.colorTexture;
            var csp = (0, base_1.getColorSmoothingProps)(props.smoothColors, theme.color.preferSmoothing, resolution);
            if (csp) {
                (0, color_smoothing_1.applyMeshColorSmoothing)(values, csp.resolution, csp.stride, webgl, colorTexture);
                geometry.meta.colorTexture = values.tColorGrid.ref.value;
            }
        },
        dispose: function (geometry) {
            var _a;
            (_a = geometry.meta.colorTexture) === null || _a === void 0 ? void 0 : _a.destroy();
        }
    }, materialId);
}
exports.MolecularSurfaceMeshVisual = MolecularSurfaceMeshVisual;
