// molstar/lib/commonjs/mol-repr/structure/visual/nucleotide-block-mesh.js
"use strict";
/**
 * Copyright (c) 2018-2019 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NucleotideBlockVisual = exports.NucleotideBlockParams = exports.DefaultNucleotideBlockMeshProps = exports.NucleotideBlockMeshParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var linear_algebra_1 = require("../../../mol-math/linear-algebra");
var box_1 = require("../../../mol-geo/primitive/box");
var structure_1 = require("../../../mol-model/structure");
var mesh_1 = require("../../../mol-geo/geometry/mesh/mesh");
var mesh_builder_1 = require("../../../mol-geo/geometry/mesh/mesh-builder");
var int_1 = require("../../../mol-data/int");
var types_1 = require("../../../mol-model/structure/model/types");
var cylinder_1 = require("../../../mol-geo/geometry/mesh/builder/cylinder");
var units_visual_1 = require("../units-visual");
var nucleotide_1 = require("./util/nucleotide");
var base_1 = require("../../../mol-geo/geometry/base");
var geometry_1 = require("../../../mol-math/geometry");
// TODO support blocks for multiple locations (including from microheterogeneity)
var p1 = (0, linear_algebra_1.Vec3)();
var p2 = (0, linear_algebra_1.Vec3)();
var p3 = (0, linear_algebra_1.Vec3)();
var p4 = (0, linear_algebra_1.Vec3)();
var p5 = (0, linear_algebra_1.Vec3)();
var p6 = (0, linear_algebra_1.Vec3)();
var v12 = (0, linear_algebra_1.Vec3)();
var v34 = (0, linear_algebra_1.Vec3)();
var vC = (0, linear_algebra_1.Vec3)();
var center = (0, linear_algebra_1.Vec3)();
var t = linear_algebra_1.Mat4.identity();
var sVec = (0, linear_algebra_1.Vec3)();
var box = (0, box_1.Box)();
exports.NucleotideBlockMeshParams = {
    sizeFactor: param_definition_1.ParamDefinition.Numeric(0.2, { min: 0, max: 10, step: 0.01 }),
    radialSegments: param_definition_1.ParamDefinition.Numeric(16, { min: 2, max: 56, step: 2 }, base_1.BaseGeometry.CustomQualityParamInfo),
};
exports.DefaultNucleotideBlockMeshProps = param_definition_1.ParamDefinition.getDefaultValues(exports.NucleotideBlockMeshParams);
function createNucleotideBlockMesh(ctx, unit, structure, theme, props, mesh) {
    if (!structure_1.Unit.isAtomic(unit))
        return mesh_1.Mesh.createEmpty(mesh);
    var nucleotideElementCount = unit.nucleotideElements.length;
    if (!nucleotideElementCount)
        return mesh_1.Mesh.createEmpty(mesh);
    var sizeFactor = props.sizeFactor, radialSegments = props.radialSegments;
    var vertexCount = nucleotideElementCount * (box.vertices.length / 3 + radialSegments * 2);
    var builderState = mesh_builder_1.MeshBuilder.createState(vertexCount, vertexCount / 4, mesh);
    var elements = unit.elements, model = unit.model;
    var _a = model.atomicHierarchy, chainAtomSegments = _a.chainAtomSegments, residueAtomSegments = _a.residueAtomSegments, atoms = _a.atoms, atomicIndex = _a.index;
    var _b = model.atomicHierarchy.derived.residue, moleculeType = _b.moleculeType, traceElementIndex = _b.traceElementIndex;
    var label_comp_id = atoms.label_comp_id;
    var pos = unit.conformation.invariantPosition;
    var chainIt = int_1.Segmentation.transientSegments(chainAtomSegments, elements);
    var residueIt = int_1.Segmentation.transientSegments(residueAtomSegments, elements);
    var cylinderProps = { radiusTop: 1 * sizeFactor, radiusBottom: 1 * sizeFactor, radialSegments: radialSegments, bottomCap: true };
    var i = 0;
    while (chainIt.hasNext) {
        residueIt.setSegment(chainIt.move());
        while (residueIt.hasNext) {
            var residueIndex = residueIt.move().index;
            if ((0, types_1.isNucleic)(moleculeType[residueIndex])) {
                var compId = label_comp_id.value(residueAtomSegments.offsets[residueIndex]);
                var idx1 = -1, idx2 = -1, idx3 = -1, idx4 = -1, idx5 = -1, idx6 = -1;
                var width = 4.5, depth = 2.5 * sizeFactor;
                var height = 4.5;
                var isPurine = (0, types_1.isPurineBase)(compId);
                var isPyrimidine = (0, types_1.isPyrimidineBase)(compId);
                if (!isPurine && !isPyrimidine) {
                    // detect Purine or Pyrimidin based on geometry
                    var idxC4 = atomicIndex.findAtomOnResidue(residueIndex, 'C4');
                    var idxN9 = atomicIndex.findAtomOnResidue(residueIndex, 'N9');
                    if (idxC4 !== -1 && idxN9 !== -1 && linear_algebra_1.Vec3.distance(pos(idxC4, p1), pos(idxN9, p2)) < 1.6) {
                        isPurine = true;
                    }
                    else {
                        isPyrimidine = true;
                    }
                }
                if (isPurine) {
                    height = 4.5;
                    idx1 = atomicIndex.findAtomOnResidue(residueIndex, 'N1');
                    idx2 = atomicIndex.findAtomOnResidue(residueIndex, 'C4');
                    idx3 = atomicIndex.findAtomOnResidue(residueIndex, 'C6');
                    idx4 = atomicIndex.findAtomOnResidue(residueIndex, 'C2');
                    idx5 = atomicIndex.findAtomOnResidue(residueIndex, 'N9');
                    idx6 = traceElementIndex[residueIndex];
                }
                else if (isPyrimidine) {
                    height = 3.0;
                    idx1 = atomicIndex.findAtomOnResidue(residueIndex, 'N3');
                    idx2 = atomicIndex.findAtomOnResidue(residueIndex, 'C6');
                    idx3 = atomicIndex.findAtomOnResidue(residueIndex, 'C4');
                    idx4 = atomicIndex.findAtomOnResidue(residueIndex, 'C2');
                    idx5 = atomicIndex.findAtomOnResidue(residueIndex, 'N1');
                    if (idx5 === -1) {
                        // modified ring, e.g. DZ
                        idx5 = atomicIndex.findAtomOnResidue(residueIndex, 'C1');
                    }
                    idx6 = traceElementIndex[residueIndex];
                }
                if (idx5 !== -1 && idx6 !== -1) {
                    pos(idx5, p5);
                    pos(idx6, p6);
                    builderState.currentGroup = i;
                    (0, cylinder_1.addCylinder)(builderState, p5, p6, 1, cylinderProps);
                    if (idx1 !== -1 && idx2 !== -1 && idx3 !== -1 && idx4 !== -1) {
                        pos(idx1, p1);
                        pos(idx2, p2);
                        pos(idx3, p3);
                        pos(idx4, p4);
                        linear_algebra_1.Vec3.normalize(v12, linear_algebra_1.Vec3.sub(v12, p2, p1));
                        linear_algebra_1.Vec3.normalize(v34, linear_algebra_1.Vec3.sub(v34, p4, p3));
                        linear_algebra_1.Vec3.normalize(vC, linear_algebra_1.Vec3.cross(vC, v12, v34));
                        linear_algebra_1.Mat4.targetTo(t, p1, p2, vC);
                        linear_algebra_1.Vec3.scaleAndAdd(center, p1, v12, height / 2 - 0.2);
                        linear_algebra_1.Mat4.scale(t, t, linear_algebra_1.Vec3.set(sVec, width, depth, height));
                        linear_algebra_1.Mat4.setTranslation(t, center);
                        mesh_builder_1.MeshBuilder.addPrimitive(builderState, t, box);
                    }
                }
                ++i;
            }
        }
    }
    var m = mesh_builder_1.MeshBuilder.getMesh(builderState);
    var sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, 1 * props.sizeFactor);
    m.setBoundingSphere(sphere);
    return m;
}
exports.NucleotideBlockParams = tslib_1.__assign(tslib_1.__assign({}, units_visual_1.UnitsMeshParams), exports.NucleotideBlockMeshParams);
function NucleotideBlockVisual(materialId) {
    return (0, units_visual_1.UnitsMeshVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.NucleotideBlockParams),
        createGeometry: createNucleotideBlockMesh,
        createLocationIterator: nucleotide_1.NucleotideLocationIterator.fromGroup,
        getLoci: nucleotide_1.getNucleotideElementLoci,
        eachLocation: nucleotide_1.eachNucleotideElement,
        setUpdateState: function (state, newProps, currentProps) {
            state.createGeometry = (newProps.sizeFactor !== currentProps.sizeFactor ||
                newProps.radialSegments !== currentProps.radialSegments);
        }
    }, materialId);
}
exports.NucleotideBlockVisual = NucleotideBlockVisual;
