// molstar/lib/commonjs/mol-repr/structure/visual/polymer-direction-wedge.js
"use strict";
/**
 * Copyright (c) 2018 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolymerDirectionVisual = exports.PolymerDirectionParams = exports.DefaultPolymerDirectionWedgeProps = exports.PolymerDirectionWedgeParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var linear_algebra_1 = require("../../../mol-math/linear-algebra");
var wedge_1 = require("../../../mol-geo/primitive/wedge");
var mesh_1 = require("../../../mol-geo/geometry/mesh/mesh");
var mesh_builder_1 = require("../../../mol-geo/geometry/mesh/mesh-builder");
var polymer_1 = require("./util/polymer");
var types_1 = require("../../../mol-model/structure/model/types");
var units_visual_1 = require("../units-visual");
var geometry_1 = require("../../../mol-math/geometry");
var t = linear_algebra_1.Mat4.identity();
var sVec = linear_algebra_1.Vec3.zero();
var n0 = linear_algebra_1.Vec3.zero();
var n1 = linear_algebra_1.Vec3.zero();
var upVec = linear_algebra_1.Vec3.zero();
var depthFactor = 4;
var widthFactor = 4;
var heightFactor = 6;
var wedge = (0, wedge_1.Wedge)();
exports.PolymerDirectionWedgeParams = {
    sizeFactor: param_definition_1.ParamDefinition.Numeric(0.2, { min: 0, max: 10, step: 0.01 }),
};
exports.DefaultPolymerDirectionWedgeProps = param_definition_1.ParamDefinition.getDefaultValues(exports.PolymerDirectionWedgeParams);
function createPolymerDirectionWedgeMesh(ctx, unit, structure, theme, props, mesh) {
    var polymerElementCount = unit.polymerElements.length;
    if (!polymerElementCount)
        return mesh_1.Mesh.createEmpty(mesh);
    var sizeFactor = props.sizeFactor;
    var vertexCount = polymerElementCount * 24;
    var builderState = mesh_builder_1.MeshBuilder.createState(vertexCount, vertexCount / 10, mesh);
    var linearSegments = 1;
    var state = (0, polymer_1.createCurveSegmentState)(linearSegments);
    var normalVectors = state.normalVectors, binormalVectors = state.binormalVectors;
    var i = 0;
    var polymerTraceIt = (0, polymer_1.PolymerTraceIterator)(unit, structure);
    while (polymerTraceIt.hasNext) {
        var v = polymerTraceIt.move();
        builderState.currentGroup = i;
        var isNucleicType = (0, types_1.isNucleic)(v.moleculeType);
        var isSheet = types_1.SecondaryStructureType.is(v.secStrucType, 4 /* SecondaryStructureType.Flag.Beta */);
        var tension = (isNucleicType || isSheet) ? 0.5 : 0.9;
        var shift = isNucleicType ? 0.3 : 0.5;
        (0, polymer_1.interpolateCurveSegment)(state, v, tension, shift);
        if ((isSheet && !v.secStrucLast) || !isSheet) {
            var size = theme.size.size(v.center) * sizeFactor;
            var depth = depthFactor * size;
            var width = widthFactor * size;
            var height = heightFactor * size;
            var vectors = isNucleicType ? binormalVectors : normalVectors;
            linear_algebra_1.Vec3.fromArray(n0, vectors, 0);
            linear_algebra_1.Vec3.fromArray(n1, vectors, 3);
            linear_algebra_1.Vec3.normalize(upVec, linear_algebra_1.Vec3.add(upVec, n0, n1));
            linear_algebra_1.Mat4.targetTo(t, v.p3, v.p1, upVec);
            linear_algebra_1.Mat4.mul(t, t, linear_algebra_1.Mat4.rotY90Z180);
            linear_algebra_1.Mat4.scale(t, t, linear_algebra_1.Vec3.set(sVec, height, width, depth));
            linear_algebra_1.Mat4.setTranslation(t, v.p2);
            mesh_builder_1.MeshBuilder.addPrimitive(builderState, t, wedge);
        }
        ++i;
    }
    var m = mesh_builder_1.MeshBuilder.getMesh(builderState);
    var sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, 1 * props.sizeFactor);
    m.setBoundingSphere(sphere);
    return m;
}
exports.PolymerDirectionParams = tslib_1.__assign(tslib_1.__assign({}, units_visual_1.UnitsMeshParams), exports.PolymerDirectionWedgeParams);
function PolymerDirectionVisual(materialId) {
    return (0, units_visual_1.UnitsMeshVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.PolymerDirectionParams),
        createGeometry: createPolymerDirectionWedgeMesh,
        createLocationIterator: polymer_1.PolymerLocationIterator.fromGroup,
        getLoci: polymer_1.getPolymerElementLoci,
        eachLocation: polymer_1.eachPolymerElement,
        setUpdateState: function (state, newProps, currentProps) {
            state.createGeometry = (newProps.sizeFactor !== currentProps.sizeFactor);
        }
    }, materialId);
}
exports.PolymerDirectionVisual = PolymerDirectionVisual;
