// molstar/lib/commonjs/mol-repr/structure/visual/polymer-tube-mesh.js
"use strict";
/**
 * Copyright (c) 2019-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolymerTubeVisual = exports.PolymerTubeParams = exports.DefaultPolymerTubeMeshProps = exports.PolymerTubeMeshParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var mesh_1 = require("../../../mol-geo/geometry/mesh/mesh");
var mesh_builder_1 = require("../../../mol-geo/geometry/mesh/mesh-builder");
var polymer_1 = require("./util/polymer");
var types_1 = require("../../../mol-model/structure/model/types");
var tube_1 = require("../../../mol-geo/geometry/mesh/builder/tube");
var units_visual_1 = require("../units-visual");
var sheet_1 = require("../../../mol-geo/geometry/mesh/builder/sheet");
var ribbon_1 = require("../../../mol-geo/geometry/mesh/builder/ribbon");
var linear_algebra_1 = require("../../../mol-math/linear-algebra");
var sphere_1 = require("../../../mol-geo/geometry/mesh/builder/sphere");
var base_1 = require("../../../mol-geo/geometry/base");
var geometry_1 = require("../../../mol-math/geometry");
exports.PolymerTubeMeshParams = {
    sizeFactor: param_definition_1.ParamDefinition.Numeric(0.2, { min: 0, max: 10, step: 0.01 }),
    detail: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 3, step: 1 }, base_1.BaseGeometry.CustomQualityParamInfo),
    linearSegments: param_definition_1.ParamDefinition.Numeric(8, { min: 1, max: 48, step: 1 }, base_1.BaseGeometry.CustomQualityParamInfo),
    radialSegments: param_definition_1.ParamDefinition.Numeric(16, { min: 2, max: 56, step: 2 }, base_1.BaseGeometry.CustomQualityParamInfo),
};
exports.DefaultPolymerTubeMeshProps = param_definition_1.ParamDefinition.getDefaultValues(exports.PolymerTubeMeshParams);
var tmpV1 = (0, linear_algebra_1.Vec3)();
function createPolymerTubeMesh(ctx, unit, structure, theme, props, mesh) {
    var polymerElementCount = unit.polymerElements.length;
    if (!polymerElementCount)
        return mesh_1.Mesh.createEmpty(mesh);
    var sizeFactor = props.sizeFactor, detail = props.detail, linearSegments = props.linearSegments, radialSegments = props.radialSegments;
    var vertexCount = linearSegments * radialSegments * polymerElementCount + (radialSegments + 1) * polymerElementCount * 2;
    var builderState = mesh_builder_1.MeshBuilder.createState(vertexCount, vertexCount / 10, mesh);
    var state = (0, polymer_1.createCurveSegmentState)(linearSegments);
    var curvePoints = state.curvePoints, normalVectors = state.normalVectors, binormalVectors = state.binormalVectors, widthValues = state.widthValues, heightValues = state.heightValues;
    var i = 0;
    var polymerTraceIt = (0, polymer_1.PolymerTraceIterator)(unit, structure, { ignoreSecondaryStructure: true });
    while (polymerTraceIt.hasNext) {
        var v = polymerTraceIt.move();
        builderState.currentGroup = i;
        var isNucleicType = (0, types_1.isNucleic)(v.moleculeType);
        var shift = isNucleicType ? polymer_1.NucleicShift : polymer_1.StandardShift;
        (0, polymer_1.interpolateCurveSegment)(state, v, polymer_1.StandardTension, shift);
        var startCap = v.coarseBackboneFirst || v.first;
        var endCap = v.coarseBackboneLast || v.last;
        var s0 = theme.size.size(v.centerPrev) * sizeFactor;
        var s1 = theme.size.size(v.center) * sizeFactor;
        var s2 = theme.size.size(v.centerNext) * sizeFactor;
        (0, polymer_1.interpolateSizes)(state, s0, s1, s2, s0, s1, s2, shift);
        var segmentCount = linearSegments;
        if (v.initial) {
            segmentCount = Math.max(Math.round(linearSegments * shift), 1);
            var offset = linearSegments - segmentCount;
            curvePoints.copyWithin(0, offset * 3);
            binormalVectors.copyWithin(0, offset * 3);
            normalVectors.copyWithin(0, offset * 3);
            widthValues.copyWithin(0, offset * 3);
            heightValues.copyWithin(0, offset * 3);
            linear_algebra_1.Vec3.fromArray(tmpV1, curvePoints, 3);
            linear_algebra_1.Vec3.normalize(tmpV1, linear_algebra_1.Vec3.sub(tmpV1, v.p2, tmpV1));
            linear_algebra_1.Vec3.scaleAndAdd(tmpV1, v.p2, tmpV1, s1 * polymer_1.OverhangFactor);
            linear_algebra_1.Vec3.toArray(tmpV1, curvePoints, 0);
        }
        else if (v.final) {
            segmentCount = Math.max(Math.round(linearSegments * (1 - shift)), 1);
            linear_algebra_1.Vec3.fromArray(tmpV1, curvePoints, segmentCount * 3 - 3);
            linear_algebra_1.Vec3.normalize(tmpV1, linear_algebra_1.Vec3.sub(tmpV1, v.p2, tmpV1));
            linear_algebra_1.Vec3.scaleAndAdd(tmpV1, v.p2, tmpV1, s1 * polymer_1.OverhangFactor);
            linear_algebra_1.Vec3.toArray(tmpV1, curvePoints, segmentCount * 3);
        }
        if (v.initial === true && v.final === true) {
            (0, sphere_1.addSphere)(builderState, v.p2, s1 * 2, detail);
        }
        else if (radialSegments === 2) {
            (0, ribbon_1.addRibbon)(builderState, curvePoints, normalVectors, binormalVectors, segmentCount, widthValues, heightValues, 0);
        }
        else if (radialSegments === 4) {
            (0, sheet_1.addSheet)(builderState, curvePoints, normalVectors, binormalVectors, segmentCount, widthValues, heightValues, 0, startCap, endCap);
        }
        else {
            (0, tube_1.addTube)(builderState, curvePoints, normalVectors, binormalVectors, segmentCount, radialSegments, widthValues, heightValues, startCap, endCap, 'elliptical');
        }
        ++i;
    }
    var m = mesh_builder_1.MeshBuilder.getMesh(builderState);
    var sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, 1 * props.sizeFactor);
    m.setBoundingSphere(sphere);
    return m;
}
exports.PolymerTubeParams = tslib_1.__assign(tslib_1.__assign({}, units_visual_1.UnitsMeshParams), exports.PolymerTubeMeshParams);
function PolymerTubeVisual(materialId) {
    return (0, units_visual_1.UnitsMeshVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.PolymerTubeParams),
        createGeometry: createPolymerTubeMesh,
        createLocationIterator: polymer_1.PolymerLocationIterator.fromGroup,
        getLoci: polymer_1.getPolymerElementLoci,
        eachLocation: polymer_1.eachPolymerElement,
        setUpdateState: function (state, newProps, currentProps) {
            state.createGeometry = (newProps.sizeFactor !== currentProps.sizeFactor ||
                newProps.detail !== currentProps.detail ||
                newProps.linearSegments !== currentProps.linearSegments ||
                newProps.radialSegments !== currentProps.radialSegments);
        }
    }, materialId);
}
exports.PolymerTubeVisual = PolymerTubeVisual;
