// molstar/lib/commonjs/mol-repr/structure/visual/polymer-trace-mesh.js
"use strict";
/**
 * Copyright (c) 2018-2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolymerTraceVisual = exports.PolymerTraceParams = exports.DefaultPolymerTraceMeshProps = exports.PolymerTraceMeshParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var structure_1 = require("../../../mol-model/structure");
var mesh_1 = require("../../../mol-geo/geometry/mesh/mesh");
var mesh_builder_1 = require("../../../mol-geo/geometry/mesh/mesh-builder");
var polymer_1 = require("./util/polymer");
var types_1 = require("../../../mol-model/structure/model/types");
var sheet_1 = require("../../../mol-geo/geometry/mesh/builder/sheet");
var tube_1 = require("../../../mol-geo/geometry/mesh/builder/tube");
var units_visual_1 = require("../units-visual");
var secondary_structure_1 = require("../../../mol-model-props/computed/secondary-structure");
var ribbon_1 = require("../../../mol-geo/geometry/mesh/builder/ribbon");
var sphere_1 = require("../../../mol-geo/geometry/mesh/builder/sphere");
var linear_algebra_1 = require("../../../mol-math/linear-algebra");
var base_1 = require("../../../mol-geo/geometry/base");
var geometry_1 = require("../../../mol-math/geometry");
exports.PolymerTraceMeshParams = {
    sizeFactor: param_definition_1.ParamDefinition.Numeric(0.2, { min: 0, max: 10, step: 0.01 }),
    aspectRatio: param_definition_1.ParamDefinition.Numeric(5, { min: 0.1, max: 10, step: 0.1 }),
    arrowFactor: param_definition_1.ParamDefinition.Numeric(1.5, { min: 0, max: 3, step: 0.1 }, { description: 'Size factor for sheet arrows' }),
    tubularHelices: param_definition_1.ParamDefinition.Boolean(false, { description: 'Draw alpha helices as tubes' }),
    helixProfile: param_definition_1.ParamDefinition.Select('elliptical', param_definition_1.ParamDefinition.arrayToOptions(['elliptical', 'rounded', 'square']), { description: 'Protein and nucleic helix trace profile' }),
    detail: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 3, step: 1 }, base_1.BaseGeometry.CustomQualityParamInfo),
    linearSegments: param_definition_1.ParamDefinition.Numeric(8, { min: 1, max: 48, step: 1 }, base_1.BaseGeometry.CustomQualityParamInfo),
    radialSegments: param_definition_1.ParamDefinition.Numeric(16, { min: 2, max: 56, step: 2 }, base_1.BaseGeometry.CustomQualityParamInfo)
};
exports.DefaultPolymerTraceMeshProps = param_definition_1.ParamDefinition.getDefaultValues(exports.PolymerTraceMeshParams);
var tmpV1 = (0, linear_algebra_1.Vec3)();
function createPolymerTraceMesh(ctx, unit, structure, theme, props, mesh) {
    var polymerElementCount = unit.polymerElements.length;
    if (!polymerElementCount)
        return mesh_1.Mesh.createEmpty(mesh);
    var sizeFactor = props.sizeFactor, detail = props.detail, linearSegments = props.linearSegments, radialSegments = props.radialSegments, aspectRatio = props.aspectRatio, arrowFactor = props.arrowFactor, tubularHelices = props.tubularHelices, helixProfile = props.helixProfile;
    var vertexCount = linearSegments * radialSegments * polymerElementCount + (radialSegments + 1) * polymerElementCount * 2;
    var builderState = mesh_builder_1.MeshBuilder.createState(vertexCount, vertexCount / 10, mesh);
    var isCoarse = structure_1.Unit.isCoarse(unit);
    var state = (0, polymer_1.createCurveSegmentState)(linearSegments);
    var curvePoints = state.curvePoints, normalVectors = state.normalVectors, binormalVectors = state.binormalVectors, widthValues = state.widthValues, heightValues = state.heightValues;
    var i = 0;
    var polymerTraceIt = (0, polymer_1.PolymerTraceIterator)(unit, structure, { ignoreSecondaryStructure: false, useHelixOrientation: tubularHelices });
    while (polymerTraceIt.hasNext) {
        var v = polymerTraceIt.move();
        builderState.currentGroup = i;
        var isNucleicType = (0, types_1.isNucleic)(v.moleculeType);
        var isSheet = types_1.SecondaryStructureType.is(v.secStrucType, 4 /* SecondaryStructureType.Flag.Beta */);
        var isHelix = types_1.SecondaryStructureType.is(v.secStrucType, 2 /* SecondaryStructureType.Flag.Helix */);
        var tension = isHelix && !tubularHelices ? polymer_1.HelixTension : polymer_1.StandardTension;
        var shift = isNucleicType ? polymer_1.NucleicShift : polymer_1.StandardShift;
        (0, polymer_1.interpolateCurveSegment)(state, v, tension, shift);
        var w0 = theme.size.size(v.centerPrev) * sizeFactor;
        var w1 = theme.size.size(v.center) * sizeFactor;
        var w2 = theme.size.size(v.centerNext) * sizeFactor;
        if (isCoarse) {
            w0 *= aspectRatio / 2;
            w1 *= aspectRatio / 2;
            w2 *= aspectRatio / 2;
        }
        var startCap = v.secStrucFirst || v.coarseBackboneFirst || v.first;
        var endCap = v.secStrucLast || v.coarseBackboneLast || v.last;
        var segmentCount = linearSegments;
        if (v.initial) {
            segmentCount = Math.max(Math.round(linearSegments * shift), 1);
            var offset = linearSegments - segmentCount;
            curvePoints.copyWithin(0, offset * 3);
            binormalVectors.copyWithin(0, offset * 3);
            normalVectors.copyWithin(0, offset * 3);
            linear_algebra_1.Vec3.fromArray(tmpV1, curvePoints, 3);
            linear_algebra_1.Vec3.normalize(tmpV1, linear_algebra_1.Vec3.sub(tmpV1, v.p2, tmpV1));
            linear_algebra_1.Vec3.scaleAndAdd(tmpV1, v.p2, tmpV1, w1 * polymer_1.OverhangFactor);
            linear_algebra_1.Vec3.toArray(tmpV1, curvePoints, 0);
        }
        else if (v.final) {
            segmentCount = Math.max(Math.round(linearSegments * (1 - shift)), 1);
            linear_algebra_1.Vec3.fromArray(tmpV1, curvePoints, segmentCount * 3 - 3);
            linear_algebra_1.Vec3.normalize(tmpV1, linear_algebra_1.Vec3.sub(tmpV1, v.p2, tmpV1));
            linear_algebra_1.Vec3.scaleAndAdd(tmpV1, v.p2, tmpV1, w1 * polymer_1.OverhangFactor);
            linear_algebra_1.Vec3.toArray(tmpV1, curvePoints, segmentCount * 3);
        }
        if (v.initial === true && v.final === true) {
            (0, sphere_1.addSphere)(builderState, v.p2, w1 * 2, detail);
        }
        else if (isSheet) {
            var h0 = w0 * aspectRatio;
            var h1 = w1 * aspectRatio;
            var h2 = w2 * aspectRatio;
            var arrowHeight = v.secStrucLast ? h1 * arrowFactor : 0;
            (0, polymer_1.interpolateSizes)(state, w0, w1, w2, h0, h1, h2, shift);
            if (radialSegments === 2) {
                (0, ribbon_1.addRibbon)(builderState, curvePoints, normalVectors, binormalVectors, segmentCount, widthValues, heightValues, arrowHeight);
            }
            else {
                (0, sheet_1.addSheet)(builderState, curvePoints, normalVectors, binormalVectors, segmentCount, widthValues, heightValues, arrowHeight, startCap, endCap);
            }
        }
        else {
            var h0 = void 0, h1 = void 0, h2 = void 0;
            if (isHelix && !v.isCoarseBackbone) {
                if (tubularHelices) {
                    w0 *= aspectRatio * 1.5;
                    w1 *= aspectRatio * 1.5;
                    w2 *= aspectRatio * 1.5;
                    h0 = w0;
                    h1 = w1;
                    h2 = w2;
                }
                else {
                    h0 = w0 * aspectRatio;
                    h1 = w1 * aspectRatio;
                    h2 = w2 * aspectRatio;
                }
            }
            else if (isNucleicType && !v.isCoarseBackbone) {
                h0 = w0 * aspectRatio;
                h1 = w1 * aspectRatio;
                h2 = w2 * aspectRatio;
            }
            else {
                h0 = w0;
                h1 = w1;
                h2 = w2;
            }
            (0, polymer_1.interpolateSizes)(state, w0, w1, w2, h0, h1, h2, shift);
            var _a = isNucleicType && !v.isCoarseBackbone ? [binormalVectors, normalVectors] : [normalVectors, binormalVectors], normals = _a[0], binormals = _a[1];
            if (isNucleicType && !v.isCoarseBackbone) {
                // TODO: find a cleaner way to swap normal and binormal for nucleic types
                for (var i_1 = 0, il = normals.length; i_1 < il; i_1++)
                    normals[i_1] *= -1;
            }
            if (radialSegments === 2) {
                if (isNucleicType && !v.isCoarseBackbone) {
                    (0, ribbon_1.addRibbon)(builderState, curvePoints, normals, binormals, segmentCount, heightValues, widthValues, 0);
                }
                else {
                    (0, ribbon_1.addRibbon)(builderState, curvePoints, normals, binormals, segmentCount, widthValues, heightValues, 0);
                }
            }
            else if (radialSegments === 4) {
                (0, sheet_1.addSheet)(builderState, curvePoints, normals, binormals, segmentCount, widthValues, heightValues, 0, startCap, endCap);
            }
            else if (h1 === w1) {
                (0, tube_1.addTube)(builderState, curvePoints, normals, binormals, segmentCount, radialSegments, widthValues, heightValues, startCap, endCap, 'elliptical');
            }
            else if (helixProfile === 'square') {
                (0, sheet_1.addSheet)(builderState, curvePoints, normals, binormals, segmentCount, widthValues, heightValues, 0, startCap, endCap);
            }
            else {
                (0, tube_1.addTube)(builderState, curvePoints, normals, binormals, segmentCount, radialSegments, widthValues, heightValues, startCap, endCap, helixProfile);
            }
        }
        ++i;
    }
    var m = mesh_builder_1.MeshBuilder.getMesh(builderState);
    var sphere = geometry_1.Sphere3D.expand((0, geometry_1.Sphere3D)(), unit.boundary.sphere, 1 * props.sizeFactor);
    m.setBoundingSphere(sphere);
    return m;
}
exports.PolymerTraceParams = tslib_1.__assign(tslib_1.__assign({}, units_visual_1.UnitsMeshParams), exports.PolymerTraceMeshParams);
function PolymerTraceVisual(materialId) {
    return (0, units_visual_1.UnitsMeshVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.PolymerTraceParams),
        createGeometry: createPolymerTraceMesh,
        createLocationIterator: polymer_1.PolymerLocationIterator.fromGroup,
        getLoci: polymer_1.getPolymerElementLoci,
        eachLocation: polymer_1.eachPolymerElement,
        setUpdateState: function (state, newProps, currentProps, newTheme, currentTheme, newStructureGroup, currentStructureGroup) {
            state.createGeometry = (newProps.sizeFactor !== currentProps.sizeFactor ||
                newProps.tubularHelices !== currentProps.tubularHelices ||
                newProps.detail !== currentProps.detail ||
                newProps.linearSegments !== currentProps.linearSegments ||
                newProps.radialSegments !== currentProps.radialSegments ||
                newProps.aspectRatio !== currentProps.aspectRatio ||
                newProps.arrowFactor !== currentProps.arrowFactor ||
                newProps.helixProfile !== currentProps.helixProfile);
            var secondaryStructureHash = secondary_structure_1.SecondaryStructureProvider.get(newStructureGroup.structure).version;
            if (state.info.secondaryStructureHash !== secondaryStructureHash) {
                state.createGeometry = state.info.secondaryStructureHash !== undefined;
                state.info.secondaryStructureHash = secondaryStructureHash;
            }
        }
    }, materialId);
}
exports.PolymerTraceVisual = PolymerTraceVisual;
