// molstar/lib/commonjs/extensions/mvs/tree/mvs/mvs-builder.js
"use strict";
/**
 * Copyright (c) 2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Adam Midlik <midlik@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Representation = exports.Component = exports.Structure = exports.Parse = exports.Download = exports.Root = void 0;
exports.createMVSBuilder = createMVSBuilder;
exports.builderDemo = builderDemo;
const object_1 = require("../../../../mol-util/object");
const mvs_data_1 = require("../../mvs-data");
const mvs_defaults_1 = require("./mvs-defaults");
const mvs_tree_1 = require("./mvs-tree");
/** Create a new MolViewSpec builder containing only a root node. Example of MVS builder usage:
 *
 * ```
 * const builder = createMVSBuilder();
 * builder.canvas({ background_color: 'white' });
 * const struct = builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og2_updated.cif' }).parse({ format: 'mmcif' }).modelStructure();
 * struct.component().representation().color({ color: '#3050F8' });
 * console.log(JSON.stringify(builder.getState()));
 * ```
 */
function createMVSBuilder() {
    return new Root();
}
/** Base class for MVS builder pointing to anything */
class _Base {
    constructor(_root, _node) {
        this._root = _root;
        this._node = _node;
    }
    /** Create a new node, append as child to current _node, and return the new node */
    addChild(kind, params) {
        var _a;
        var _b;
        const allowedParamNames = Object.keys(mvs_tree_1.MVSTreeSchema.nodes[kind].params);
        const node = {
            kind,
            params: (0, object_1.pickObjectKeys)(params, allowedParamNames),
        };
        (_a = (_b = this._node).children) !== null && _a !== void 0 ? _a : (_b.children = []);
        this._node.children.push(node);
        return node;
    }
}
/** MVS builder pointing to the 'root' node */
class Root extends _Base {
    constructor() {
        const node = { kind: 'root' };
        super(undefined, node);
        this._root = this;
    }
    /** Return the current state of the builder as object in MVS format. */
    getState(metadata) {
        return {
            root: (0, object_1.deepClone)(this._node),
            metadata: {
                ...metadata,
                version: `${mvs_data_1.MVSData.SupportedVersion}`,
                timestamp: utcNowISO(),
            },
        };
    }
    // omitting `saveState`, filesystem operations are responsibility of the caller code (platform-dependent)
    /** Add a 'camera' node and return builder pointing to the root. 'camera' node instructs to set the camera position and orientation. */
    camera(params) {
        this.addChild('camera', params);
        return this;
    }
    /** Add a 'canvas' node and return builder pointing to the root. 'canvas' node sets canvas properties. */
    canvas(params) {
        this.addChild('canvas', params);
        return this;
    }
    /** Add a 'download' node and return builder pointing to it. 'download' node instructs to retrieve a data resource. */
    download(params) {
        return new Download(this._root, this.addChild('download', params));
    }
}
exports.Root = Root;
/** MVS builder pointing to a 'download' node */
class Download extends _Base {
    /** Add a 'parse' node and return builder pointing to it. 'parse' node instructs to parse a data resource. */
    parse(params) {
        return new Parse(this._root, this.addChild('parse', params));
    }
}
exports.Download = Download;
/** Subsets of 'structure' node params which will be passed to individual builder functions. */
const StructureParamsSubsets = {
    model: ['block_header', 'block_index', 'model_index'],
    assembly: ['block_header', 'block_index', 'model_index', 'assembly_id'],
    symmetry: ['block_header', 'block_index', 'model_index', 'ijk_min', 'ijk_max'],
    symmetry_mates: ['block_header', 'block_index', 'model_index', 'radius'],
};
/** MVS builder pointing to a 'parse' node */
class Parse extends _Base {
    /** Add a 'structure' node representing a "model structure", i.e. includes all coordinates from the original model without applying any transformations.
     * Return builder pointing to the new node. */
    modelStructure(params = {}) {
        return new Structure(this._root, this.addChild('structure', {
            type: 'model',
            ...(0, object_1.pickObjectKeys)(params, StructureParamsSubsets.model),
        }));
    }
    /** Add a 'structure' node representing an "assembly structure", i.e. may apply filters and symmetry operators to the original model coordinates.
     * Return builder pointing to the new node. */
    assemblyStructure(params = {}) {
        return new Structure(this._root, this.addChild('structure', {
            type: 'assembly',
            ...(0, object_1.pickObjectKeys)(params, StructureParamsSubsets.assembly),
        }));
    }
    /** Add a 'structure' node representing a "symmetry structure", i.e. applies symmetry operators to build crystal unit cells within given Miller indices.
     * Return builder pointing to the new node. */
    symmetryStructure(params = {}) {
        return new Structure(this._root, this.addChild('structure', {
            type: 'symmetry',
            ...(0, object_1.pickObjectKeys)(params, StructureParamsSubsets.symmetry),
        }));
    }
    /** Add a 'structure' node representing a "symmetry mates structure", i.e. applies symmetry operators to build asymmetric units within a radius from the original model.
     * Return builder pointing to the new node. */
    symmetryMatesStructure(params = {}) {
        return new Structure(this._root, this.addChild('structure', {
            type: 'symmetry_mates',
            ...(0, object_1.pickObjectKeys)(params, StructureParamsSubsets.symmetry_mates),
        }));
    }
}
exports.Parse = Parse;
/** MVS builder pointing to a 'structure' node */
class Structure extends _Base {
    /** Add a 'component' node and return builder pointing to it. 'component' node instructs to create a component (i.e. a subset of the parent structure). */
    component(params = {}) {
        var _a;
        const fullParams = { ...params, selector: (_a = params.selector) !== null && _a !== void 0 ? _a : mvs_defaults_1.MVSDefaults.component.selector };
        return new Component(this._root, this.addChild('component', fullParams));
    }
    /** Add a 'component_from_uri' node and return builder pointing to it. 'component_from_uri' node instructs to create a component defined by an external annotation resource. */
    componentFromUri(params) {
        return new Component(this._root, this.addChild('component_from_uri', params));
    }
    /** Add a 'component_from_source' node and return builder pointing to it. 'component_from_source' node instructs to create a component defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    componentFromSource(params) {
        return new Component(this._root, this.addChild('component_from_source', params));
    }
    /** Add a 'label_from_uri' node and return builder pointing back to the structure node. 'label_from_uri' node instructs to add labels (textual visual representations) to parts of a structure. The labels are defined by an external annotation resource. */
    labelFromUri(params) {
        this.addChild('label_from_uri', params);
        return this;
    }
    /** Add a 'label_from_source' node and return builder pointing back to the structure node. 'label_from_source' node instructs to add labels (textual visual representations) to parts of a structure. The labels are defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    labelFromSource(params) {
        this.addChild('label_from_source', params);
        return this;
    }
    /** Add a 'tooltip_from_uri' node and return builder pointing back to the structure node. 'tooltip_from_uri' node instructs to add tooltips to parts of a structure. The tooltips are defined by an external annotation resource. */
    tooltipFromUri(params) {
        this.addChild('tooltip_from_uri', params);
        return this;
    }
    /** Add a 'tooltip_from_source' node and return builder pointing back to the structure node. 'tooltip_from_source' node instructs to add tooltips to parts of a structure. The tooltips are defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    tooltipFromSource(params) {
        this.addChild('tooltip_from_source', params);
        return this;
    }
    /** Add a 'transform' node and return builder pointing back to the structure node. 'transform' node instructs to rotate and/or translate structure coordinates. */
    transform(params = {}) {
        if (params.rotation && params.rotation.length !== 9) {
            throw new Error('ValueError: `rotation` parameter must be an array of 9 numbers');
        }
        this.addChild('transform', params);
        return this;
    }
}
exports.Structure = Structure;
/** MVS builder pointing to a 'component' or 'component_from_uri' or 'component_from_source' node */
class Component extends _Base {
    /** Add a 'representation' node and return builder pointing to it. 'representation' node instructs to create a visual representation of a component. */
    representation(params = {}) {
        var _a;
        const fullParams = { ...params, type: (_a = params.type) !== null && _a !== void 0 ? _a : 'cartoon' };
        return new Representation(this._root, this.addChild('representation', fullParams));
    }
    /** Add a 'label' node and return builder pointing back to the component node. 'label' node instructs to add a label (textual visual representation) to a component. */
    label(params) {
        this.addChild('label', params);
        return this;
    }
    /** Add a 'tooltip' node and return builder pointing back to the component node. 'tooltip' node instructs to add a text which is not a part of the visualization but should be presented to the users when they interact with the component (typically, the tooltip will be shown somewhere on the screen when the user hovers over a visual representation of the component). */
    tooltip(params) {
        this.addChild('tooltip', params);
        return this;
    }
    /** Add a 'focus' node and return builder pointing back to the component node. 'focus' node instructs to set the camera focus to a component (zoom in). */
    focus(params = {}) {
        this.addChild('focus', params);
        return this;
    }
}
exports.Component = Component;
/** MVS builder pointing to a 'representation' node */
class Representation extends _Base {
    /** Add a 'color' node and return builder pointing back to the representation node. 'color' node instructs to apply color to a visual representation. */
    color(params) {
        this.addChild('color', params);
        return this;
    }
    /** Add a 'color_from_uri' node and return builder pointing back to the representation node. 'color_from_uri' node instructs to apply colors to a visual representation. The colors are defined by an external annotation resource. */
    colorFromUri(params) {
        this.addChild('color_from_uri', params);
        return this;
    }
    /** Add a 'color_from_source' node and return builder pointing back to the representation node. 'color_from_source' node instructs to apply colors to a visual representation. The colors are defined by an annotation resource included in the same file this structure was loaded from. Only applicable if the structure was loaded from an mmCIF or BinaryCIF file. */
    colorFromSource(params) {
        this.addChild('color_from_source', params);
        return this;
    }
}
exports.Representation = Representation;
/** Demonstration of usage of MVS builder */
function builderDemo() {
    const builder = createMVSBuilder();
    builder.canvas({ background_color: 'white' });
    const struct = builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og2_updated.cif' }).parse({ format: 'mmcif' }).modelStructure();
    struct.component().representation().color({ color: 'white' });
    struct.component({ selector: 'ligand' }).representation({ type: 'ball_and_stick' })
        .color({ color: '#555555' })
        .color({ selector: { type_symbol: 'N' }, color: '#3050F8' })
        .color({ selector: { type_symbol: 'O' }, color: '#FF0D0D' })
        .color({ selector: { type_symbol: 'S' }, color: '#FFFF30' })
        .color({ selector: { type_symbol: 'FE' }, color: '#E06633' });
    builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og5_updated.cif' }).parse({ format: 'mmcif' }).assemblyStructure({ assembly_id: '1' }).component().representation().color({ color: 'cyan' });
    builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1og5_updated.cif' }).parse({ format: 'mmcif' }).assemblyStructure({ assembly_id: '2' }).component().representation().color({ color: 'blue' });
    const cif = builder.download({ url: 'https://www.ebi.ac.uk/pdbe/entry-files/download/1wrf_updated.cif' }).parse({ format: 'mmcif' });
    cif.modelStructure({ model_index: 0 }).component().representation().color({ color: '#CC0000' });
    cif.modelStructure({ model_index: 1 }).component().representation().color({ color: '#EE7700' });
    cif.modelStructure({ model_index: 2 }).component().representation().color({ color: '#FFFF00' });
    cif.modelStructure({ model_index: 0 }).transform({ translation: [30, 0, 0] }).component().representation().color({ color: '#ff88bb' });
    cif.modelStructure({ model_index: 0 }).transform({ translation: [60, 0, 0], rotation: [0, 1, 0, -1, 0, 0, 0, 0, 1] }).component().representation().color({ color: '#aa0077' });
    return builder.getState();
}
/** Return the current universal time, in ISO format, e.g. '2023-11-24T10:45:49.873Z' */
function utcNowISO() {
    return new Date().toISOString();
}
