// molstar/lib/commonjs/mol-repr/shape/loci/orientation.js
"use strict";
/**
 * Copyright (c) 2019-2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrientationRepresentation = exports.OrientationParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var names_1 = require("../../../mol-util/color/names");
var representation_1 = require("../representation");
var representation_2 = require("../../representation");
var shape_1 = require("../../../mol-model/shape");
var mesh_1 = require("../../../mol-geo/geometry/mesh/mesh");
var mesh_builder_1 = require("../../../mol-geo/geometry/mesh/mesh-builder");
var label_1 = require("../../../mol-theme/label");
var axes_1 = require("../../../mol-geo/geometry/mesh/builder/axes");
var box_1 = require("../../../mol-geo/geometry/mesh/builder/box");
var ellipsoid_1 = require("../../../mol-geo/geometry/mesh/builder/ellipsoid");
var geometry_1 = require("../../../mol-math/geometry");
var linear_algebra_1 = require("../../../mol-math/linear-algebra");
var marker_action_1 = require("../../../mol-util/marker-action");
var structure_1 = require("../../../mol-model/structure");
var SharedParams = {
    color: param_definition_1.ParamDefinition.Color(names_1.ColorNames.orange),
    scaleFactor: param_definition_1.ParamDefinition.Numeric(1, { min: 0.1, max: 10, step: 0.1 }),
    radiusScale: param_definition_1.ParamDefinition.Numeric(2, { min: 0.1, max: 10, step: 0.1 })
};
var AxesParams = tslib_1.__assign(tslib_1.__assign({}, mesh_1.Mesh.Params), SharedParams);
var BoxParams = tslib_1.__assign(tslib_1.__assign({}, mesh_1.Mesh.Params), SharedParams);
var EllipsoidParams = tslib_1.__assign(tslib_1.__assign({}, mesh_1.Mesh.Params), SharedParams);
var OrientationVisuals = {
    'axes': function (ctx, getParams) { return (0, representation_1.ShapeRepresentation)(getAxesShape, mesh_1.Mesh.Utils); },
    'box': function (ctx, getParams) { return (0, representation_1.ShapeRepresentation)(getBoxShape, mesh_1.Mesh.Utils); },
    'ellipsoid': function (ctx, getParams) { return (0, representation_1.ShapeRepresentation)(getEllipsoidShape, mesh_1.Mesh.Utils); },
};
exports.OrientationParams = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, AxesParams), BoxParams), EllipsoidParams), { visuals: param_definition_1.ParamDefinition.MultiSelect(['box'], param_definition_1.ParamDefinition.objectToOptions(OrientationVisuals)) });
//
function getAxesName(locis) {
    var label = (0, label_1.structureElementLociLabelMany)(locis, { countsOnly: true });
    return "Principal Axes of ".concat(label);
}
function buildAxesMesh(data, props, mesh) {
    var state = mesh_builder_1.MeshBuilder.createState(256, 128, mesh);
    var principalAxes = structure_1.StructureElement.Loci.getPrincipalAxesMany(data.locis);
    geometry_1.Axes3D.scale(principalAxes.momentsAxes, principalAxes.momentsAxes, props.scaleFactor);
    state.currentGroup = 0;
    (0, axes_1.addAxes)(state, principalAxes.momentsAxes, props.radiusScale, 2, 20);
    return mesh_builder_1.MeshBuilder.getMesh(state);
}
function getAxesShape(ctx, data, props, shape) {
    var mesh = buildAxesMesh(data, props, shape && shape.geometry);
    var name = getAxesName(data.locis);
    return shape_1.Shape.create(name, data, mesh, function () { return props.color; }, function () { return 1; }, function () { return name; });
}
//
function getBoxName(locis) {
    var label = (0, label_1.structureElementLociLabelMany)(locis, { countsOnly: true });
    return "Oriented Box of ".concat(label);
}
function buildBoxMesh(data, props, mesh) {
    var state = mesh_builder_1.MeshBuilder.createState(256, 128, mesh);
    var principalAxes = structure_1.StructureElement.Loci.getPrincipalAxesMany(data.locis);
    geometry_1.Axes3D.scale(principalAxes.boxAxes, principalAxes.boxAxes, props.scaleFactor);
    state.currentGroup = 0;
    (0, box_1.addOrientedBox)(state, principalAxes.boxAxes, props.radiusScale, 2, 20);
    return mesh_builder_1.MeshBuilder.getMesh(state);
}
function getBoxShape(ctx, data, props, shape) {
    var mesh = buildBoxMesh(data, props, shape && shape.geometry);
    var name = getBoxName(data.locis);
    return shape_1.Shape.create(name, data, mesh, function () { return props.color; }, function () { return 1; }, function () { return name; });
}
//
function getEllipsoidName(locis) {
    var label = (0, label_1.structureElementLociLabelMany)(locis, { countsOnly: true });
    return "Oriented Ellipsoid of ".concat(label);
}
function buildEllipsoidMesh(data, props, mesh) {
    var state = mesh_builder_1.MeshBuilder.createState(256, 128, mesh);
    var principalAxes = structure_1.StructureElement.Loci.getPrincipalAxesMany(data.locis);
    var axes = principalAxes.boxAxes;
    var origin = axes.origin, dirA = axes.dirA, dirB = axes.dirB;
    var size = geometry_1.Axes3D.size((0, linear_algebra_1.Vec3)(), axes);
    linear_algebra_1.Vec3.scale(size, size, 0.5 * props.scaleFactor);
    var radiusScale = linear_algebra_1.Vec3.create(size[2], size[1], size[0]);
    state.currentGroup = 0;
    (0, ellipsoid_1.addEllipsoid)(state, origin, dirA, dirB, radiusScale, 2);
    return mesh_builder_1.MeshBuilder.getMesh(state);
}
function getEllipsoidShape(ctx, data, props, shape) {
    var mesh = buildEllipsoidMesh(data, props, shape && shape.geometry);
    var name = getEllipsoidName(data.locis);
    return shape_1.Shape.create(name, data, mesh, function () { return props.color; }, function () { return 1; }, function () { return name; });
}
function OrientationRepresentation(ctx, getParams) {
    var repr = representation_2.Representation.createMulti('Orientation', ctx, getParams, representation_2.Representation.StateBuilder, OrientationVisuals);
    repr.setState({ markerActions: marker_action_1.MarkerActions.Highlighting });
    return repr;
}
exports.OrientationRepresentation = OrientationRepresentation;
