// molstar/lib/commonjs/mol-repr/structure/visual/util/polymer.js
"use strict";
/**
 * Copyright (c) 2018-2023 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolymerGapLocationIterator = exports.PolymerLocationIterator = exports.OverhangFactor = exports.NucleicShift = exports.StandardShift = exports.HelixTension = exports.StandardTension = void 0;
exports.getPolymerRanges = getPolymerRanges;
exports.getGapRanges = getGapRanges;
exports.getPolymerElementLoci = getPolymerElementLoci;
exports.eachAtomicUnitTracedElement = eachAtomicUnitTracedElement;
exports.eachPolymerElement = eachPolymerElement;
exports.getPolymerGapElementLoci = getPolymerGapElementLoci;
exports.eachPolymerGapElement = eachPolymerGapElement;
const tslib_1 = require("tslib");
const structure_1 = require("../../../../mol-model/structure");
const int_1 = require("../../../../mol-data/int");
const loci_1 = require("../../../../mol-model/loci");
const location_iterator_1 = require("../../../../mol-geo/util/location-iterator");
const common_1 = require("./common");
tslib_1.__exportStar(require("./polymer/backbone"), exports);
tslib_1.__exportStar(require("./polymer/gap-iterator"), exports);
tslib_1.__exportStar(require("./polymer/trace-iterator"), exports);
tslib_1.__exportStar(require("./polymer/curve-segment"), exports);
exports.StandardTension = 0.5;
exports.HelixTension = 0.9;
exports.StandardShift = 0.5;
exports.NucleicShift = 0.3;
exports.OverhangFactor = 2;
function getPolymerRanges(unit) {
    switch (unit.kind) {
        case 0 /* Unit.Kind.Atomic */: return unit.model.atomicRanges.polymerRanges;
        case 1 /* Unit.Kind.Spheres */: return unit.model.coarseHierarchy.spheres.polymerRanges;
        case 2 /* Unit.Kind.Gaussians */: return unit.model.coarseHierarchy.gaussians.polymerRanges;
    }
}
function getGapRanges(unit) {
    switch (unit.kind) {
        case 0 /* Unit.Kind.Atomic */: return unit.model.atomicRanges.gapRanges;
        case 1 /* Unit.Kind.Spheres */: return unit.model.coarseHierarchy.spheres.gapRanges;
        case 2 /* Unit.Kind.Gaussians */: return unit.model.coarseHierarchy.gaussians.gapRanges;
    }
}
var PolymerLocationIterator;
(function (PolymerLocationIterator) {
    function fromGroup(structureGroup, options) {
        const { group, structure } = structureGroup;
        const polymerElements = group.units[0].polymerElements;
        const groupCount = polymerElements.length;
        const instanceCount = group.units.length;
        const location = structure_1.StructureElement.Location.create(structure);
        const getLocation = (groupIndex, instanceIndex) => {
            const unit = group.units[instanceIndex];
            location.unit = unit;
            location.element = polymerElements[groupIndex];
            return location;
        };
        const asSecondary = !!(options === null || options === void 0 ? void 0 : options.asSecondary);
        function isSecondary(elementIndex, instanceIndex) {
            return asSecondary;
        }
        return (0, location_iterator_1.LocationIterator)(groupCount, instanceCount, 1, getLocation, false, isSecondary);
    }
    PolymerLocationIterator.fromGroup = fromGroup;
})(PolymerLocationIterator || (exports.PolymerLocationIterator = PolymerLocationIterator = {}));
var PolymerGapLocationIterator;
(function (PolymerGapLocationIterator) {
    function fromGroup(structureGroup) {
        const { group, structure } = structureGroup;
        const gapElements = group.units[0].gapElements;
        const groupCount = gapElements.length;
        const instanceCount = group.units.length;
        const location = structure_1.StructureElement.Location.create(structure);
        const getLocation = (groupIndex, instanceIndex) => {
            const unit = group.units[instanceIndex];
            location.unit = unit;
            location.element = gapElements[groupIndex];
            return location;
        };
        return (0, location_iterator_1.LocationIterator)(groupCount, instanceCount, 1, getLocation);
    }
    PolymerGapLocationIterator.fromGroup = fromGroup;
})(PolymerGapLocationIterator || (exports.PolymerGapLocationIterator = PolymerGapLocationIterator = {}));
/** Return a Loci for the elements of the whole residue of a polymer element. */
function getPolymerElementLoci(pickingId, structureGroup, id) {
    const { objectId, instanceId, groupId } = pickingId;
    if (id === objectId) {
        const { structure, group } = structureGroup;
        const unit = group.units[instanceId];
        if (structure_1.Unit.isAtomic(unit)) {
            return (0, common_1.getResidueLoci)(structure, unit, unit.polymerElements[groupId]);
        }
        else {
            const { elements } = unit;
            const elementIndex = unit.polymerElements[groupId];
            const unitIndex = int_1.OrderedSet.indexOf(elements, elementIndex);
            if (unitIndex !== -1) {
                const indices = int_1.OrderedSet.ofSingleton(unitIndex);
                return structure_1.StructureElement.Loci(structure, [{ unit, indices }]);
            }
        }
    }
    return loci_1.EmptyLoci;
}
function tryApplyResidueInterval(offset, elements, traceElementIndex, apply, r1, r2) {
    let start = -1, startIdx = -1;
    for (let rI = r1; rI <= r2; rI++) {
        const eI = traceElementIndex[rI];
        if (eI < 0)
            continue;
        start = int_1.OrderedSet.indexOf(elements, eI);
        if (start >= 0) {
            startIdx = rI;
            break;
        }
    }
    if (start < 0) {
        return false;
    }
    let end = start;
    for (let rI = r2; rI > startIdx; rI--) {
        const eI = traceElementIndex[rI];
        if (eI < 0)
            continue;
        const e = int_1.OrderedSet.indexOf(elements, eI);
        if (e >= 0) {
            end = e;
            break;
        }
    }
    return apply(int_1.Interval.ofRange(offset + start, offset + end));
}
function eachAtomicUnitTracedElement(offset, groupSize, elementsSelector, apply, e) {
    let changed = false;
    const { elements } = e.unit;
    const { traceElementIndex } = e.unit.model.atomicHierarchy.derived.residue;
    const { index: resIndex } = e.unit.model.atomicHierarchy.residueAtomSegments;
    const tracedElements = elementsSelector(e.unit);
    if (int_1.Interval.is(e.indices)) {
        if (int_1.Interval.start(e.indices) === 0 && int_1.Interval.end(e.indices) === e.unit.elements.length) {
            // full unit here
            changed = apply(int_1.Interval.ofBounds(offset, offset + groupSize)) || changed;
        }
        else {
            const r1 = resIndex[elements[int_1.Interval.min(e.indices)]];
            const r2 = resIndex[elements[int_1.Interval.max(e.indices)]];
            changed = tryApplyResidueInterval(offset, tracedElements, traceElementIndex, apply, r1, r2) || changed;
        }
    }
    else {
        const { indices } = e;
        for (let i = 0, _i = indices.length; i < _i; i++) {
            const r1 = resIndex[elements[indices[i]]];
            let r2 = r1;
            let endI = i + 1;
            while (endI < _i) {
                const _r = resIndex[elements[indices[endI]]];
                if (_r - r2 > 1)
                    break;
                r2 = _r;
                endI++;
            }
            i = endI - 1;
            changed = tryApplyResidueInterval(offset, tracedElements, traceElementIndex, apply, r1, r2) || changed;
        }
    }
    return changed;
}
function selectPolymerElements(u) { return u.polymerElements; }
/** Mark a polymer element (e.g. part of a cartoon trace) */
function eachPolymerElement(loci, structureGroup, apply) {
    let changed = false;
    if (!structure_1.StructureElement.Loci.is(loci))
        return false;
    const { structure, group } = structureGroup;
    if (!structure_1.Structure.areEquivalent(loci.structure, structure))
        return false;
    const groupCount = group.units[0].polymerElements.length;
    for (const e of loci.elements) {
        if (!group.unitIndexMap.has(e.unit.id))
            continue;
        const offset = group.unitIndexMap.get(e.unit.id) * groupCount; // to target unit instance
        if (structure_1.Unit.isAtomic(e.unit)) {
            changed = eachAtomicUnitTracedElement(offset, groupCount, selectPolymerElements, apply, e) || changed;
        }
        else {
            if (int_1.Interval.is(e.indices)) {
                const start = offset + int_1.Interval.start(e.indices);
                const end = offset + int_1.Interval.end(e.indices);
                changed = apply(int_1.Interval.ofBounds(start, end)) || changed;
            }
            else {
                for (let i = 0, _i = e.indices.length; i < _i; i++) {
                    const start = e.indices[i];
                    let endI = i + 1;
                    while (endI < _i && e.indices[endI] === start)
                        endI++;
                    i = endI - 1;
                    const end = e.indices[i];
                    changed = apply(int_1.Interval.ofRange(offset + start, offset + end)) || changed;
                }
            }
        }
    }
    return changed;
}
/** Return a Loci for both directions of the polymer gap element. */
function getPolymerGapElementLoci(pickingId, structureGroup, id) {
    const { objectId, instanceId, groupId } = pickingId;
    if (id === objectId) {
        const { structure, group } = structureGroup;
        const unit = group.units[instanceId];
        const unitIndexA = int_1.OrderedSet.indexOf(unit.elements, unit.gapElements[groupId]);
        const unitIndexB = int_1.OrderedSet.indexOf(unit.elements, unit.gapElements[groupId % 2 ? groupId - 1 : groupId + 1]);
        if (unitIndexA !== -1 && unitIndexB !== -1) {
            return structure_1.Bond.Loci(structure, [
                structure_1.Bond.Location(structure, unit, unitIndexA, structure, unit, unitIndexB),
                structure_1.Bond.Location(structure, unit, unitIndexB, structure, unit, unitIndexA)
            ]);
        }
    }
    return loci_1.EmptyLoci;
}
function eachPolymerGapElement(loci, structureGroup, apply) {
    let changed = false;
    if (structure_1.Bond.isLoci(loci)) {
        const { structure, group } = structureGroup;
        if (!structure_1.Structure.areRootsEquivalent(loci.structure, structure))
            return false;
        loci = structure_1.Bond.remapLoci(loci, structure);
        const groupCount = group.units[0].gapElements.length;
        for (const b of loci.bonds) {
            const unitIdx = group.unitIndexMap.get(b.aUnit.id);
            if (unitIdx !== undefined) {
                const idxA = int_1.OrderedSet.indexOf(b.aUnit.gapElements, b.aUnit.elements[b.aIndex]);
                const idxB = int_1.OrderedSet.indexOf(b.bUnit.gapElements, b.bUnit.elements[b.bIndex]);
                if (idxA !== -1 && idxB !== -1) {
                    if (apply(int_1.Interval.ofSingleton(unitIdx * groupCount + idxA)))
                        changed = true;
                }
            }
        }
    }
    else if (structure_1.StructureElement.Loci.is(loci)) {
        const { structure, group } = structureGroup;
        if (!structure_1.Structure.areRootsEquivalent(loci.structure, structure))
            return false;
        loci = structure_1.StructureElement.Loci.remap(loci, structure);
        const groupCount = group.units[0].gapElements.length;
        for (const e of loci.elements) {
            const unitIdx = group.unitIndexMap.get(e.unit.id);
            if (unitIdx !== undefined) {
                int_1.OrderedSet.forEach(e.indices, v => {
                    const idx = int_1.OrderedSet.indexOf(e.unit.gapElements, e.unit.elements[v]);
                    if (idx !== -1) {
                        if (apply(int_1.Interval.ofSingleton(unitIdx * groupCount + idx)))
                            changed = true;
                    }
                });
            }
        }
    }
    return changed;
}
