// molstar/lib/commonjs/mol-geo/geometry/substance-data.js
"use strict";
/**
 * Copyright (c) 2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEmptySubstance = exports.createSubstance = exports.clearSubstance = exports.applySubstanceMaterial = void 0;
var value_cell_1 = require("../../mol-util/value-cell");
var linear_algebra_1 = require("../../mol-math/linear-algebra");
var util_1 = require("../../mol-gl/renderable/util");
var texture_1 = require("../../mol-gl/webgl/texture");
var material_1 = require("../../mol-util/material");
function applySubstanceMaterial(array, start, end, material) {
    for (var i = start; i < end; ++i) {
        material_1.Material.toArray(material, array, i * 4);
        array[i * 4 + 3] = 255;
    }
    return true;
}
exports.applySubstanceMaterial = applySubstanceMaterial;
function clearSubstance(array, start, end) {
    array.fill(0, start * 4, end * 4);
    return true;
}
exports.clearSubstance = clearSubstance;
function createSubstance(count, substanceData) {
    var substance = (0, util_1.createTextureImage)(Math.max(1, count), 4, Uint8Array, substanceData && substanceData.tSubstance.ref.value.array);
    if (substanceData) {
        value_cell_1.ValueCell.update(substanceData.tSubstance, substance);
        value_cell_1.ValueCell.update(substanceData.uSubstanceTexDim, linear_algebra_1.Vec2.create(substance.width, substance.height));
        value_cell_1.ValueCell.updateIfChanged(substanceData.dSubstance, count > 0);
        return substanceData;
    }
    else {
        return {
            tSubstance: value_cell_1.ValueCell.create(substance),
            uSubstanceTexDim: value_cell_1.ValueCell.create(linear_algebra_1.Vec2.create(substance.width, substance.height)),
            dSubstance: value_cell_1.ValueCell.create(count > 0),
            tSubstanceGrid: value_cell_1.ValueCell.create((0, texture_1.createNullTexture)()),
            uSubstanceGridDim: value_cell_1.ValueCell.create(linear_algebra_1.Vec3.create(1, 1, 1)),
            uSubstanceGridTransform: value_cell_1.ValueCell.create(linear_algebra_1.Vec4.create(0, 0, 0, 1)),
            dSubstanceType: value_cell_1.ValueCell.create('groupInstance'),
        };
    }
}
exports.createSubstance = createSubstance;
var emptySubstanceTexture = { array: new Uint8Array(4), width: 1, height: 1 };
function createEmptySubstance(substanceData) {
    if (substanceData) {
        value_cell_1.ValueCell.update(substanceData.tSubstance, emptySubstanceTexture);
        value_cell_1.ValueCell.update(substanceData.uSubstanceTexDim, linear_algebra_1.Vec2.create(1, 1));
        return substanceData;
    }
    else {
        return {
            tSubstance: value_cell_1.ValueCell.create(emptySubstanceTexture),
            uSubstanceTexDim: value_cell_1.ValueCell.create(linear_algebra_1.Vec2.create(1, 1)),
            dSubstance: value_cell_1.ValueCell.create(false),
            tSubstanceGrid: value_cell_1.ValueCell.create((0, texture_1.createNullTexture)()),
            uSubstanceGridDim: value_cell_1.ValueCell.create(linear_algebra_1.Vec3.create(1, 1, 1)),
            uSubstanceGridTransform: value_cell_1.ValueCell.create(linear_algebra_1.Vec4.create(0, 0, 0, 1)),
            dSubstanceType: value_cell_1.ValueCell.create('groupInstance'),
        };
    }
}
exports.createEmptySubstance = createEmptySubstance;
