// molstar/lib/commonjs/extensions/rcsb/assembly-symmetry/color.js
"use strict";
/**
 * Copyright (c) 2018-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssemblySymmetryClusterColorThemeProvider = exports.AssemblySymmetryClusterColorTheme = exports.getAssemblySymmetryClusterColorThemeParams = exports.AssemblySymmetryClusterColorThemeParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var prop_1 = require("./prop");
var color_1 = require("../../../mol-util/color");
var structure_1 = require("../../../mol-model/structure");
var palette_1 = require("../../../mol-util/color/palette");
var DefaultColor = (0, color_1.Color)(0xCCCCCC);
function getAsymId(unit) {
    switch (unit.kind) {
        case 0 /* Unit.Kind.Atomic */:
            return structure_1.StructureProperties.chain.label_asym_id;
        case 1 /* Unit.Kind.Spheres */:
        case 2 /* Unit.Kind.Gaussians */:
            return structure_1.StructureProperties.coarse.asym_id;
    }
}
function clusterMemberKey(asymId, operList) {
    return "".concat(asymId, "-").concat(operList.join('|'));
}
exports.AssemblySymmetryClusterColorThemeParams = tslib_1.__assign({}, (0, palette_1.getPaletteParams)({ colorList: 'red-yellow-blue' }));
function getAssemblySymmetryClusterColorThemeParams(ctx) {
    var params = param_definition_1.ParamDefinition.clone(exports.AssemblySymmetryClusterColorThemeParams);
    return params;
}
exports.getAssemblySymmetryClusterColorThemeParams = getAssemblySymmetryClusterColorThemeParams;
function AssemblySymmetryClusterColorTheme(ctx, props) {
    var _a;
    var color = function () { return DefaultColor; };
    var legend;
    var assemblySymmetry = ctx.structure && prop_1.AssemblySymmetryProvider.get(ctx.structure);
    var contextHash = assemblySymmetry === null || assemblySymmetry === void 0 ? void 0 : assemblySymmetry.version;
    var clusters = (_a = assemblySymmetry === null || assemblySymmetry === void 0 ? void 0 : assemblySymmetry.value) === null || _a === void 0 ? void 0 : _a.clusters;
    if ((clusters === null || clusters === void 0 ? void 0 : clusters.length) && ctx.structure) {
        var l_1 = structure_1.StructureElement.Location.create(ctx.structure);
        var clusterByMember_1 = new Map();
        for (var i = 0, il = clusters.length; i < il; ++i) {
            var members = clusters[i].members;
            for (var j = 0, jl = members.length; j < jl; ++j) {
                var asymId = members[j].asym_id;
                var operList = tslib_1.__spreadArray([], members[j].pdbx_struct_oper_list_ids || [], true);
                clusterByMember_1.set(clusterMemberKey(asymId, operList), i);
                if (operList.length === 0) {
                    operList.push('1'); // TODO hack assuming '1' is the id of the identity operator
                    clusterByMember_1.set(clusterMemberKey(asymId, operList), i);
                }
            }
        }
        var palette_2 = (0, palette_1.getPalette)(clusters.length, props);
        legend = palette_2.legend;
        var _emptyList_1 = [];
        var getColor_1 = function (location) {
            var assembly = location.unit.conformation.operator.assembly;
            var asymId = getAsymId(location.unit)(location);
            var cluster = clusterByMember_1.get(clusterMemberKey(asymId, (assembly === null || assembly === void 0 ? void 0 : assembly.operList) || _emptyList_1));
            return cluster !== undefined ? palette_2.color(cluster) : DefaultColor;
        };
        color = function (location) {
            if (structure_1.StructureElement.Location.is(location)) {
                return getColor_1(location);
            }
            else if (structure_1.Bond.isLocation(location)) {
                l_1.unit = location.aUnit;
                l_1.element = location.aUnit.elements[location.aIndex];
                return getColor_1(l_1);
            }
            return DefaultColor;
        };
    }
    return {
        factory: AssemblySymmetryClusterColorTheme,
        granularity: 'instance',
        color: color,
        props: props,
        contextHash: contextHash,
        description: 'Assigns chain colors according to assembly symmetry cluster membership calculated with BioJava and obtained via RCSB PDB.',
        legend: legend
    };
}
exports.AssemblySymmetryClusterColorTheme = AssemblySymmetryClusterColorTheme;
exports.AssemblySymmetryClusterColorThemeProvider = {
    name: prop_1.AssemblySymmetry.Tag.Cluster,
    label: 'Assembly Symmetry Cluster',
    category: "Symmetry" /* ColorTheme.Category.Symmetry */,
    factory: AssemblySymmetryClusterColorTheme,
    getParams: getAssemblySymmetryClusterColorThemeParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.AssemblySymmetryClusterColorThemeParams),
    isApplicable: function (ctx) { return prop_1.AssemblySymmetry.isApplicable(ctx.structure); },
    ensureCustomProperties: {
        attach: function (ctx, data) { return data.structure ? prop_1.AssemblySymmetryProvider.attach(ctx, data.structure, void 0, true) : Promise.resolve(); },
        detach: function (data) { return data.structure && prop_1.AssemblySymmetryProvider.ref(data.structure, false); }
    }
};
