// molstar/lib/extensions/cellpack/property.js
/**
 * Copyright (c) 2019-2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
import { __assign, __awaiter, __generator } from "tslib";
import { CustomStructureProperty } from '../../mol-model-props/common/custom-structure-property';
import { ParamDefinition as PD } from '../../mol-util/param-definition';
import { CustomPropertyDescriptor } from '../../mol-model/custom-property';
var CellPackInfoParams = {
    info: PD.Value({ packingsCount: 1, packingIndex: 0, colors: undefined }, { isHidden: true })
};
export var CellPackInfoProvider = CustomStructureProperty.createProvider({
    label: 'CellPack Info',
    descriptor: CustomPropertyDescriptor({ name: 'cellpack-info' }),
    type: 'root',
    defaultParams: CellPackInfoParams,
    getParams: function (data) { return CellPackInfoParams; },
    isApplicable: function (data) { return true; },
    obtain: function (ctx, data, props) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, {
                    value: __assign(__assign({}, CellPackInfoParams.info.defaultValue), props.info)
                }];
        });
    }); }
});
