// molstar/lib/commonjs/extensions/cellpack/color/generate.js
"use strict";
/**
 * Copyright (c) 2019-2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CellPackGenerateColorThemeProvider = exports.CellPackGenerateColorTheme = exports.getCellPackGenerateColorThemeParams = exports.CellPackGenerateColorThemeParams = void 0;
var param_definition_1 = require("../../../mol-util/param-definition");
var color_1 = require("../../../mol-util/color");
var palette_1 = require("../../../mol-util/color/palette");
var structure_1 = require("../../../mol-model/structure");
var property_1 = require("../property");
var distinct_1 = require("../../../mol-util/color/distinct");
var hcl_1 = require("../../../mol-util/color/spaces/hcl");
var DefaultColor = (0, color_1.Color)(0xCCCCCC);
var Description = 'Gives every model in a CellPack packing a unique generated color similar to other models in the packing.';
exports.CellPackGenerateColorThemeParams = {};
function getCellPackGenerateColorThemeParams(ctx) {
    return exports.CellPackGenerateColorThemeParams; // TODO return copy
}
exports.getCellPackGenerateColorThemeParams = getCellPackGenerateColorThemeParams;
function CellPackGenerateColorTheme(ctx, props) {
    var color;
    var legend;
    var info = ctx.structure && property_1.CellPackInfoProvider.get(ctx.structure).value;
    if (ctx.structure && info) {
        var colors = (0, distinct_1.distinctColors)(info.packingsCount);
        var hcl = hcl_1.Hcl.fromColor((0, hcl_1.Hcl)(), colors[info.packingIndex]);
        var hue = [Math.max(0, hcl[0] - 35), Math.min(360, hcl[0] + 35)];
        var models = ctx.structure.root.models;
        var size = 0;
        for (var _i = 0, models_1 = models; _i < models_1.length; _i++) {
            var m = models_1[_i];
            size = Math.max(size, structure_1.Model.TrajectoryInfo.get(m).size);
        }
        var palette = (0, palette_1.getPalette)(size, { palette: {
                name: 'generate',
                params: {
                    hue: hue,
                    chroma: [30, 80], luminance: [15, 85],
                    clusteringStepCount: 50, minSampleCount: 800, maxCount: 75
                }
            } }, { minLabel: 'Min', maxLabel: 'Max' });
        legend = palette.legend;
        var modelColor_1 = new Map();
        for (var i = 0, il = models.length; i < il; ++i) {
            var idx = structure_1.Model.TrajectoryInfo.get(models[i]).index;
            modelColor_1.set(structure_1.Model.TrajectoryInfo.get(models[i]).index, palette.color(idx));
        }
        color = function (location) {
            if (structure_1.StructureElement.Location.is(location)) {
                return modelColor_1.get(structure_1.Model.TrajectoryInfo.get(location.unit.model).index);
            }
            else if (structure_1.Bond.isLocation(location)) {
                return modelColor_1.get(structure_1.Model.TrajectoryInfo.get(location.aUnit.model).index);
            }
            return DefaultColor;
        };
    }
    else {
        color = function () { return DefaultColor; };
    }
    return {
        factory: CellPackGenerateColorTheme,
        granularity: 'instance',
        color: color,
        props: props,
        description: Description,
        legend: legend
    };
}
exports.CellPackGenerateColorTheme = CellPackGenerateColorTheme;
exports.CellPackGenerateColorThemeProvider = {
    name: 'cellpack-generate',
    label: 'CellPack Generate',
    category: "Chain Property" /* ColorTheme.Category.Chain */,
    factory: CellPackGenerateColorTheme,
    getParams: getCellPackGenerateColorThemeParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.CellPackGenerateColorThemeParams),
    isApplicable: function (ctx) {
        return (!!ctx.structure && ctx.structure.elementCount > 0 &&
            !!property_1.CellPackInfoProvider.get(ctx.structure).value);
    }
};
