// molstar/lib/commonjs/extensions/rcsb/validation-report/color/density-fit.js
"use strict";
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DensityFitColorThemeProvider = void 0;
exports.DensityFitColorTheme = DensityFitColorTheme;
const color_1 = require("../../../../mol-theme/color");
const param_definition_1 = require("../../../../mol-util/param-definition");
const color_2 = require("../../../../mol-util/color");
const structure_1 = require("../../../../mol-model/structure");
const prop_1 = require("../prop");
const DefaultColor = (0, color_2.Color)(0xCCCCCC);
function DensityFitColorTheme(ctx, props) {
    var _a;
    let color = () => DefaultColor;
    const scaleRsrz = color_2.ColorScale.create({
        minLabel: 'Poor',
        maxLabel: 'Better',
        domain: [2, 0],
        listOrName: 'red-yellow-blue',
    });
    const scaleRscc = color_2.ColorScale.create({
        minLabel: 'Poor',
        maxLabel: 'Better',
        domain: [0.678, 1.0],
        listOrName: 'red-yellow-blue',
    });
    const validationReport = ctx.structure && prop_1.ValidationReportProvider.get(ctx.structure.models[0]);
    const contextHash = validationReport === null || validationReport === void 0 ? void 0 : validationReport.version;
    const model = (_a = ctx.structure) === null || _a === void 0 ? void 0 : _a.models[0];
    if ((validationReport === null || validationReport === void 0 ? void 0 : validationReport.value) && model) {
        const { rsrz, rscc } = validationReport.value;
        const residueIndex = model.atomicHierarchy.residueAtomSegments.index;
        const getColor = (element) => {
            const rsrzValue = rsrz.get(residueIndex[element]);
            if (rsrzValue !== undefined)
                return scaleRsrz.color(rsrzValue);
            const rsccValue = rscc.get(residueIndex[element]);
            if (rsccValue !== undefined)
                return scaleRscc.color(rsccValue);
            return DefaultColor;
        };
        color = (location) => {
            if (structure_1.StructureElement.Location.is(location) && location.unit.model === model) {
                return getColor(location.element);
            }
            else if (structure_1.Bond.isLocation(location) && location.aUnit.model === model) {
                return getColor(location.aUnit.elements[location.aIndex]);
            }
            return DefaultColor;
        };
    }
    return {
        factory: DensityFitColorTheme,
        granularity: 'group',
        preferSmoothing: true,
        color,
        props,
        contextHash,
        description: 'Assigns residue colors according to the density fit using normalized Real Space R (RSRZ) for polymer residues and real space correlation coefficient (RSCC) for ligands. Colors range from poor (RSRZ = 2 or RSCC = 0.678) - to better (RSRZ = 0 or RSCC = 1.0). Data from wwPDB Validation Report, obtained via RCSB PDB.',
        legend: scaleRsrz.legend
    };
}
exports.DensityFitColorThemeProvider = {
    name: prop_1.ValidationReport.Tag.DensityFit,
    label: 'Density Fit',
    category: color_1.ColorTheme.Category.Validation,
    factory: DensityFitColorTheme,
    getParams: () => ({}),
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues({}),
    isApplicable: (ctx) => !!ctx.structure && prop_1.ValidationReport.isApplicable(ctx.structure.models[0]) && structure_1.Model.isFromXray(ctx.structure.models[0]) && structure_1.Model.probablyHasDensityMap(ctx.structure.models[0]),
    ensureCustomProperties: {
        attach: (ctx, data) => data.structure ? prop_1.ValidationReportProvider.attach(ctx, data.structure.models[0], void 0, true) : Promise.resolve(),
        detach: (data) => data.structure && prop_1.ValidationReportProvider.ref(data.structure.models[0], false)
    }
};
