// molstar/lib/commonjs/extensions/rcsb/validation-report/color/density-fit.js
"use strict";
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DensityFitColorThemeProvider = exports.DensityFitColorTheme = void 0;
var param_definition_1 = require("../../../../mol-util/param-definition");
var color_1 = require("../../../../mol-util/color");
var structure_1 = require("../../../../mol-model/structure");
var prop_1 = require("../prop");
var DefaultColor = (0, color_1.Color)(0xCCCCCC);
function DensityFitColorTheme(ctx, props) {
    var _a;
    var color = function () { return DefaultColor; };
    var scaleRsrz = color_1.ColorScale.create({
        minLabel: 'Poor',
        maxLabel: 'Better',
        domain: [2, 0],
        listOrName: 'red-yellow-blue',
    });
    var scaleRscc = color_1.ColorScale.create({
        minLabel: 'Poor',
        maxLabel: 'Better',
        domain: [0.678, 1.0],
        listOrName: 'red-yellow-blue',
    });
    var validationReport = ctx.structure && prop_1.ValidationReportProvider.get(ctx.structure.models[0]);
    var contextHash = validationReport === null || validationReport === void 0 ? void 0 : validationReport.version;
    var model = (_a = ctx.structure) === null || _a === void 0 ? void 0 : _a.models[0];
    if ((validationReport === null || validationReport === void 0 ? void 0 : validationReport.value) && model) {
        var _b = validationReport.value, rsrz_1 = _b.rsrz, rscc_1 = _b.rscc;
        var residueIndex_1 = model.atomicHierarchy.residueAtomSegments.index;
        var getColor_1 = function (element) {
            var rsrzValue = rsrz_1.get(residueIndex_1[element]);
            if (rsrzValue !== undefined)
                return scaleRsrz.color(rsrzValue);
            var rsccValue = rscc_1.get(residueIndex_1[element]);
            if (rsccValue !== undefined)
                return scaleRscc.color(rsccValue);
            return DefaultColor;
        };
        color = function (location) {
            if (structure_1.StructureElement.Location.is(location) && location.unit.model === model) {
                return getColor_1(location.element);
            }
            else if (structure_1.Bond.isLocation(location) && location.aUnit.model === model) {
                return getColor_1(location.aUnit.elements[location.aIndex]);
            }
            return DefaultColor;
        };
    }
    return {
        factory: DensityFitColorTheme,
        granularity: 'group',
        preferSmoothing: true,
        color: color,
        props: props,
        contextHash: contextHash,
        description: 'Assigns residue colors according to the density fit using normalized Real Space R (RSRZ) for polymer residues and real space correlation coefficient (RSCC) for ligands. Colors range from poor (RSRZ = 2 or RSCC = 0.678) - to better (RSRZ = 0 or RSCC = 1.0). Data from wwPDB Validation Report, obtained via RCSB PDB.',
        legend: scaleRsrz.legend
    };
}
exports.DensityFitColorTheme = DensityFitColorTheme;
exports.DensityFitColorThemeProvider = {
    name: prop_1.ValidationReport.Tag.DensityFit,
    label: 'Density Fit',
    category: "Validation" /* ColorTheme.Category.Validation */,
    factory: DensityFitColorTheme,
    getParams: function () { return ({}); },
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues({}),
    isApplicable: function (ctx) { return !!ctx.structure && prop_1.ValidationReport.isApplicable(ctx.structure.models[0]) && structure_1.Model.isFromXray(ctx.structure.models[0]) && structure_1.Model.probablyHasDensityMap(ctx.structure.models[0]); },
    ensureCustomProperties: {
        attach: function (ctx, data) { return data.structure ? prop_1.ValidationReportProvider.attach(ctx, data.structure.models[0], void 0, true) : Promise.resolve(); },
        detach: function (data) { return data.structure && prop_1.ValidationReportProvider.ref(data.structure.models[0], false); }
    }
};
