// molstar/lib/commonjs/extensions/rcsb/validation-report/color/geometry-quality.js
"use strict";
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeometryQualityColorThemeProvider = void 0;
exports.getGeometricQualityColorThemeParams = getGeometricQualityColorThemeParams;
exports.GeometryQualityColorTheme = GeometryQualityColorTheme;
const color_1 = require("../../../../mol-theme/color");
const param_definition_1 = require("../../../../mol-util/param-definition");
const color_2 = require("../../../../mol-util/color");
const structure_1 = require("../../../../mol-model/structure");
const prop_1 = require("../prop");
const legend_1 = require("../../../../mol-util/legend");
const set_1 = require("../../../../mol-util/set");
const DefaultColor = (0, color_2.Color)(0x909090);
const NoIssuesColor = (0, color_2.Color)(0x2166ac);
const OneIssueColor = (0, color_2.Color)(0xfee08b);
const TwoIssuesColor = (0, color_2.Color)(0xf46d43);
const ThreeOrMoreIssuesColor = (0, color_2.Color)(0xa50026);
const ColorLegend = (0, legend_1.TableLegend)([
    ['Data unavailable', DefaultColor],
    ['No issues', NoIssuesColor],
    ['One issue', OneIssueColor],
    ['Two issues', TwoIssuesColor],
    ['Three or more issues', ThreeOrMoreIssuesColor],
]);
function getGeometricQualityColorThemeParams(ctx) {
    const validationReport = !!ctx.structure && ctx.structure.models.length > 0 && prop_1.ValidationReportProvider.get(ctx.structure.models[0]).value;
    const options = [];
    if (validationReport) {
        const kinds = new Set();
        validationReport.geometryIssues.forEach(v => v.forEach(k => kinds.add(k)));
        kinds.forEach(k => options.push([k, k]));
    }
    return {
        ignore: param_definition_1.ParamDefinition.MultiSelect([], options)
    };
}
function GeometryQualityColorTheme(ctx, props) {
    var _a;
    let color = () => DefaultColor;
    const validationReport = !!ctx.structure && ctx.structure.models.length > 0 ? prop_1.ValidationReportProvider.get(ctx.structure.models[0]) : void 0;
    const contextHash = validationReport === null || validationReport === void 0 ? void 0 : validationReport.version;
    const value = validationReport === null || validationReport === void 0 ? void 0 : validationReport.value;
    const model = (_a = ctx.structure) === null || _a === void 0 ? void 0 : _a.models[0];
    if (value && model) {
        const { geometryIssues, clashes, bondOutliers, angleOutliers } = value;
        const residueIndex = model.atomicHierarchy.residueAtomSegments.index;
        const { polymerType } = model.atomicHierarchy.derived.residue;
        const ignore = new Set(props.ignore);
        const getColor = (element) => {
            const rI = residueIndex[element];
            const value = geometryIssues.get(rI);
            if (value === undefined)
                return DefaultColor;
            let count = set_1.SetUtils.differenceSize(value, ignore);
            if (count > 0 && polymerType[rI] === 0 /* PolymerType.NA */) {
                count = 0;
                if (!ignore.has('clash') && clashes.getVertexEdgeCount(element) > 0)
                    count += 1;
                if (!ignore.has('mog-bond-outlier') && bondOutliers.index.has(element))
                    count += 1;
                if (!ignore.has('mog-angle-outlier') && angleOutliers.index.has(element))
                    count += 1;
            }
            switch (count) {
                case undefined: return DefaultColor;
                case 0: return NoIssuesColor;
                case 1: return OneIssueColor;
                case 2: return TwoIssuesColor;
                default: return ThreeOrMoreIssuesColor;
            }
        };
        color = (location) => {
            if (structure_1.StructureElement.Location.is(location) && location.unit.model === model) {
                return getColor(location.element);
            }
            else if (structure_1.Bond.isLocation(location) && location.aUnit.model === model) {
                return getColor(location.aUnit.elements[location.aIndex]);
            }
            return DefaultColor;
        };
    }
    return {
        factory: GeometryQualityColorTheme,
        granularity: 'group',
        preferSmoothing: true,
        color,
        props,
        contextHash,
        description: 'Assigns residue colors according to the number of (filtered) geometry issues. Data from wwPDB Validation Report, obtained via RCSB PDB.',
        legend: ColorLegend
    };
}
exports.GeometryQualityColorThemeProvider = {
    name: prop_1.ValidationReport.Tag.GeometryQuality,
    label: 'Geometry Quality',
    category: color_1.ColorTheme.Category.Validation,
    factory: GeometryQualityColorTheme,
    getParams: getGeometricQualityColorThemeParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(getGeometricQualityColorThemeParams({})),
    isApplicable: (ctx) => { var _a; return prop_1.ValidationReport.isApplicable((_a = ctx.structure) === null || _a === void 0 ? void 0 : _a.models[0]); },
    ensureCustomProperties: {
        attach: (ctx, data) => data.structure ? prop_1.ValidationReportProvider.attach(ctx, data.structure.models[0], void 0, true) : Promise.resolve(),
        detach: (data) => data.structure && prop_1.ValidationReportProvider.ref(data.structure.models[0], false)
    }
};
