// molstar/lib/commonjs/extensions/rcsb/validation-report/color/random-coil-index.js
"use strict";
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RandomCoilIndexColorThemeProvider = void 0;
exports.RandomCoilIndexColorTheme = RandomCoilIndexColorTheme;
const color_1 = require("../../../../mol-theme/color");
const param_definition_1 = require("../../../../mol-util/param-definition");
const color_2 = require("../../../../mol-util/color");
const structure_1 = require("../../../../mol-model/structure");
const prop_1 = require("../prop");
const DefaultColor = (0, color_2.Color)(0xCCCCCC);
function RandomCoilIndexColorTheme(ctx, props) {
    var _a, _b;
    let color = () => DefaultColor;
    const scale = color_2.ColorScale.create({
        reverse: true,
        domain: [0, 0.6],
        listOrName: 'red-yellow-blue',
    });
    const validationReport = ctx.structure && prop_1.ValidationReportProvider.get(ctx.structure.models[0]);
    const contextHash = validationReport === null || validationReport === void 0 ? void 0 : validationReport.version;
    const rci = (_a = validationReport === null || validationReport === void 0 ? void 0 : validationReport.value) === null || _a === void 0 ? void 0 : _a.rci;
    const model = (_b = ctx.structure) === null || _b === void 0 ? void 0 : _b.models[0];
    if (rci && model) {
        const residueIndex = model.atomicHierarchy.residueAtomSegments.index;
        const getColor = (element) => {
            const value = rci.get(residueIndex[element]);
            return value === undefined ? DefaultColor : scale.color(value);
        };
        color = (location) => {
            if (structure_1.StructureElement.Location.is(location) && location.unit.model === model) {
                return getColor(location.element);
            }
            else if (structure_1.Bond.isLocation(location) && location.aUnit.model === model) {
                return getColor(location.aUnit.elements[location.aIndex]);
            }
            return DefaultColor;
        };
    }
    return {
        factory: RandomCoilIndexColorTheme,
        granularity: 'group',
        preferSmoothing: true,
        color,
        props,
        contextHash,
        description: 'Assigns residue colors according to the Random Coil Index value. Data from wwPDB Validation Report, obtained via RCSB PDB.',
        legend: scale.legend
    };
}
exports.RandomCoilIndexColorThemeProvider = {
    name: prop_1.ValidationReport.Tag.RandomCoilIndex,
    label: 'Random Coil Index',
    category: color_1.ColorTheme.Category.Validation,
    factory: RandomCoilIndexColorTheme,
    getParams: () => ({}),
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues({}),
    isApplicable: (ctx) => !!ctx.structure && prop_1.ValidationReport.isApplicable(ctx.structure.models[0]) && structure_1.Model.isFromNmr(ctx.structure.models[0]),
    ensureCustomProperties: {
        attach: (ctx, data) => data.structure ? prop_1.ValidationReportProvider.attach(ctx, data.structure.models[0], void 0, true) : Promise.resolve(),
        detach: (data) => data.structure && prop_1.ValidationReportProvider.ref(data.structure.models[0], false)
    }
};
