// molstar/lib/commonjs/extensions/rcsb/validation-report/color/random-coil-index.js
"use strict";
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RandomCoilIndexColorThemeProvider = exports.RandomCoilIndexColorTheme = void 0;
var param_definition_1 = require("../../../../mol-util/param-definition");
var color_1 = require("../../../../mol-util/color");
var structure_1 = require("../../../../mol-model/structure");
var prop_1 = require("../prop");
var DefaultColor = (0, color_1.Color)(0xCCCCCC);
function RandomCoilIndexColorTheme(ctx, props) {
    var _a, _b;
    var color = function () { return DefaultColor; };
    var scale = color_1.ColorScale.create({
        reverse: true,
        domain: [0, 0.6],
        listOrName: 'red-yellow-blue',
    });
    var validationReport = ctx.structure && prop_1.ValidationReportProvider.get(ctx.structure.models[0]);
    var contextHash = validationReport === null || validationReport === void 0 ? void 0 : validationReport.version;
    var rci = (_a = validationReport === null || validationReport === void 0 ? void 0 : validationReport.value) === null || _a === void 0 ? void 0 : _a.rci;
    var model = (_b = ctx.structure) === null || _b === void 0 ? void 0 : _b.models[0];
    if (rci && model) {
        var residueIndex_1 = model.atomicHierarchy.residueAtomSegments.index;
        var getColor_1 = function (element) {
            var value = rci.get(residueIndex_1[element]);
            return value === undefined ? DefaultColor : scale.color(value);
        };
        color = function (location) {
            if (structure_1.StructureElement.Location.is(location) && location.unit.model === model) {
                return getColor_1(location.element);
            }
            else if (structure_1.Bond.isLocation(location) && location.aUnit.model === model) {
                return getColor_1(location.aUnit.elements[location.aIndex]);
            }
            return DefaultColor;
        };
    }
    return {
        factory: RandomCoilIndexColorTheme,
        granularity: 'group',
        preferSmoothing: true,
        color: color,
        props: props,
        contextHash: contextHash,
        description: 'Assigns residue colors according to the Random Coil Index value. Data from wwPDB Validation Report, obtained via RCSB PDB.',
        legend: scale.legend
    };
}
exports.RandomCoilIndexColorTheme = RandomCoilIndexColorTheme;
exports.RandomCoilIndexColorThemeProvider = {
    name: prop_1.ValidationReport.Tag.RandomCoilIndex,
    label: 'Random Coil Index',
    category: "Validation" /* Validation */,
    factory: RandomCoilIndexColorTheme,
    getParams: function () { return ({}); },
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues({}),
    isApplicable: function (ctx) { return !!ctx.structure && prop_1.ValidationReport.isApplicable(ctx.structure.models[0]) && structure_1.Model.isFromNmr(ctx.structure.models[0]); },
    ensureCustomProperties: {
        attach: function (ctx, data) { return data.structure ? prop_1.ValidationReportProvider.attach(ctx, data.structure.models[0], void 0, true) : Promise.resolve(); },
        detach: function (data) { return data.structure && prop_1.ValidationReportProvider.ref(data.structure.models[0], false); }
    }
};
