// molstar/lib/commonjs/mol-plugin-state/builder/structure/hierarchy.js
"use strict";
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrajectoryHierarchyBuilder = void 0;
var util_1 = require("../../../mol-data/util");
var mol_state_1 = require("../../../mol-state");
var mol_task_1 = require("../../../mol-task");
var debug_1 = require("../../../mol-util/debug");
var object_1 = require("../../../mol-util/object");
var param_definition_1 = require("../../../mol-util/param-definition");
var hierarchy_preset_1 = require("./hierarchy-preset");
var array_1 = require("../../../mol-util/array");
var TrajectoryHierarchyBuilder = /** @class */ (function () {
    function TrajectoryHierarchyBuilder(plugin) {
        var _this = this;
        this.plugin = plugin;
        this._providers = [];
        this.providerMap = new Map();
        this.defaultProvider = hierarchy_preset_1.PresetTrajectoryHierarchy.default;
        (0, object_1.objectForEach)(hierarchy_preset_1.PresetTrajectoryHierarchy, function (r) { return _this.registerPreset(r); });
    }
    TrajectoryHierarchyBuilder.prototype.resolveProvider = function (ref) {
        var _a;
        return typeof ref === 'string'
            ? (_a = hierarchy_preset_1.PresetTrajectoryHierarchy[ref]) !== null && _a !== void 0 ? _a : (0, util_1.arrayFind)(this._providers, function (p) { return p.id === ref; })
            : ref;
    };
    TrajectoryHierarchyBuilder.prototype.hasPreset = function (t) {
        for (var _i = 0, _a = this._providers; _i < _a.length; _i++) {
            var p = _a[_i];
            if (!p.isApplicable || p.isApplicable(t, this.plugin))
                return true;
        }
        return false;
    };
    Object.defineProperty(TrajectoryHierarchyBuilder.prototype, "providers", {
        get: function () { return this._providers; },
        enumerable: false,
        configurable: true
    });
    TrajectoryHierarchyBuilder.prototype.getPresets = function (t) {
        if (!t)
            return this.providers;
        var ret = [];
        for (var _i = 0, _a = this._providers; _i < _a.length; _i++) {
            var p = _a[_i];
            if (p.isApplicable && !p.isApplicable(t, this.plugin))
                continue;
            ret.push(p);
        }
        return ret;
    };
    TrajectoryHierarchyBuilder.prototype.getPresetSelect = function (t) {
        var options = [];
        for (var _i = 0, _a = this._providers; _i < _a.length; _i++) {
            var p = _a[_i];
            if (t && p.isApplicable && !p.isApplicable(t, this.plugin))
                continue;
            options.push([p.id, p.display.name]);
        }
        return param_definition_1.ParamDefinition.Select('auto', options);
    };
    TrajectoryHierarchyBuilder.prototype.getPresetsWithOptions = function (t) {
        var options = [];
        var map = Object.create(null);
        for (var _i = 0, _a = this._providers; _i < _a.length; _i++) {
            var p = _a[_i];
            if (p.isApplicable && !p.isApplicable(t, this.plugin))
                continue;
            options.push([p.id, p.display.name]);
            map[p.id] = p.params ? param_definition_1.ParamDefinition.Group(p.params(t, this.plugin)) : param_definition_1.ParamDefinition.EmptyGroup();
        }
        if (options.length === 0)
            return param_definition_1.ParamDefinition.MappedStatic('', { '': param_definition_1.ParamDefinition.EmptyGroup() });
        return param_definition_1.ParamDefinition.MappedStatic(options[0][0], map, { options: options });
    };
    TrajectoryHierarchyBuilder.prototype.registerPreset = function (provider) {
        if (this.providerMap.has(provider.id)) {
            throw new Error("Hierarchy provider with id '".concat(provider.id, "' already registered."));
        }
        this._providers.push(provider);
        this.providerMap.set(provider.id, provider);
    };
    TrajectoryHierarchyBuilder.prototype.unregisterPreset = function (provider) {
        this.providerMap.delete(provider.id);
        (0, array_1.arrayRemoveInPlace)(this._providers, provider);
    };
    TrajectoryHierarchyBuilder.prototype.applyPreset = function (parent, providerRef, params) {
        var _this = this;
        var provider = this.resolveProvider(providerRef);
        if (!provider)
            return;
        var state = this.plugin.state.data;
        var cell = mol_state_1.StateObjectRef.resolveAndCheck(state, parent);
        if (!cell) {
            if (!debug_1.isProductionMode)
                console.warn("Applying hierarchy preset provider to bad cell.");
            return;
        }
        var prms = params || (provider.params
            ? param_definition_1.ParamDefinition.getDefaultValues(provider.params(cell.obj, this.plugin))
            : {});
        var task = mol_task_1.Task.create("".concat(provider.display.name), function () { return provider.apply(cell, prms, _this.plugin); });
        return this.plugin.runTask(task);
    };
    return TrajectoryHierarchyBuilder;
}());
exports.TrajectoryHierarchyBuilder = TrajectoryHierarchyBuilder;
