// molstar/lib/commonjs/mol-model/structure/structure/util/unique-subset-builder.js
"use strict";
/**
 * Copyright (c) 2017 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StructureUniqueSubsetBuilder = void 0;
var int_1 = require("../../../../mol-data/int");
var util_1 = require("../../../../mol-data/util");
var symmetry_1 = require("../symmetry");
var structure_1 = require("../structure");
var generic_1 = require("../../../../mol-data/generic");
var StructureUniqueSubsetBuilder = /** @class */ (function () {
    function StructureUniqueSubsetBuilder(parent) {
        this.parent = parent;
        this.ids = [];
        this.unitMap = int_1.IntMap.Mutable();
        this.parentId = -1;
        this.currentUnit = generic_1.UniqueArray.create();
        this.elementCount = 0;
    }
    StructureUniqueSubsetBuilder.prototype.addToUnit = function (parentId, e) {
        var unit = this.unitMap.get(parentId);
        if (!!unit) {
            if (generic_1.UniqueArray.add(unit, e, e))
                this.elementCount++;
        }
        else {
            var arr = generic_1.UniqueArray.create();
            generic_1.UniqueArray.add(arr, e, e);
            this.unitMap.set(parentId, arr);
            this.ids[this.ids.length] = parentId;
            this.elementCount++;
        }
    };
    StructureUniqueSubsetBuilder.prototype.has = function (parentId, e) {
        var unit = this.unitMap.get(parentId);
        if (!unit)
            return false;
        return generic_1.UniqueArray.has(unit, e);
    };
    StructureUniqueSubsetBuilder.prototype.beginUnit = function (parentId) {
        this.parentId = parentId;
        if (this.unitMap.has(parentId)) {
            this.currentUnit = this.unitMap.get(parentId);
        }
        else {
            this.currentUnit = this.currentUnit.array.length > 0 ? generic_1.UniqueArray.create() : this.currentUnit;
        }
    };
    StructureUniqueSubsetBuilder.prototype.addElement = function (e) {
        if (generic_1.UniqueArray.add(this.currentUnit, e, e))
            this.elementCount++;
    };
    StructureUniqueSubsetBuilder.prototype.commitUnit = function () {
        if (this.currentUnit.array.length === 0 || this.unitMap.has(this.parentId))
            return;
        this.ids[this.ids.length] = this.parentId;
        this.unitMap.set(this.parentId, this.currentUnit);
        this.parentId = -1;
    };
    StructureUniqueSubsetBuilder.prototype.getStructure = function () {
        if (this.isEmpty)
            return structure_1.Structure.Empty;
        var newUnits = [];
        (0, util_1.sortArray)(this.ids);
        var symmGroups = symmetry_1.StructureSymmetry.UnitEquivalenceBuilder();
        for (var i = 0, _i = this.ids.length; i < _i; i++) {
            var id = this.ids[i];
            var parent_1 = this.parent.unitMap.get(id);
            var unit = this.unitMap.get(id).array;
            var l = unit.length;
            // if the length is the same, just copy the old unit.
            if (unit.length === parent_1.elements.length) {
                newUnits[newUnits.length] = parent_1;
                symmGroups.add(parent_1.id, parent_1);
                continue;
            }
            if (l > 1)
                (0, util_1.sortArray)(unit);
            var child = parent_1.getChild(int_1.SortedArray.ofSortedArray(unit));
            var pivot = symmGroups.add(child.id, child);
            if (child !== pivot)
                child = pivot.applyOperator(child.id, child.conformation.operator, true);
            newUnits[newUnits.length] = child;
        }
        return structure_1.Structure.create(newUnits, { parent: this.parent });
    };
    Object.defineProperty(StructureUniqueSubsetBuilder.prototype, "isEmpty", {
        get: function () {
            return this.elementCount === 0;
        },
        enumerable: false,
        configurable: true
    });
    return StructureUniqueSubsetBuilder;
}());
exports.StructureUniqueSubsetBuilder = StructureUniqueSubsetBuilder;
