// molstar/lib/commonjs/servers/model/server/api-web.js
"use strict";
/**
 * Copyright (c) 2018-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.initWebApi = exports.createResultWriter = void 0;
var tslib_1 = require("tslib");
var fs = tslib_1.__importStar(require("fs"));
var path = tslib_1.__importStar(require("path"));
var bodyParser = tslib_1.__importStar(require("body-parser"));
var config_1 = require("../config");
var console_logger_1 = require("../../../mol-util/console-logger");
var query_1 = require("./query");
var jobs_1 = require("./jobs");
var api_1 = require("./api");
var api_schema_1 = require("./api-schema");
var swagger_ui_1 = require("../../common/swagger-ui");
var api_web_multiple_1 = require("./api-web-multiple");
var writer_1 = require("../utils/writer");
var string_1 = require("../../../mol-util/string");
function makePath(p) {
    return config_1.ModelServerConfig.apiPrefix + '/' + p;
}
var responseMap = new Map();
function processNextJob() {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var job, writer, e_1;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!jobs_1.JobManager.hasNext())
                        return [2 /*return*/];
                    job = jobs_1.JobManager.getNext();
                    responseMap.delete(job.id);
                    writer = job.writer;
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, 4, 5]);
                    return [4 /*yield*/, (0, query_1.resolveJob)(job)];
                case 2:
                    _a.sent();
                    return [3 /*break*/, 5];
                case 3:
                    e_1 = _a.sent();
                    console_logger_1.ConsoleLogger.errorId(job.id, '' + e_1);
                    writer.doError(404, '' + e_1);
                    return [3 /*break*/, 5];
                case 4:
                    writer.end();
                    console_logger_1.ConsoleLogger.logId(job.id, 'Query', 'Finished.');
                    setImmediate(processNextJob);
                    return [7 /*endfinally*/];
                case 5: return [2 /*return*/];
            }
        });
    });
}
function createResultWriter(response, params) {
    var filenameBase = params.entryId && params.queryName
        ? "".concat(params.entryId, "_").concat((0, string_1.splitCamelCase)(params.queryName.replace(/\s/g, '_'), '-').toLowerCase())
        : "result";
    return new writer_1.SimpleResponseResultWriter(params.filename || "".concat(filenameBase, ".").concat(params.encoding), response, params.encoding === 'bcif', params.download);
}
exports.createResultWriter = createResultWriter;
function mapQuery(app, queryName, queryDefinition) {
    function createJob(queryParams, req, res) {
        var entryId = req.params.id;
        var commonParams = (0, api_1.normalizeRestCommonParams)(req.query);
        var resultWriterParams = { encoding: commonParams.encoding, download: !!commonParams.download, filename: commonParams.filename, entryId: entryId, queryName: queryName };
        var jobId = jobs_1.JobManager.add({
            entries: [(0, jobs_1.JobEntry)({
                    sourceId: commonParams.data_source || config_1.ModelServerConfig.defaultSource,
                    entryId: entryId,
                    queryName: queryName,
                    queryParams: queryParams,
                    modelNums: commonParams.model_nums,
                    copyAllCategories: !!commonParams.copy_all_categories,
                    transform: commonParams.transform
                })],
            writer: createResultWriter(res, resultWriterParams),
            options: { binary: commonParams.encoding === 'bcif', encoding: commonParams.encoding }
        });
        responseMap.set(jobId, res);
        if (jobs_1.JobManager.size === 1)
            processNextJob();
    }
    app.get(makePath('v1/:id/' + queryName), function (req, res) {
        var queryParams = (0, api_1.normalizeRestQueryParams)(queryDefinition, req.query);
        createJob(queryParams, req, res);
    });
    app.post(makePath('v1/:id/' + queryName), function (req, res) {
        var queryParams = req.body;
        createJob(queryParams, req, res);
    });
}
function serveStatic(req, res) {
    var source = req.params.source === 'bcif'
        ? 'pdb-bcif'
        : req.params.source === 'cif'
            ? 'pdb-cif'
            : req.params.source;
    var id = req.params.id;
    var _a = (0, config_1.mapSourceAndIdToFilename)(source, id), fn = _a[0], format = _a[1];
    var binary = format === 'bcif' || fn.indexOf('.bcif') > 0;
    if (!fn || !fs.existsSync(fn)) {
        res.status(404);
        res.end();
        return;
    }
    fs.readFile(fn, function (err, data) {
        if (err) {
            res.status(404);
            res.end();
            return;
        }
        var f = path.parse(fn);
        res.writeHead(200, {
            'Content-Type': binary ? 'application/octet-stream' : 'text/plain; charset=utf-8',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Headers': 'X-Requested-With',
            'Content-Disposition': "inline; filename=\"".concat(f.name).concat(f.ext, "\"")
        });
        res.write(data);
        res.end();
    });
}
function createMultiJob(spec, res) {
    var _a;
    var writer = spec.asTarGz
        ? new writer_1.TarballResponseResultWriter((0, api_web_multiple_1.getMultiQuerySpecFilename)(), res)
        : createResultWriter(res, { encoding: spec.encoding, download: !!spec.download, filename: spec.filename });
    if (spec.queries.length > config_1.ModelServerConfig.maxQueryManyQueries) {
        writer.doError(400, "query-many queries limit (".concat(config_1.ModelServerConfig.maxQueryManyQueries, ") exceeded."));
        return;
    }
    var jobId = jobs_1.JobManager.add({
        entries: spec.queries.map(function (q) { return (0, jobs_1.JobEntry)({
            sourceId: q.data_source || config_1.ModelServerConfig.defaultSource,
            entryId: q.entryId,
            queryName: q.query,
            queryParams: q.params || {},
            modelNums: q.model_nums,
            copyAllCategories: !!q.copy_all_categories
        }); }),
        writer: writer,
        options: { binary: ((_a = spec.encoding) === null || _a === void 0 ? void 0 : _a.toLowerCase()) === 'bcif', tarball: spec.asTarGz }
    });
    responseMap.set(jobId, res);
    if (jobs_1.JobManager.size === 1)
        processNextJob();
}
function initWebApi(app) {
    app.use(bodyParser.json({ limit: '1mb' }));
    app.get(makePath('static/:source/:id'), function (req, res) { return serveStatic(req, res); });
    app.get(makePath('v1/static/:source/:id'), function (req, res) { return serveStatic(req, res); });
    app.get(makePath('v1/query-many'), function (req, res) {
        var query = /\?query=(.*)$/.exec(req.url)[1];
        var params = JSON.parse(decodeURIComponent(query));
        createMultiJob(params, res);
    });
    app.post(makePath('v1/query-many'), function (req, res) {
        var params = req.body;
        req.setTimeout;
        createMultiJob(params, res);
    });
    app.use(bodyParser.json({ limit: '20mb' }));
    for (var _i = 0, QueryList_1 = api_1.QueryList; _i < QueryList_1.length; _i++) {
        var q = QueryList_1[_i];
        mapQuery(app, q.name, q.definition);
    }
    var schema = (0, api_schema_1.getApiSchema)();
    app.get(makePath('openapi.json'), function (req, res) {
        res.writeHead(200, {
            'Content-Type': 'application/json; charset=utf-8',
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Headers': 'X-Requested-With'
        });
        res.end(JSON.stringify(schema));
    });
    app.use(makePath(''), (0, swagger_ui_1.swaggerUiAssetsHandler)());
    app.get(makePath(''), (0, swagger_ui_1.swaggerUiIndexHandler)({
        openapiJsonUrl: makePath('openapi.json'),
        apiPrefix: config_1.ModelServerConfig.apiPrefix,
        title: 'ModelServer API',
        shortcutIconLink: api_schema_1.shortcutIconLink
    }));
}
exports.initWebApi = initWebApi;
