// molstar/lib/commonjs/servers/model/server/jobs.js
"use strict";
/**
 * Copyright (c) 2018-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobManager = exports.createJob = exports.JobEntry = void 0;
var mol_util_1 = require("../../../mol-util");
var api_1 = require("./api");
var generic_1 = require("../../../mol-data/generic");
function JobEntry(definition) {
    var queryDefinition = (0, api_1.getQueryByName)(definition.queryName);
    if (!queryDefinition)
        throw new Error("Query '".concat(definition.queryName, "' is not supported."));
    var normalizedParams = definition.queryParams;
    var sourceId = definition.sourceId || '_local_';
    return {
        job: void 0,
        key: "".concat(sourceId, "/").concat(definition.entryId),
        sourceId: sourceId,
        entryId: definition.entryId,
        queryDefinition: queryDefinition,
        normalizedParams: normalizedParams,
        modelNums: definition.modelNums,
        copyAllCategories: !!definition.copyAllCategories,
        transform: definition.transform
    };
}
exports.JobEntry = JobEntry;
function createJob(definition) {
    var _a, _b;
    var job = {
        id: mol_util_1.UUID.create22(),
        datetime_utc: "".concat(new Date().toISOString().replace(/T/, ' ').replace(/\..+/, '')),
        entries: definition.entries,
        writer: definition.writer,
        responseFormat: {
            tarball: !!((_a = definition === null || definition === void 0 ? void 0 : definition.options) === null || _a === void 0 ? void 0 : _a.tarball),
            encoding: ((_b = definition === null || definition === void 0 ? void 0 : definition.options) === null || _b === void 0 ? void 0 : _b.encoding) ? definition.options.encoding : !!(definition.options && definition.options.binary) ? 'bcif' : 'cif'
        },
        outputFilename: definition.options && definition.options.outputFilename
    };
    definition.entries.forEach(function (e) { return e.job = job; });
    return job;
}
exports.createJob = createJob;
var _JobQueue = /** @class */ (function () {
    function _JobQueue() {
        this.list = (0, generic_1.LinkedList)();
    }
    Object.defineProperty(_JobQueue.prototype, "size", {
        get: function () {
            return this.list.count;
        },
        enumerable: false,
        configurable: true
    });
    _JobQueue.prototype.add = function (definition) {
        var job = createJob(definition);
        this.list.addLast(job);
        return job.id;
    };
    _JobQueue.prototype.hasNext = function () {
        return this.list.count > 0;
    };
    _JobQueue.prototype.getNext = function () {
        return this.list.removeFirst();
    };
    /** Sort the job list by key = sourceId/entryId */
    _JobQueue.prototype.sort = function () {
        if (this.list.count === 0)
            return;
        var jobs = [];
        for (var j = this.list.first; !!j; j = j.next) {
            jobs[jobs.length] = j.value;
        }
        jobs.sort(function (a, b) { var _a, _b; return ((_a = a.entries[0]) === null || _a === void 0 ? void 0 : _a.key) < ((_b = b.entries[0]) === null || _b === void 0 ? void 0 : _b.key) ? -1 : 1; });
        this.list = (0, generic_1.LinkedList)();
        for (var _i = 0, jobs_1 = jobs; _i < jobs_1.length; _i++) {
            var j = jobs_1[_i];
            this.list.addLast(j);
        }
    };
    return _JobQueue;
}());
exports.JobManager = new _JobQueue();
