// molstar/lib/commonjs/mol-model-props/computed/helix-orientation.js
"use strict";
/**
 * Copyright (c) 2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HelixOrientationProvider = exports.HelixOrientationParams = void 0;
var tslib_1 = require("tslib");
var custom_property_1 = require("../../mol-model/custom-property");
var custom_model_property_1 = require("../common/custom-model-property");
var helix_orientation_1 = require("./helix-orientation/helix-orientation");
exports.HelixOrientationParams = {};
exports.HelixOrientationProvider = custom_model_property_1.CustomModelProperty.createProvider({
    label: 'Helix Orientation',
    descriptor: (0, custom_property_1.CustomPropertyDescriptor)({
        name: 'molstar_helix_orientation'
    }),
    type: 'dynamic',
    defaultParams: {},
    getParams: function () { return ({}); },
    isApplicable: function (data) { return true; },
    obtain: function (ctx, data) { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, { value: (0, helix_orientation_1.calcHelixOrientation)(data) }];
        });
    }); }
});
