// molstar/lib/servers/model/server/api-local.js
/**
 * Copyright (c) 2018 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
import { __awaiter, __generator } from "tslib";
import { Mat4 } from '../../../mol-math/linear-algebra';
import { ConsoleLogger } from '../../../mol-util/console-logger';
import { now } from '../../../mol-util/now';
import { PerformanceMonitor } from '../../../mol-util/performance-monitor';
import { FileResultWriter, TarballFileResultWriter } from '../utils/writer';
import { JobEntry, JobManager } from './jobs';
import { resolveJob } from './query';
import { StructureCache } from './structure-wrapper';
export function runLocal(input) {
    return __awaiter(this, void 0, void 0, function () {
        var _i, input_1, job_1, binary, started, job, key, progress, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!input.length) {
                        ConsoleLogger.error('Local', 'No input');
                        return [2 /*return*/];
                    }
                    for (_i = 0, input_1 = input; _i < input_1.length; _i++) {
                        job_1 = input_1[_i];
                        binary = /\.bcif/.test(job_1.output);
                        JobManager.add({
                            entries: job_1.queries.map(function (q) {
                                var _a;
                                return JobEntry({
                                    entryId: q.input,
                                    queryName: q.query,
                                    queryParams: q.params || {},
                                    modelNums: q.modelNums,
                                    transform: (_a = q.transform) !== null && _a !== void 0 ? _a : Mat4.identity(),
                                    copyAllCategories: !!q.copyAllCategories
                                });
                            }),
                            writer: job_1.asTarGz
                                ? new TarballFileResultWriter(job_1.output, job_1.gzipLevel)
                                : new FileResultWriter(job_1.output),
                            options: {
                                outputFilename: job_1.output,
                                binary: binary,
                                tarball: job_1.asTarGz
                            }
                        });
                    }
                    JobManager.sort();
                    started = now();
                    job = JobManager.getNext();
                    key = job.entries[0].key;
                    progress = 0;
                    _a.label = 1;
                case 1:
                    if (!job) return [3 /*break*/, 6];
                    _a.label = 2;
                case 2:
                    _a.trys.push([2, 4, , 5]);
                    return [4 /*yield*/, resolveJob(job)];
                case 3:
                    _a.sent();
                    job.writer.end();
                    ConsoleLogger.logId(job.id, 'Query', 'Written.');
                    if (job.entries.length > 0)
                        StructureCache.expireAll();
                    if (JobManager.hasNext()) {
                        job = JobManager.getNext();
                        if (key !== job.entries[0].key)
                            StructureCache.expire(key);
                        key = job.entries[0].key;
                    }
                    else {
                        return [3 /*break*/, 6];
                    }
                    return [3 /*break*/, 5];
                case 4:
                    e_1 = _a.sent();
                    ConsoleLogger.errorId(job.id, e_1);
                    if (JobManager.hasNext()) {
                        job = JobManager.getNext();
                        if (key !== job.entries[0].key)
                            StructureCache.expire(key);
                        key = job.entries[0].key;
                    }
                    else {
                        return [3 /*break*/, 6];
                    }
                    return [3 /*break*/, 5];
                case 5:
                    ConsoleLogger.log('Progress', "[".concat(++progress, "/").concat(input.length, "] after ").concat(PerformanceMonitor.format(now() - started), "."));
                    return [3 /*break*/, 1];
                case 6:
                    ConsoleLogger.log('Progress', "Done in ".concat(PerformanceMonitor.format(now() - started), "."));
                    StructureCache.expireAll();
                    return [2 /*return*/];
            }
        });
    });
}
