// molstar/lib/commonjs/mol-plugin-state/formats/structure.js
"use strict";
/**
 * Copyright (c) 2018-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuiltInStructureFormats = exports.XtcProvider = exports.DcdProvider = exports.PsfProvider = exports.StructureFormatCategory = void 0;
var tslib_1 = require("tslib");
var transforms_1 = require("../transforms");
var provider_1 = require("./provider");
exports.StructureFormatCategory = 'Structure';
exports.PsfProvider = (0, provider_1.DataFormatProvider)({
    label: 'PSF',
    description: 'PSF',
    category: exports.StructureFormatCategory,
    stringExtensions: ['psf'],
    parse: function (plugin, data) { return (0, tslib_1.__awaiter)(void 0, void 0, void 0, function () {
        var format, topology;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0:
                    format = plugin.state.data.build()
                        .to(data)
                        .apply(transforms_1.StateTransforms.Data.ParsePsf, {}, { state: { isGhost: true } });
                    topology = format.apply(transforms_1.StateTransforms.Model.TopologyFromPsf);
                    return [4 /*yield*/, format.commit()];
                case 1:
                    _a.sent();
                    return [2 /*return*/, { format: format.selector, topology: topology.selector }];
            }
        });
    }); }
});
exports.DcdProvider = (0, provider_1.DataFormatProvider)({
    label: 'DCD',
    description: 'DCD',
    category: exports.StructureFormatCategory,
    binaryExtensions: ['dcd'],
    parse: function (plugin, data) {
        var coordinates = plugin.state.data.build()
            .to(data)
            .apply(transforms_1.StateTransforms.Model.CoordinatesFromDcd);
        return coordinates.commit();
    }
});
exports.XtcProvider = (0, provider_1.DataFormatProvider)({
    label: 'XTC',
    description: 'XTC',
    category: exports.StructureFormatCategory,
    binaryExtensions: ['xtc'],
    parse: function (plugin, data) {
        var coordinates = plugin.state.data.build()
            .to(data)
            .apply(transforms_1.StateTransforms.Model.CoordinatesFromXtc);
        return coordinates.commit();
    }
});
exports.BuiltInStructureFormats = [
    ['psf', exports.PsfProvider],
    ['dcd', exports.DcdProvider],
    ['xtc', exports.XtcProvider],
];
