// molstar/lib/commonjs/mol-canvas3d/helper/handle-helper.js
"use strict";
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHandleLoci = exports.HandleGroup = exports.HandleHelper = exports.HandleHelperParams = void 0;
var tslib_1 = require("tslib");
var scene_1 = require("../../mol-gl/scene");
var mesh_builder_1 = require("../../mol-geo/geometry/mesh/mesh-builder");
var linear_algebra_1 = require("../../mol-math/linear-algebra");
var sphere_1 = require("../../mol-geo/geometry/mesh/builder/sphere");
var mesh_1 = require("../../mol-geo/geometry/mesh/mesh");
var names_1 = require("../../mol-util/color/names");
var cylinder_1 = require("../../mol-geo/geometry/mesh/builder/cylinder");
var mol_util_1 = require("../../mol-util");
var geometry_1 = require("../../mol-math/geometry");
var param_definition_1 = require("../../mol-util/param-definition");
var immer_1 = (0, tslib_1.__importDefault)(require("immer"));
var shape_1 = require("../../mol-model/shape");
var loci_1 = require("../../mol-model/loci");
var marker_action_1 = require("../../mol-util/marker-action");
var visual_1 = require("../../mol-repr/visual");
var int_1 = require("../../mol-data/int");
var HandleParams = (0, tslib_1.__assign)((0, tslib_1.__assign)({}, mesh_1.Mesh.Params), { alpha: (0, tslib_1.__assign)((0, tslib_1.__assign)({}, mesh_1.Mesh.Params.alpha), { defaultValue: 1 }), ignoreLight: (0, tslib_1.__assign)((0, tslib_1.__assign)({}, mesh_1.Mesh.Params.ignoreLight), { defaultValue: true }), colorX: param_definition_1.ParamDefinition.Color(names_1.ColorNames.red, { isEssential: true }), colorY: param_definition_1.ParamDefinition.Color(names_1.ColorNames.green, { isEssential: true }), colorZ: param_definition_1.ParamDefinition.Color(names_1.ColorNames.blue, { isEssential: true }), scale: param_definition_1.ParamDefinition.Numeric(0.33, { min: 0.1, max: 2, step: 0.1 }, { isEssential: true }) });
exports.HandleHelperParams = {
    handle: param_definition_1.ParamDefinition.MappedStatic('off', {
        on: param_definition_1.ParamDefinition.Group(HandleParams),
        off: param_definition_1.ParamDefinition.Group({})
    }, { cycle: true, description: 'Show handle tool' }),
};
var HandleHelper = /** @class */ (function () {
    function HandleHelper(webgl, props) {
        var _this = this;
        if (props === void 0) { props = {}; }
        this.webgl = webgl;
        this.props = {
            handle: { name: 'off', params: {} }
        };
        this._transform = (0, linear_algebra_1.Mat4)();
        this.eachGroup = function (loci, apply) {
            if (!_this.renderObject)
                return false;
            if (!isHandleLoci(loci))
                return false;
            var changed = false;
            var groupCount = _this.renderObject.values.uGroupCount.ref.value;
            var elements = loci.elements;
            for (var _i = 0, elements_1 = elements; _i < elements_1.length; _i++) {
                var _a = elements_1[_i], groupId = _a.groupId, instanceId = _a.instanceId;
                var idx = instanceId * groupCount + groupId;
                if (apply(int_1.Interval.ofSingleton(idx)))
                    changed = true;
            }
            return changed;
        };
        this.scene = scene_1.Scene.create(webgl);
        this.setProps(props);
    }
    HandleHelper.prototype.getBoundingSphere = function (out, instanceId) {
        if (this.renderObject) {
            geometry_1.Sphere3D.copy(out, this.renderObject.values.invariantBoundingSphere.ref.value);
            linear_algebra_1.Mat4.fromArray(this._transform, this.renderObject.values.aTransform.ref.value, instanceId * 16);
            geometry_1.Sphere3D.transform(out, out, this._transform);
        }
        return out;
    };
    HandleHelper.prototype.setProps = function (props) {
        var _this = this;
        this.props = (0, immer_1.default)(this.props, function (p) {
            if (props.handle !== undefined) {
                p.handle.name = props.handle.name;
                if (props.handle.name === 'on') {
                    _this.scene.clear();
                    var params = (0, tslib_1.__assign)((0, tslib_1.__assign)({}, props.handle.params), { scale: props.handle.params.scale * _this.webgl.pixelRatio });
                    _this.renderObject = createHandleRenderObject(params);
                    _this.renderObject.state.noClip = true;
                    _this.scene.add(_this.renderObject);
                    _this.scene.commit();
                    p.handle.params = (0, tslib_1.__assign)({}, props.handle.params);
                }
            }
        });
    };
    Object.defineProperty(HandleHelper.prototype, "isEnabled", {
        get: function () {
            return this.props.handle.name === 'on';
        },
        enumerable: false,
        configurable: true
    });
    // TODO could be a lists of position/rotation if we want to show more than one handle tool,
    //      they would be distingishable by their instanceId
    HandleHelper.prototype.update = function (camera, position, rotation) {
        if (!this.renderObject)
            return;
        linear_algebra_1.Mat4.setTranslation(this.renderObject.values.aTransform.ref.value, position);
        linear_algebra_1.Mat4.fromMat3(this.renderObject.values.aTransform.ref.value, rotation);
        // TODO make invariant to camera scaling by adjusting renderObject transform
        mol_util_1.ValueCell.update(this.renderObject.values.aTransform, this.renderObject.values.aTransform.ref.value);
        this.scene.update([this.renderObject], true);
    };
    HandleHelper.prototype.getLoci = function (pickingId) {
        var objectId = pickingId.objectId, groupId = pickingId.groupId, instanceId = pickingId.instanceId;
        if (!this.renderObject || objectId !== this.renderObject.id)
            return loci_1.EmptyLoci;
        return HandleLoci(this, groupId, instanceId);
    };
    HandleHelper.prototype.mark = function (loci, action) {
        if (!marker_action_1.MarkerActions.is(marker_action_1.MarkerActions.Highlighting, action))
            return false;
        if (!isHandleLoci(loci))
            return false;
        if (loci.data !== this)
            return false;
        return visual_1.Visual.mark(this.renderObject, loci, action, this.eachGroup);
    };
    return HandleHelper;
}());
exports.HandleHelper = HandleHelper;
function createHandleMesh(scale, mesh) {
    var state = mesh_builder_1.MeshBuilder.createState(512, 256, mesh);
    var radius = 0.05 * scale;
    var x = linear_algebra_1.Vec3.scale((0, linear_algebra_1.Vec3)(), linear_algebra_1.Vec3.unitX, scale);
    var y = linear_algebra_1.Vec3.scale((0, linear_algebra_1.Vec3)(), linear_algebra_1.Vec3.unitY, scale);
    var z = linear_algebra_1.Vec3.scale((0, linear_algebra_1.Vec3)(), linear_algebra_1.Vec3.unitZ, scale);
    var cylinderProps = { radiusTop: radius, radiusBottom: radius, radialSegments: 32 };
    state.currentGroup = exports.HandleGroup.TranslateScreenXY;
    (0, sphere_1.addSphere)(state, linear_algebra_1.Vec3.origin, radius * 3, 2);
    state.currentGroup = exports.HandleGroup.TranslateObjectX;
    (0, sphere_1.addSphere)(state, x, radius, 2);
    (0, cylinder_1.addCylinder)(state, linear_algebra_1.Vec3.origin, x, 1, cylinderProps);
    state.currentGroup = exports.HandleGroup.TranslateObjectY;
    (0, sphere_1.addSphere)(state, y, radius, 2);
    (0, cylinder_1.addCylinder)(state, linear_algebra_1.Vec3.origin, y, 1, cylinderProps);
    state.currentGroup = exports.HandleGroup.TranslateObjectZ;
    (0, sphere_1.addSphere)(state, z, radius, 2);
    (0, cylinder_1.addCylinder)(state, linear_algebra_1.Vec3.origin, z, 1, cylinderProps);
    // TODO add more helper geometries for the other HandleGroup options
    // TODO add props to create subset of geometries
    return mesh_builder_1.MeshBuilder.getMesh(state);
}
exports.HandleGroup = {
    None: 0,
    TranslateScreenXY: 1,
    // TranslateScreenZ: 2,
    TranslateObjectX: 3,
    TranslateObjectY: 4,
    TranslateObjectZ: 5,
    // TranslateObjectXY: 6,
    // TranslateObjectXZ: 7,
    // TranslateObjectYZ: 8,
    // RotateScreenZ: 9,
    // RotateObjectX: 10,
    // RotateObjectY: 11,
    // RotateObjectZ: 12,
};
function HandleLoci(handleHelper, groupId, instanceId) {
    return (0, loci_1.DataLoci)('handle', handleHelper, [{ groupId: groupId, instanceId: instanceId }], function (boundingSphere) { return handleHelper.getBoundingSphere(boundingSphere, instanceId); }, function () { return "Handle Helper | Group Id ".concat(groupId, " | Instance Id ").concat(instanceId); });
}
function isHandleLoci(x) {
    return x.kind === 'data-loci' && x.tag === 'handle';
}
exports.isHandleLoci = isHandleLoci;
function getHandleShape(props, shape) {
    var scale = 10 * props.scale;
    var mesh = createHandleMesh(scale, shape === null || shape === void 0 ? void 0 : shape.geometry);
    mesh.setBoundingSphere(geometry_1.Sphere3D.create(linear_algebra_1.Vec3.create(scale / 2, scale / 2, scale / 2), scale + scale / 4));
    var getColor = function (groupId) {
        switch (groupId) {
            case exports.HandleGroup.TranslateObjectX: return props.colorX;
            case exports.HandleGroup.TranslateObjectY: return props.colorY;
            case exports.HandleGroup.TranslateObjectZ: return props.colorZ;
            default: return names_1.ColorNames.grey;
        }
    };
    return shape_1.Shape.create('handle', {}, mesh, getColor, function () { return 1; }, function () { return ''; });
}
function createHandleRenderObject(props) {
    var shape = getHandleShape(props);
    return shape_1.Shape.createRenderObject(shape, props);
}
