// molstar/lib/commonjs/mol-repr/structure/params.js
"use strict";
/**
 * Copyright (c) 2018-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StructureTextureMeshParams = exports.StructureDirectVolumeParams = exports.StructureTextParams = exports.StructureLinesParams = exports.StructurePointsParams = exports.StructureCylindersParams = exports.StructureSpheresParams = exports.StructureMeshParams = exports.StructureParams = void 0;
exports.getUnitKindsParam = getUnitKindsParam;
const direct_volume_1 = require("../../mol-geo/geometry/direct-volume/direct-volume");
const lines_1 = require("../../mol-geo/geometry/lines/lines");
const mesh_1 = require("../../mol-geo/geometry/mesh/mesh");
const points_1 = require("../../mol-geo/geometry/points/points");
const spheres_1 = require("../../mol-geo/geometry/spheres/spheres");
const cylinders_1 = require("../../mol-geo/geometry/cylinders/cylinders");
const text_1 = require("../../mol-geo/geometry/text/text");
const texture_mesh_1 = require("../../mol-geo/geometry/texture-mesh/texture-mesh");
const param_definition_1 = require("../../mol-util/param-definition");
const common_1 = require("./visual/util/common");
function getUnitKindsParam(defaultValue) {
    return param_definition_1.ParamDefinition.MultiSelect(defaultValue, common_1.UnitKindOptions, { description: 'For which kinds of units/chains to show the representation visuals.' });
}
exports.StructureParams = {
    unitKinds: getUnitKindsParam(['atomic', 'spheres']),
    includeParent: param_definition_1.ParamDefinition.Boolean(false, { isHidden: true }),
};
exports.StructureMeshParams = { ...mesh_1.Mesh.Params };
exports.StructureSpheresParams = { ...spheres_1.Spheres.Params };
exports.StructureCylindersParams = { ...cylinders_1.Cylinders.Params };
exports.StructurePointsParams = { ...points_1.Points.Params };
exports.StructureLinesParams = { ...lines_1.Lines.Params };
exports.StructureTextParams = { ...text_1.Text.Params };
exports.StructureDirectVolumeParams = { ...direct_volume_1.DirectVolume.Params };
exports.StructureTextureMeshParams = { ...texture_mesh_1.TextureMesh.Params };
