// molstar/lib/commonjs/mol-model-props/computed/representations/interactions.js
"use strict";
/**
 * Copyright (c) 2019-2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractionsRepresentationProvider = exports.InteractionRepresentation = exports.getInteractionParams = exports.InteractionsParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var representation_1 = require("../../../mol-repr/representation");
var representation_2 = require("../../../mol-repr/structure/representation");
var interactions_intra_unit_cylinder_1 = require("./interactions-intra-unit-cylinder");
var interactions_1 = require("../interactions");
var interactions_inter_unit_cylinder_1 = require("./interactions-inter-unit-cylinder");
var params_1 = require("../../../mol-repr/structure/params");
var InteractionsVisuals = {
    'intra-unit': function (ctx, getParams) { return (0, representation_2.UnitsRepresentation)('Intra-unit interactions cylinder', ctx, getParams, interactions_intra_unit_cylinder_1.InteractionsIntraUnitVisual); },
    'inter-unit': function (ctx, getParams) { return (0, representation_2.ComplexRepresentation)('Inter-unit interactions cylinder', ctx, getParams, interactions_inter_unit_cylinder_1.InteractionsInterUnitVisual); },
};
exports.InteractionsParams = (0, tslib_1.__assign)((0, tslib_1.__assign)((0, tslib_1.__assign)({}, interactions_intra_unit_cylinder_1.InteractionsIntraUnitParams), interactions_inter_unit_cylinder_1.InteractionsInterUnitParams), { unitKinds: (0, params_1.getUnitKindsParam)(['atomic']), sizeFactor: param_definition_1.ParamDefinition.Numeric(0.2, { min: 0.01, max: 1, step: 0.01 }), visuals: param_definition_1.ParamDefinition.MultiSelect(['intra-unit', 'inter-unit'], param_definition_1.ParamDefinition.objectToOptions(InteractionsVisuals)) });
function getInteractionParams(ctx, structure) {
    return param_definition_1.ParamDefinition.clone(exports.InteractionsParams);
}
exports.getInteractionParams = getInteractionParams;
function InteractionRepresentation(ctx, getParams) {
    return representation_1.Representation.createMulti('Interactions', ctx, getParams, representation_2.StructureRepresentationStateBuilder, InteractionsVisuals);
}
exports.InteractionRepresentation = InteractionRepresentation;
exports.InteractionsRepresentationProvider = (0, representation_2.StructureRepresentationProvider)({
    name: 'interactions',
    label: 'Non-covalent Interactions',
    description: 'Displays non-covalent interactions as dashed cylinders.',
    factory: InteractionRepresentation,
    getParams: getInteractionParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.InteractionsParams),
    defaultColorTheme: { name: 'interaction-type' },
    defaultSizeTheme: { name: 'uniform' },
    isApplicable: function (structure) { return structure.elementCount > 0 && interactions_1.InteractionsProvider.isApplicable(structure); },
    ensureCustomProperties: {
        attach: function (ctx, structure) { return interactions_1.InteractionsProvider.attach(ctx, structure, void 0, true); },
        detach: function (data) { return interactions_1.InteractionsProvider.ref(data, false); }
    },
    getData: function (structure, props) {
        return props.includeParent ? structure.asParent() : structure;
    },
    mustRecreate: function (oldProps, newProps) {
        return oldProps.includeParent !== newProps.includeParent;
    }
});
