// molstar/lib/commonjs/mol-plugin-state/builder/data.js
"use strict";
/**
 * Copyright (c) 2019 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataBuilder = void 0;
var tslib_1 = require("tslib");
var data_1 = require("../transforms/data");
var file_info_1 = require("../../mol-util/file-info");
var DataBuilder = /** @class */ (function () {
    function DataBuilder(plugin) {
        this.plugin = plugin;
    }
    Object.defineProperty(DataBuilder.prototype, "dataState", {
        get: function () {
            return this.plugin.state.data;
        },
        enumerable: false,
        configurable: true
    });
    DataBuilder.prototype.rawData = function (params, options) {
        var data = this.dataState.build().toRoot().apply(data_1.RawData, params, options);
        return data.commit({ revertOnError: true });
    };
    DataBuilder.prototype.download = function (params, options) {
        var data = this.dataState.build().toRoot().apply(data_1.Download, params, options);
        return data.commit({ revertOnError: true });
    };
    DataBuilder.prototype.downloadBlob = function (params, options) {
        var data = this.dataState.build().toRoot().apply(data_1.DownloadBlob, params, options);
        return data.commit({ revertOnError: true });
    };
    DataBuilder.prototype.readFile = function (params, options) {
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var data, fileInfo;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.dataState.build().toRoot().apply(data_1.ReadFile, params, options).commit({ revertOnError: true })];
                    case 1:
                        data = _b.sent();
                        fileInfo = (0, file_info_1.getFileInfo)(((_a = params.file) === null || _a === void 0 ? void 0 : _a.file) || '');
                        return [2 /*return*/, { data: data, fileInfo: fileInfo }];
                }
            });
        });
    };
    return DataBuilder;
}());
exports.DataBuilder = DataBuilder;
