// molstar/lib/extensions/rcsb/assembly-symmetry/behavior.js
/**
 * Copyright (c) 2018-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
import { __assign, __awaiter, __extends, __generator } from "tslib";
import { ParamDefinition as PD } from '../../../mol-util/param-definition';
import { AssemblySymmetryProvider, AssemblySymmetry, AssemblySymmetryDataProvider } from './prop';
import { PluginBehavior } from '../../../mol-plugin/behavior/behavior';
import { AssemblySymmetryParams, AssemblySymmetryRepresentation } from './representation';
import { AssemblySymmetryClusterColorThemeProvider } from './color';
import { PluginStateTransform, PluginStateObject } from '../../../mol-plugin-state/objects';
import { Task } from '../../../mol-task';
import { StateTransformer, StateAction, StateObject, StateObjectRef } from '../../../mol-state';
import { AssemblySymmetryControls } from './ui';
import { StructureRepresentationPresetProvider, PresetStructureRepresentations } from '../../../mol-plugin-state/builder/structure/representation-preset';
var Tag = AssemblySymmetry.Tag;
export var RCSBAssemblySymmetry = PluginBehavior.create({
    name: 'rcsb-assembly-symmetry-prop',
    category: 'custom-props',
    display: {
        name: 'Assembly Symmetry',
        description: 'Assembly Symmetry data calculated with BioJava, obtained via RCSB PDB.'
    },
    ctor: /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.provider = AssemblySymmetryProvider;
            return _this;
        }
        class_1.prototype.register = function () {
            this.ctx.state.data.actions.add(InitAssemblySymmetry3D);
            this.ctx.customStructureProperties.register(this.provider, this.params.autoAttach);
            this.ctx.representation.structure.themes.colorThemeRegistry.add(AssemblySymmetryClusterColorThemeProvider);
            this.ctx.genericRepresentationControls.set(Tag.Representation, function (selection) {
                var refs = [];
                selection.structures.forEach(function (structure) {
                    var _a;
                    var symmRepr = (_a = structure.genericRepresentations) === null || _a === void 0 ? void 0 : _a.filter(function (r) { return r.cell.transform.transformer.id === AssemblySymmetry3D.id; })[0];
                    if (symmRepr)
                        refs.push(symmRepr);
                });
                return [refs, 'Symmetries'];
            });
            this.ctx.customStructureControls.set(Tag.Representation, AssemblySymmetryControls);
            this.ctx.builders.structure.representation.registerPreset(AssemblySymmetryPreset);
        };
        class_1.prototype.update = function (p) {
            var updated = this.params.autoAttach !== p.autoAttach;
            this.params.autoAttach = p.autoAttach;
            this.ctx.customStructureProperties.setDefaultAutoAttach(this.provider.descriptor.name, this.params.autoAttach);
            return updated;
        };
        class_1.prototype.unregister = function () {
            this.ctx.state.data.actions.remove(InitAssemblySymmetry3D);
            this.ctx.customStructureProperties.unregister(this.provider.descriptor.name);
            this.ctx.representation.structure.themes.colorThemeRegistry.remove(AssemblySymmetryClusterColorThemeProvider);
            this.ctx.genericRepresentationControls.delete(Tag.Representation);
            this.ctx.customStructureControls.delete(Tag.Representation);
            this.ctx.builders.structure.representation.unregisterPreset(AssemblySymmetryPreset);
        };
        return class_1;
    }(PluginBehavior.Handler)),
    params: function () { return ({
        autoAttach: PD.Boolean(false),
        serverUrl: PD.Text(AssemblySymmetry.DefaultServerUrl)
    }); }
});
//
export var InitAssemblySymmetry3D = StateAction.build({
    display: {
        name: 'Assembly Symmetry',
        description: 'Initialize Assembly Symmetry axes and cage. Data calculated with BioJava, obtained via RCSB PDB.'
    },
    from: PluginStateObject.Molecule.Structure,
    isApplicable: function (a) { return AssemblySymmetry.isApplicable(a.data); }
})(function (_a, plugin) {
    var a = _a.a, ref = _a.ref, state = _a.state;
    return Task.create('Init Assembly Symmetry', function (ctx) { return __awaiter(void 0, void 0, void 0, function () {
        var propCtx, assemblySymmetryData, symmetryIndex, e_1, tree;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 3, , 4]);
                    propCtx = { runtime: ctx, assetManager: plugin.managers.asset };
                    return [4 /*yield*/, AssemblySymmetryDataProvider.attach(propCtx, a.data)];
                case 1:
                    _a.sent();
                    assemblySymmetryData = AssemblySymmetryDataProvider.get(a.data).value;
                    symmetryIndex = assemblySymmetryData ? AssemblySymmetry.firstNonC1(assemblySymmetryData) : -1;
                    return [4 /*yield*/, AssemblySymmetryProvider.attach(propCtx, a.data, { symmetryIndex: symmetryIndex })];
                case 2:
                    _a.sent();
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _a.sent();
                    plugin.log.error("Assembly Symmetry: ".concat(e_1));
                    return [2 /*return*/];
                case 4:
                    tree = state.build().to(ref)
                        .applyOrUpdateTagged(AssemblySymmetry.Tag.Representation, AssemblySymmetry3D);
                    return [4 /*yield*/, state.updateTree(tree).runInContext(ctx)];
                case 5:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    }); });
});
export { AssemblySymmetry3D };
var AssemblySymmetry3D = PluginStateTransform.BuiltIn({
    name: Tag.Representation,
    display: {
        name: 'Assembly Symmetry',
        description: 'Assembly Symmetry axes and cage. Data calculated with BioJava, obtained via RCSB PDB.'
    },
    from: PluginStateObject.Molecule.Structure,
    to: PluginStateObject.Shape.Representation3D,
    params: function (a) {
        return __assign({}, AssemblySymmetryParams);
    }
})({
    canAutoUpdate: function (_a) {
        var oldParams = _a.oldParams, newParams = _a.newParams;
        return true;
    },
    apply: function (_a, plugin) {
        var _this = this;
        var a = _a.a, params = _a.params;
        return Task.create('Assembly Symmetry', function (ctx) { return __awaiter(_this, void 0, void 0, function () {
            var assemblySymmetry, repr, type, kind, symbol;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, AssemblySymmetryProvider.attach({ runtime: ctx, assetManager: plugin.managers.asset }, a.data)];
                    case 1:
                        _b.sent();
                        assemblySymmetry = AssemblySymmetryProvider.get(a.data).value;
                        if (!assemblySymmetry || assemblySymmetry.symbol === 'C1') {
                            return [2 /*return*/, StateObject.Null];
                        }
                        repr = AssemblySymmetryRepresentation(__assign({ webgl: (_a = plugin.canvas3d) === null || _a === void 0 ? void 0 : _a.webgl }, plugin.representation.structure.themes), function () { return AssemblySymmetryParams; });
                        return [4 /*yield*/, repr.createOrUpdate(params, a.data).runInContext(ctx)];
                    case 2:
                        _b.sent();
                        type = assemblySymmetry.type, kind = assemblySymmetry.kind, symbol = assemblySymmetry.symbol;
                        return [2 /*return*/, new PluginStateObject.Shape.Representation3D({ repr: repr, sourceData: a.data }, { label: kind, description: "".concat(type, " (").concat(symbol, ")") })];
                }
            });
        }); });
    },
    update: function (_a, plugin) {
        var _this = this;
        var a = _a.a, b = _a.b, newParams = _a.newParams;
        return Task.create('Assembly Symmetry', function (ctx) { return __awaiter(_this, void 0, void 0, function () {
            var assemblySymmetry, props, type, kind, symbol;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, AssemblySymmetryProvider.attach({ runtime: ctx, assetManager: plugin.managers.asset }, a.data)];
                    case 1:
                        _a.sent();
                        assemblySymmetry = AssemblySymmetryProvider.get(a.data).value;
                        if (!assemblySymmetry || assemblySymmetry.symbol === 'C1') {
                            // this should NOT be StateTransformer.UpdateResult.Null
                            // because that keeps the old object
                            return [2 /*return*/, StateTransformer.UpdateResult.Recreate];
                        }
                        props = __assign(__assign({}, b.data.repr.props), newParams);
                        return [4 /*yield*/, b.data.repr.createOrUpdate(props, a.data).runInContext(ctx)];
                    case 2:
                        _a.sent();
                        b.data.sourceData = a.data;
                        type = assemblySymmetry.type, kind = assemblySymmetry.kind, symbol = assemblySymmetry.symbol;
                        b.label = kind;
                        b.description = "".concat(type, " (").concat(symbol, ")");
                        return [2 /*return*/, StateTransformer.UpdateResult.Updated];
                }
            });
        }); });
    },
    isApplicable: function (a) {
        return AssemblySymmetry.isApplicable(a.data);
    }
});
//
export var AssemblySymmetryPresetParams = __assign({}, StructureRepresentationPresetProvider.CommonParams);
export var AssemblySymmetryPreset = StructureRepresentationPresetProvider({
    id: 'preset-structure-representation-rcsb-assembly-symmetry',
    display: {
        name: 'Assembly Symmetry', group: 'Annotation',
        description: 'Shows Assembly Symmetry axes and cage; colors structure according to assembly symmetry cluster membership. Data calculated with BioJava, obtained via RCSB PDB.'
    },
    isApplicable: function (a) {
        return AssemblySymmetry.isApplicable(a.data);
    },
    params: function () { return AssemblySymmetryPresetParams; },
    apply: function (ref, params, plugin) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var structureCell, structure, assemblySymmetry, colorTheme, preset;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        structureCell = StateObjectRef.resolveAndCheck(plugin.state.data, ref);
                        structure = (_a = structureCell === null || structureCell === void 0 ? void 0 : structureCell.obj) === null || _a === void 0 ? void 0 : _a.data;
                        if (!structureCell || !structure)
                            return [2 /*return*/, {}];
                        if (!!AssemblySymmetryDataProvider.get(structure).value) return [3 /*break*/, 2];
                        return [4 /*yield*/, plugin.runTask(Task.create('Assembly Symmetry', function (runtime) { return __awaiter(_this, void 0, void 0, function () {
                                var propCtx, assemblySymmetryData, symmetryIndex;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            propCtx = { runtime: runtime, assetManager: plugin.managers.asset };
                                            return [4 /*yield*/, AssemblySymmetryDataProvider.attach(propCtx, structure)];
                                        case 1:
                                            _a.sent();
                                            assemblySymmetryData = AssemblySymmetryDataProvider.get(structure).value;
                                            symmetryIndex = assemblySymmetryData ? AssemblySymmetry.firstNonC1(assemblySymmetryData) : -1;
                                            return [4 /*yield*/, AssemblySymmetryProvider.attach(propCtx, structure, { symmetryIndex: symmetryIndex })];
                                        case 2:
                                            _a.sent();
                                            return [2 /*return*/];
                                    }
                                });
                            }); }))];
                    case 1:
                        _b.sent();
                        _b.label = 2;
                    case 2: return [4 /*yield*/, tryCreateAssemblySymmetry(plugin, structureCell)];
                    case 3:
                        assemblySymmetry = _b.sent();
                        colorTheme = assemblySymmetry.isOk ? Tag.Cluster : undefined;
                        return [4 /*yield*/, PresetStructureRepresentations.auto.apply(ref, __assign(__assign({}, params), { theme: { globalName: colorTheme, focus: { name: colorTheme } } }), plugin)];
                    case 4:
                        preset = _b.sent();
                        return [2 /*return*/, { components: preset.components, representations: __assign(__assign({}, preset.representations), { assemblySymmetry: assemblySymmetry }) }];
                }
            });
        });
    }
});
export function tryCreateAssemblySymmetry(plugin, structure, params, initialState) {
    var state = plugin.state.data;
    var assemblySymmetry = state.build().to(structure)
        .applyOrUpdateTagged(AssemblySymmetry.Tag.Representation, AssemblySymmetry3D, params, { state: initialState });
    return assemblySymmetry.commit({ revertOnError: true });
}
