// molstar/lib/extensions/rcsb/assembly-symmetry/color.js
/**
 * Copyright (c) 2018-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
import { __assign, __spreadArray } from "tslib";
import { ParamDefinition as PD } from '../../../mol-util/param-definition';
import { AssemblySymmetryProvider, AssemblySymmetry } from './prop';
import { Color } from '../../../mol-util/color';
import { StructureElement, StructureProperties, Bond } from '../../../mol-model/structure';
import { getPalette, getPaletteParams } from '../../../mol-util/color/palette';
var DefaultColor = Color(0xCCCCCC);
function getAsymId(unit) {
    switch (unit.kind) {
        case 0 /* Unit.Kind.Atomic */:
            return StructureProperties.chain.label_asym_id;
        case 1 /* Unit.Kind.Spheres */:
        case 2 /* Unit.Kind.Gaussians */:
            return StructureProperties.coarse.asym_id;
    }
}
function clusterMemberKey(asymId, operList) {
    return "".concat(asymId, "-").concat(operList.join('|'));
}
export var AssemblySymmetryClusterColorThemeParams = __assign({}, getPaletteParams({ colorList: 'red-yellow-blue' }));
export function getAssemblySymmetryClusterColorThemeParams(ctx) {
    var params = PD.clone(AssemblySymmetryClusterColorThemeParams);
    return params;
}
export function AssemblySymmetryClusterColorTheme(ctx, props) {
    var _a;
    var color = function () { return DefaultColor; };
    var legend;
    var assemblySymmetry = ctx.structure && AssemblySymmetryProvider.get(ctx.structure);
    var contextHash = assemblySymmetry === null || assemblySymmetry === void 0 ? void 0 : assemblySymmetry.version;
    var clusters = (_a = assemblySymmetry === null || assemblySymmetry === void 0 ? void 0 : assemblySymmetry.value) === null || _a === void 0 ? void 0 : _a.clusters;
    if ((clusters === null || clusters === void 0 ? void 0 : clusters.length) && ctx.structure) {
        var l_1 = StructureElement.Location.create(ctx.structure);
        var clusterByMember_1 = new Map();
        for (var i = 0, il = clusters.length; i < il; ++i) {
            var members = clusters[i].members;
            for (var j = 0, jl = members.length; j < jl; ++j) {
                var asymId = members[j].asym_id;
                var operList = __spreadArray([], members[j].pdbx_struct_oper_list_ids || [], true);
                clusterByMember_1.set(clusterMemberKey(asymId, operList), i);
                if (operList.length === 0) {
                    operList.push('1'); // TODO hack assuming '1' is the id of the identity operator
                    clusterByMember_1.set(clusterMemberKey(asymId, operList), i);
                }
            }
        }
        var palette_1 = getPalette(clusters.length, props);
        legend = palette_1.legend;
        var _emptyList_1 = [];
        var getColor_1 = function (location) {
            var assembly = location.unit.conformation.operator.assembly;
            var asymId = getAsymId(location.unit)(location);
            var cluster = clusterByMember_1.get(clusterMemberKey(asymId, (assembly === null || assembly === void 0 ? void 0 : assembly.operList) || _emptyList_1));
            return cluster !== undefined ? palette_1.color(cluster) : DefaultColor;
        };
        color = function (location) {
            if (StructureElement.Location.is(location)) {
                return getColor_1(location);
            }
            else if (Bond.isLocation(location)) {
                l_1.unit = location.aUnit;
                l_1.element = location.aUnit.elements[location.aIndex];
                return getColor_1(l_1);
            }
            return DefaultColor;
        };
    }
    return {
        factory: AssemblySymmetryClusterColorTheme,
        granularity: 'instance',
        color: color,
        props: props,
        contextHash: contextHash,
        description: 'Assigns chain colors according to assembly symmetry cluster membership calculated with BioJava and obtained via RCSB PDB.',
        legend: legend
    };
}
export var AssemblySymmetryClusterColorThemeProvider = {
    name: AssemblySymmetry.Tag.Cluster,
    label: 'Assembly Symmetry Cluster',
    category: "Symmetry" /* ColorTheme.Category.Symmetry */,
    factory: AssemblySymmetryClusterColorTheme,
    getParams: getAssemblySymmetryClusterColorThemeParams,
    defaultValues: PD.getDefaultValues(AssemblySymmetryClusterColorThemeParams),
    isApplicable: function (ctx) { return AssemblySymmetry.isApplicable(ctx.structure); },
    ensureCustomProperties: {
        attach: function (ctx, data) { return data.structure ? AssemblySymmetryProvider.attach(ctx, data.structure, void 0, true) : Promise.resolve(); },
        detach: function (data) { return data.structure && AssemblySymmetryProvider.ref(data.structure, false); }
    }
};
