// molstar/lib/commonjs/mol-model-formats/structure/basic/coarse.js
"use strict";
/**
 * Copyright (c) 2018-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmptyCoarse = void 0;
exports.getCoarse = getCoarse;
const coarse_1 = require("../../../mol-model/structure/model/properties/coarse");
const coarse_keys_1 = require("../../../mol-model/structure/model/properties/utils/coarse-keys");
const mol_util_1 = require("../../../mol-util");
const int_1 = require("../../../mol-data/int");
const linear_algebra_1 = require("../../../mol-math/linear-algebra");
const coarse_ranges_1 = require("../../../mol-model/structure/model/properties/utils/coarse-ranges");
const schema_1 = require("./schema");
exports.EmptyCoarse = { hierarchy: coarse_1.CoarseHierarchy.Empty, conformation: void 0 };
function getCoarse(data, chemicalComponentMap) {
    const { ihm_sphere_obj_site, ihm_gaussian_obj_site } = data;
    if (ihm_sphere_obj_site._rowCount === 0 && ihm_gaussian_obj_site._rowCount === 0)
        return exports.EmptyCoarse;
    const sphereData = getData(ihm_sphere_obj_site);
    const sphereConformation = getSphereConformation(ihm_sphere_obj_site);
    const sphereKeys = (0, coarse_keys_1.getCoarseKeys)(sphereData, data.entities);
    const sphereRanges = (0, coarse_ranges_1.getCoarseRanges)(sphereData, chemicalComponentMap);
    const gaussianData = getData(ihm_gaussian_obj_site);
    const gaussianConformation = getGaussianConformation(ihm_gaussian_obj_site);
    const gaussianKeys = (0, coarse_keys_1.getCoarseKeys)(gaussianData, data.entities);
    const gaussianRanges = (0, coarse_ranges_1.getCoarseRanges)(gaussianData, chemicalComponentMap);
    return {
        hierarchy: {
            isDefined: true,
            spheres: { ...sphereData, ...sphereKeys, ...sphereRanges },
            gaussians: { ...gaussianData, ...gaussianKeys, ...gaussianRanges },
        },
        conformation: {
            id: mol_util_1.UUID.create22(),
            spheres: sphereConformation,
            gaussians: gaussianConformation
        }
    };
}
function getSphereConformation(data) {
    return {
        x: data.Cartn_x.toArray({ array: Float32Array }),
        y: data.Cartn_y.toArray({ array: Float32Array }),
        z: data.Cartn_z.toArray({ array: Float32Array }),
        radius: data.object_radius.toArray({ array: Float32Array }),
        rmsf: data.rmsf.toArray({ array: Float32Array })
    };
}
function getGaussianConformation(data) {
    const matrix_space = schema_1.BasicSchema.ihm_gaussian_obj_site.covariance_matrix.space;
    const covariance_matrix = [];
    const { covariance_matrix: cm } = data;
    for (let i = 0, _i = cm.rowCount; i < _i; i++) {
        covariance_matrix[i] = linear_algebra_1.Tensor.toMat3((0, linear_algebra_1.Mat3)(), matrix_space, cm.value(i));
    }
    return {
        x: data.mean_Cartn_x.toArray({ array: Float32Array }),
        y: data.mean_Cartn_y.toArray({ array: Float32Array }),
        z: data.mean_Cartn_z.toArray({ array: Float32Array }),
        weight: data.weight.toArray({ array: Float32Array }),
        covariance_matrix
    };
}
function getSegments(asym_id, seq_id_begin, seq_id_end) {
    const chainOffsets = [0];
    for (let i = 1, _i = asym_id.rowCount; i < _i; i++) {
        const newChain = !asym_id.areValuesEqual(i - 1, i);
        if (newChain)
            chainOffsets[chainOffsets.length] = i;
    }
    return {
        chainElementSegments: int_1.Segmentation.ofOffsets(chainOffsets, int_1.Interval.ofBounds(0, asym_id.rowCount))
    };
}
function getData(data) {
    const { entity_id, seq_id_begin, seq_id_end, asym_id } = data;
    return { count: entity_id.rowCount, entity_id, asym_id, seq_id_begin, seq_id_end, ...getSegments(asym_id, seq_id_begin, seq_id_end) };
}
