// molstar/lib/commonjs/extensions/cellpack/property.js
"use strict";
/**
 * Copyright (c) 2019-2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CellPackInfoProvider = void 0;
var tslib_1 = require("tslib");
var custom_structure_property_1 = require("../../mol-model-props/common/custom-structure-property");
var param_definition_1 = require("../../mol-util/param-definition");
var custom_property_1 = require("../../mol-model/custom-property");
var CellPackInfoParams = {
    info: param_definition_1.ParamDefinition.Value({ packingsCount: 1, packingIndex: 0, colors: undefined }, { isHidden: true })
};
exports.CellPackInfoProvider = custom_structure_property_1.CustomStructureProperty.createProvider({
    label: 'CellPack Info',
    descriptor: (0, custom_property_1.CustomPropertyDescriptor)({ name: 'cellpack-info' }),
    type: 'root',
    defaultParams: CellPackInfoParams,
    getParams: function (data) { return CellPackInfoParams; },
    isApplicable: function (data) { return true; },
    obtain: function (ctx, data, props) { return (0, tslib_1.__awaiter)(void 0, void 0, void 0, function () {
        return (0, tslib_1.__generator)(this, function (_a) {
            return [2 /*return*/, {
                    value: (0, tslib_1.__assign)((0, tslib_1.__assign)({}, CellPackInfoParams.info.defaultValue), props.info)
                }];
        });
    }); }
});
