// molstar/lib/commonjs/mol-script/language/symbol-table/internal.js
"use strict";
/**
 * Copyright (c) 2019 Mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.internal = void 0;
var tslib_1 = require("tslib");
var type_1 = require("../type");
var Struct = (0, tslib_1.__importStar)(require("./structure-query"));
var symbol_1 = require("../symbol");
var helpers_1 = require("../helpers");
var generator = {
    '@header': 'Generators',
    bundleElement: (0, helpers_1.symbol)(symbol_1.Arguments.Dictionary({
        // TODO: should we use more universal unit keys? (i.e. based on chain and "operator name")
        groupedUnits: (0, symbol_1.Argument)(type_1.Type.Any),
        set: (0, symbol_1.Argument)(type_1.Type.Any),
        ranges: (0, symbol_1.Argument)(type_1.Type.Any) // SortedArray<UnitIndex>
    }), type_1.Type.Any),
    bundle: (0, helpers_1.symbol)(symbol_1.Arguments.Dictionary({
        elements: (0, symbol_1.Argument)(type_1.Type.Any) // BundleElement[]
    }), Struct.Types.ElementSelectionQuery, 'A selection with single structure containing represented by the bundle.'),
    // Use with caution as this is not "state saveable"
    // This query should never be used in any State Transform!
    current: (0, helpers_1.symbol)(symbol_1.Arguments.None, Struct.Types.ElementSelectionQuery, 'Current selection provided by the query context. Avoid using this in State Transforms.')
};
exports.internal = {
    '@header': 'Internal Queries',
    generator: generator
};
