// molstar/lib/mol-plugin-ui/controls.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
/**
 * Copyright (c) 2018-2024 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
import * as React from 'react';
import Markdown from 'react-markdown';
import { UpdateTrajectory } from '../mol-plugin-state/actions/structure';
import { StateTransforms } from '../mol-plugin-state/transforms';
import { PluginCommands } from '../mol-plugin/commands';
import { PluginReactContext, PluginUIComponent } from './base';
import { IconButton } from './controls/common';
import { Icon, NavigateBeforeSvg, NavigateNextSvg, SkipPreviousSvg, StopSvg, PlayArrowSvg, SubscriptionsOutlinedSvg, BuildSvg } from './controls/icons';
import { AnimationControls } from './state/animation';
import { StructureComponentControls } from './structure/components';
import { StructureMeasurementsControls } from './structure/measurements';
import { StructureSelectionActionsControls } from './structure/selection';
import { StructureSourceControls } from './structure/source';
import { VolumeStreamingControls, VolumeSourceControls } from './structure/volume';
import { PluginConfig } from '../mol-plugin/config';
import { StructureSuperpositionControls } from './structure/superposition';
import { StructureQuickStylesControls } from './structure/quick-styles';
export class TrajectoryViewportControls extends PluginUIComponent {
    constructor() {
        super(...arguments);
        this.state = { show: false, label: '' };
        this.update = () => {
            const state = this.plugin.state.data;
            const models = state.selectQ(q => q.ofTransformer(StateTransforms.Model.ModelFromTrajectory));
            if (models.length === 0) {
                this.setState({ show: false });
                return;
            }
            let label = '', count = 0;
            const parents = new Set();
            for (const m of models) {
                if (!m.sourceRef)
                    continue;
                const parent = state.cells.get(m.sourceRef).obj;
                if (!parent)
                    continue;
                if (parent.data.frameCount > 1) {
                    if (parents.has(m.sourceRef)) {
                        // do not show the controls if there are 2 models of the same trajectory present
                        this.setState({ show: false });
                        return;
                    }
                    parents.add(m.sourceRef);
                    count++;
                    if (!label) {
                        const idx = m.transform.params.modelIndex;
                        label = `Model ${idx + 1} / ${parent.data.frameCount}`;
                    }
                }
            }
            if (count > 1)
                label = '';
            this.setState({ show: count > 0, label });
        };
        this.reset = () => PluginCommands.State.ApplyAction(this.plugin, {
            state: this.plugin.state.data,
            action: UpdateTrajectory.create({ action: 'reset' })
        });
        this.prev = () => PluginCommands.State.ApplyAction(this.plugin, {
            state: this.plugin.state.data,
            action: UpdateTrajectory.create({ action: 'advance', by: -1 })
        });
        this.next = () => PluginCommands.State.ApplyAction(this.plugin, {
            state: this.plugin.state.data,
            action: UpdateTrajectory.create({ action: 'advance', by: 1 })
        });
    }
    componentDidMount() {
        this.subscribe(this.plugin.state.data.events.changed, this.update);
        this.subscribe(this.plugin.behaviors.state.isAnimating, this.update);
    }
    render() {
        const isAnimating = this.plugin.behaviors.state.isAnimating.value;
        if (!this.state.show || (isAnimating && !this.state.label) || !this.plugin.config.get(PluginConfig.Viewport.ShowTrajectoryControls))
            return null;
        return _jsxs("div", { className: 'msp-traj-controls', children: [!isAnimating && _jsx(IconButton, { svg: SkipPreviousSvg, title: 'First Model', onClick: this.reset, disabled: isAnimating }), !isAnimating && _jsx(IconButton, { svg: NavigateBeforeSvg, title: 'Previous Model', onClick: this.prev, disabled: isAnimating }), !isAnimating && _jsx(IconButton, { svg: NavigateNextSvg, title: 'Next Model', onClick: this.next, disabled: isAnimating }), !!this.state.label && _jsx("span", { children: this.state.label })] });
    }
}
export class StateSnapshotViewportControls extends PluginUIComponent {
    constructor() {
        super(...arguments);
        this.state = { isBusy: false, show: true };
        this.keyUp = (e) => {
            if (!e.ctrlKey || this.state.isBusy || e.target !== document.body)
                return;
            const snapshots = this.plugin.managers.snapshot;
            if (e.keyCode === 37 || e.key === 'ArrowLeft') {
                if (snapshots.state.isPlaying)
                    snapshots.stop();
                this.prev();
            }
            else if (e.keyCode === 38 || e.key === 'ArrowUp') {
                if (snapshots.state.isPlaying)
                    snapshots.stop();
                if (snapshots.state.entries.size === 0)
                    return;
                const e = snapshots.state.entries.get(0);
                this.update(e.snapshot.id);
            }
            else if (e.keyCode === 39 || e.key === 'ArrowRight') {
                if (snapshots.state.isPlaying)
                    snapshots.stop();
                this.next();
            }
            else if (e.keyCode === 40 || e.key === 'ArrowDown') {
                if (snapshots.state.isPlaying)
                    snapshots.stop();
                if (snapshots.state.entries.size === 0)
                    return;
                const e = snapshots.state.entries.get(snapshots.state.entries.size - 1);
                this.update(e.snapshot.id);
            }
        };
        this.change = (e) => {
            if (e.target.value === 'none')
                return;
            this.update(e.target.value);
        };
        this.prev = () => {
            const s = this.plugin.managers.snapshot;
            const id = s.getNextId(s.state.current, -1);
            if (id)
                this.update(id);
        };
        this.next = () => {
            const s = this.plugin.managers.snapshot;
            const id = s.getNextId(s.state.current, 1);
            if (id)
                this.update(id);
        };
        this.togglePlay = () => {
            this.plugin.managers.snapshot.togglePlay();
        };
    }
    componentDidMount() {
        // TODO: this needs to be diabled when the state is updating!
        this.subscribe(this.plugin.managers.snapshot.events.changed, () => this.forceUpdate());
        this.subscribe(this.plugin.behaviors.state.isBusy, isBusy => this.setState({ isBusy }));
        this.subscribe(this.plugin.behaviors.state.isAnimating, isBusy => this.setState({ isBusy }));
        window.addEventListener('keyup', this.keyUp, false);
    }
    componentWillUnmount() {
        super.componentWillUnmount();
        window.removeEventListener('keyup', this.keyUp, false);
    }
    async update(id) {
        this.setState({ isBusy: true });
        await PluginCommands.State.Snapshots.Apply(this.plugin, { id });
        this.setState({ isBusy: false });
    }
    render() {
        const snapshots = this.plugin.managers.snapshot;
        const count = snapshots.state.entries.size;
        if (count < 2 || !this.state.show) {
            return null;
        }
        const current = snapshots.state.current;
        const isPlaying = snapshots.state.isPlaying;
        return _jsxs("div", { className: 'msp-state-snapshot-viewport-controls', children: [_jsxs("select", { className: 'msp-form-control', value: current || 'none', onChange: this.change, disabled: this.state.isBusy || isPlaying, children: [!current && _jsx("option", { value: 'none' }, 'none'), snapshots.state.entries.valueSeq().map((e, i) => _jsxs("option", { value: e.snapshot.id, children: [`[${i + 1}/${count}]`, " ", e.name || new Date(e.timestamp).toLocaleString()] }, e.snapshot.id))] }), _jsx(IconButton, { svg: isPlaying ? StopSvg : PlayArrowSvg, title: isPlaying ? 'Pause' : 'Cycle States', onClick: this.togglePlay, disabled: isPlaying ? false : this.state.isBusy }), !isPlaying && _jsxs(_Fragment, { children: [_jsx(IconButton, { svg: NavigateBeforeSvg, title: 'Previous State', onClick: this.prev, disabled: this.state.isBusy || isPlaying }), _jsx(IconButton, { svg: NavigateNextSvg, title: 'Next State', onClick: this.next, disabled: this.state.isBusy || isPlaying })] })] });
    }
}
export function ViewportSnapshotDescription() {
    var _a;
    const plugin = React.useContext(PluginReactContext);
    const [_, setV] = React.useState(0);
    React.useEffect(() => {
        const sub = plugin.managers.snapshot.events.changed.subscribe(() => setV(v => v + 1));
        return () => sub.unsubscribe();
    }, [plugin]);
    const current = plugin.managers.snapshot.state.current;
    if (!current)
        return null;
    const e = plugin.managers.snapshot.getEntry(current);
    if (!((_a = e === null || e === void 0 ? void 0 : e.description) === null || _a === void 0 ? void 0 : _a.trim()))
        return null;
    return _jsx("div", { id: 'snapinfo', className: 'msp-snapshot-description-wrapper', children: _jsx(Markdown, { skipHtml: true, components: { a: MarkdownAnchor }, children: e.description }) });
}
function MarkdownAnchor({ href, children, element }) {
    const plugin = React.useContext(PluginReactContext);
    if (!href)
        return element;
    if (href[0] === '#') {
        return _jsx("a", { href: '#', onClick: (e) => {
                e.preventDefault();
                plugin.managers.snapshot.applyKey(href.substring(1));
            }, children: children });
    }
    // TODO: consider adding more "commands", for example !reset-camera
    return element;
}
export class AnimationViewportControls extends PluginUIComponent {
    constructor() {
        super(...arguments);
        this.state = { isEmpty: true, isExpanded: false, isBusy: false, isAnimating: false, isPlaying: false };
        this.toggleExpanded = () => this.setState({ isExpanded: !this.state.isExpanded });
        this.stop = () => {
            this.plugin.managers.animation.stop();
            this.plugin.managers.snapshot.stop();
        };
    }
    componentDidMount() {
        this.subscribe(this.plugin.managers.snapshot.events.changed, () => {
            if (this.plugin.managers.snapshot.state.isPlaying)
                this.setState({ isPlaying: true, isExpanded: false });
            else
                this.setState({ isPlaying: false });
        });
        this.subscribe(this.plugin.behaviors.state.isBusy, isBusy => {
            if (isBusy)
                this.setState({ isBusy: true, isExpanded: false, isEmpty: this.plugin.state.data.tree.transforms.size < 2 });
            else
                this.setState({ isBusy: false, isEmpty: this.plugin.state.data.tree.transforms.size < 2 });
        });
        this.subscribe(this.plugin.behaviors.state.isAnimating, isAnimating => {
            if (isAnimating)
                this.setState({ isAnimating: true, isExpanded: false });
            else
                this.setState({ isAnimating: false });
        });
    }
    render() {
        const isPlaying = this.plugin.managers.snapshot.state.isPlaying;
        if (isPlaying || this.state.isEmpty || this.plugin.managers.animation.isEmpty || !this.plugin.config.get(PluginConfig.Viewport.ShowAnimation))
            return null;
        const isAnimating = this.state.isAnimating;
        return _jsxs("div", { className: 'msp-animation-viewport-controls', children: [_jsxs("div", { children: [_jsx("div", { className: 'msp-semi-transparent-background' }), _jsx(IconButton, { svg: isAnimating || isPlaying ? StopSvg : SubscriptionsOutlinedSvg, transparent: true, title: isAnimating ? 'Stop' : 'Select Animation', onClick: isAnimating || isPlaying ? this.stop : this.toggleExpanded, toggleState: this.state.isExpanded, disabled: isAnimating || isPlaying ? false : this.state.isBusy || this.state.isPlaying || this.state.isEmpty })] }), (this.state.isExpanded && !this.state.isBusy) && _jsx("div", { className: 'msp-animation-viewport-controls-select', children: _jsx(AnimationControls, { onStart: this.toggleExpanded }) })] });
    }
}
export class SelectionViewportControls extends PluginUIComponent {
    componentDidMount() {
        this.subscribe(this.plugin.behaviors.interaction.selectionMode, () => this.forceUpdate());
    }
    render() {
        if (!this.plugin.selectionMode)
            return null;
        return _jsx("div", { className: 'msp-selection-viewport-controls', children: _jsx(StructureSelectionActionsControls, {}) });
    }
}
export class LociLabels extends PluginUIComponent {
    constructor() {
        super(...arguments);
        this.state = { labels: [] };
    }
    componentDidMount() {
        this.subscribe(this.plugin.behaviors.labels.highlight, e => this.setState({ labels: e.labels }));
    }
    render() {
        if (this.state.labels.length === 0) {
            return null;
        }
        return _jsx("div", { className: 'msp-highlight-info', children: this.state.labels.map((e, i) => {
                if (e.indexOf('\n') > 0) {
                    return _jsx("div", { className: 'msp-highlight-markdown-row', children: _jsx(Markdown, { skipHtml: true, children: e }) }, '' + i);
                }
                return _jsx("div", { className: 'msp-highlight-simple-row', dangerouslySetInnerHTML: { __html: e } }, '' + i);
            }) });
    }
}
export class CustomStructureControls extends PluginUIComponent {
    componentDidMount() {
        this.subscribe(this.plugin.state.behaviors.events.changed, () => this.forceUpdate());
    }
    render() {
        const controls = [];
        this.plugin.customStructureControls.forEach((Controls, key) => {
            controls.push(_jsx(Controls, { initiallyCollapsed: this.props.initiallyCollapsed }, key));
        });
        return controls.length > 0 ? _jsx(_Fragment, { children: controls }) : null;
    }
}
export class DefaultStructureTools extends PluginUIComponent {
    render() {
        return _jsxs(_Fragment, { children: [_jsxs("div", { className: 'msp-section-header', children: [_jsx(Icon, { svg: BuildSvg }), "Structure Tools"] }), _jsx(StructureSourceControls, {}), _jsx(StructureMeasurementsControls, {}), _jsx(StructureSuperpositionControls, {}), _jsx(StructureQuickStylesControls, {}), _jsx(StructureComponentControls, {}), this.plugin.config.get(PluginConfig.VolumeStreaming.Enabled) && _jsx(VolumeStreamingControls, {}), _jsx(VolumeSourceControls, {}), _jsx(CustomStructureControls, {})] });
    }
}
