// molstar/lib/extensions/geo-export/controls.js
/**
 * Copyright (c) 2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Sukolsak Sakshuwong <sukolsak@stanford.edu>
 */
import { __awaiter, __extends, __generator } from "tslib";
import { Box3D } from '../../mol-math/geometry';
import { PluginComponent } from '../../mol-plugin-state/component';
import { Task } from '../../mol-task';
import { PluginStateObject } from '../../mol-plugin-state/objects';
import { StateSelection } from '../../mol-state';
import { ParamDefinition as PD } from '../../mol-util/param-definition';
import { SetUtils } from '../../mol-util/set';
import { GlbExporter } from './glb-exporter';
import { ObjExporter } from './obj-exporter';
import { StlExporter } from './stl-exporter';
import { UsdzExporter } from './usdz-exporter';
export var GeometryParams = {
    format: PD.Select('glb', [
        ['glb', 'glTF 2.0 Binary (.glb)'],
        ['stl', 'Stl (.stl)'],
        ['obj', 'Wavefront (.obj)'],
        ['usdz', 'Universal Scene Description (.usdz)']
    ])
};
var GeometryControls = /** @class */ (function (_super) {
    __extends(GeometryControls, _super);
    function GeometryControls(plugin) {
        var _this = _super.call(this) || this;
        _this.plugin = plugin;
        _this.behaviors = {
            params: _this.ev.behavior(PD.getDefaultValues(GeometryParams))
        };
        return _this;
    }
    GeometryControls.prototype.getFilename = function () {
        var models = this.plugin.state.data.select(StateSelection.Generators.rootsOfType(PluginStateObject.Molecule.Model)).map(function (s) { return s.obj.data; });
        var uniqueIds = new Set();
        models.forEach(function (m) { return uniqueIds.add(m.entryId.toUpperCase()); });
        var idString = SetUtils.toArray(uniqueIds).join('-');
        return "".concat(idString || 'molstar-model');
    };
    GeometryControls.prototype.exportGeometry = function () {
        var _this = this;
        var task = Task.create('Export Geometry', function (ctx) { return __awaiter(_this, void 0, void 0, function () {
            var renderObjects, filename, boundingSphere, boundingBox, renderObjectExporter, i, il, blob, e_1;
            var _a, _b, _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        _d.trys.push([0, 7, , 8]);
                        renderObjects = (_a = this.plugin.canvas3d) === null || _a === void 0 ? void 0 : _a.getRenderObjects();
                        filename = this.getFilename();
                        boundingSphere = (_b = this.plugin.canvas3d) === null || _b === void 0 ? void 0 : _b.boundingSphereVisible;
                        boundingBox = Box3D.fromSphere3D(Box3D(), boundingSphere);
                        renderObjectExporter = void 0;
                        switch (this.behaviors.params.value.format) {
                            case 'glb':
                                renderObjectExporter = new GlbExporter(boundingBox);
                                break;
                            case 'obj':
                                renderObjectExporter = new ObjExporter(filename, boundingBox);
                                break;
                            case 'stl':
                                renderObjectExporter = new StlExporter(boundingBox);
                                break;
                            case 'usdz':
                                renderObjectExporter = new UsdzExporter(boundingBox, boundingSphere.radius);
                                break;
                            default: throw new Error('Unsupported format.');
                        }
                        i = 0, il = renderObjects.length;
                        _d.label = 1;
                    case 1:
                        if (!(i < il)) return [3 /*break*/, 5];
                        return [4 /*yield*/, ctx.update({ message: "Exporting object ".concat(i, "/").concat(il) })];
                    case 2:
                        _d.sent();
                        return [4 /*yield*/, renderObjectExporter.add(renderObjects[i], (_c = this.plugin.canvas3d) === null || _c === void 0 ? void 0 : _c.webgl, ctx)];
                    case 3:
                        _d.sent();
                        _d.label = 4;
                    case 4:
                        ++i;
                        return [3 /*break*/, 1];
                    case 5: return [4 /*yield*/, renderObjectExporter.getBlob(ctx)];
                    case 6:
                        blob = _d.sent();
                        return [2 /*return*/, {
                                blob: blob,
                                filename: filename + '.' + renderObjectExporter.fileExtension
                            }];
                    case 7:
                        e_1 = _d.sent();
                        this.plugin.log.error('Error during geometry export');
                        throw e_1;
                    case 8: return [2 /*return*/];
                }
            });
        }); });
        return this.plugin.runTask(task, { useOverlay: true });
    };
    return GeometryControls;
}(PluginComponent));
export { GeometryControls };
