// molstar/lib/extensions/geo-export/index.js
/**
 * Copyright (c) 2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Sukolsak Sakshuwong <sukolsak@stanford.edu>
 */
import { __extends } from "tslib";
import { PluginBehavior } from '../../mol-plugin/behavior/behavior';
import { GeometryExporterUI } from './ui';
export var GeometryExport = PluginBehavior.create({
    name: 'extension-geo-export',
    category: 'misc',
    display: {
        name: 'Geometry Export'
    },
    ctor: /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.register = function () {
            this.ctx.customStructureControls.set('geo-export', GeometryExporterUI);
        };
        class_1.prototype.update = function () {
            return false;
        };
        class_1.prototype.unregister = function () {
            this.ctx.customStructureControls.delete('geo-export');
        };
        return class_1;
    }(PluginBehavior.Handler)),
    params: function () { return ({}); }
});
