// molstar/lib/commonjs/mol-repr/structure/representation/gaussian-volume.js
"use strict";
/**
 * Copyright (c) 2018-2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GaussianVolumeRepresentationProvider = exports.GaussianVolumeRepresentation = exports.getGaussianVolumeParams = exports.GaussianVolumeParams = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../../mol-util/param-definition");
var gaussian_density_volume_1 = require("../visual/gaussian-density-volume");
var representation_1 = require("../representation");
var representation_2 = require("../../../mol-repr/representation");
var GaussianVolumeVisuals = {
    'gaussian-volume': function (ctx, getParams) { return (0, representation_1.ComplexRepresentation)('Gaussian volume', ctx, getParams, gaussian_density_volume_1.GaussianDensityVolumeVisual); },
    'units-gaussian-volume': function (ctx, getParams) { return (0, representation_1.UnitsRepresentation)('Units-Gaussian volume', ctx, getParams, gaussian_density_volume_1.UnitsGaussianDensityVolumeVisual); }
};
exports.GaussianVolumeParams = (0, tslib_1.__assign)((0, tslib_1.__assign)({}, gaussian_density_volume_1.GaussianDensityVolumeParams), { visuals: param_definition_1.ParamDefinition.MultiSelect(['gaussian-volume'], param_definition_1.ParamDefinition.objectToOptions(GaussianVolumeVisuals)) });
function getGaussianVolumeParams(ctx, structure) {
    var p = param_definition_1.ParamDefinition.clone(exports.GaussianVolumeParams);
    p.jumpLength = param_definition_1.ParamDefinition.Numeric(4, { min: 0, max: 20, step: 0.1 });
    return p;
}
exports.getGaussianVolumeParams = getGaussianVolumeParams;
function GaussianVolumeRepresentation(ctx, getParams) {
    return representation_2.Representation.createMulti('Gaussian Volume', ctx, getParams, representation_1.StructureRepresentationStateBuilder, GaussianVolumeVisuals);
}
exports.GaussianVolumeRepresentation = GaussianVolumeRepresentation;
exports.GaussianVolumeRepresentationProvider = (0, representation_1.StructureRepresentationProvider)({
    name: 'gaussian-volume',
    label: 'Gaussian Volume',
    description: 'Displays a gaussian molecular density using direct volume rendering.',
    factory: GaussianVolumeRepresentation,
    getParams: getGaussianVolumeParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.GaussianVolumeParams),
    defaultColorTheme: { name: 'chain-id' },
    defaultSizeTheme: { name: 'physical' },
    isApplicable: function (structure) { return structure.elementCount > 0; }
});
