// molstar/lib/commonjs/mol-repr/structure/representation/molecular-surface.js
"use strict";
/**
 * Copyright (c) 2018-2019 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MolecularSurfaceRepresentationProvider = exports.MolecularSurfaceRepresentation = exports.getMolecularSurfaceParams = exports.MolecularSurfaceParams = void 0;
var tslib_1 = require("tslib");
var molecular_surface_mesh_1 = require("../visual/molecular-surface-mesh");
var units_representation_1 = require("../units-representation");
var param_definition_1 = require("../../../mol-util/param-definition");
var representation_1 = require("../representation");
var representation_2 = require("../../../mol-repr/representation");
var molecular_surface_wireframe_1 = require("../visual/molecular-surface-wireframe");
var base_1 = require("../../../mol-geo/geometry/base");
var MolecularSurfaceVisuals = {
    'molecular-surface-mesh': function (ctx, getParams) { return (0, units_representation_1.UnitsRepresentation)('Molecular surface mesh', ctx, getParams, molecular_surface_mesh_1.MolecularSurfaceMeshVisual); },
    'molecular-surface-wireframe': function (ctx, getParams) { return (0, units_representation_1.UnitsRepresentation)('Molecular surface wireframe', ctx, getParams, molecular_surface_wireframe_1.MolecularSurfaceWireframeVisual); },
};
exports.MolecularSurfaceParams = tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, molecular_surface_mesh_1.MolecularSurfaceMeshParams), molecular_surface_wireframe_1.MolecularSurfaceWireframeParams), { visuals: param_definition_1.ParamDefinition.MultiSelect(['molecular-surface-mesh'], param_definition_1.ParamDefinition.objectToOptions(MolecularSurfaceVisuals)), bumpFrequency: param_definition_1.ParamDefinition.Numeric(1, { min: 0, max: 10, step: 0.1 }, base_1.BaseGeometry.ShadingCategory) });
function getMolecularSurfaceParams(ctx, structure) {
    return param_definition_1.ParamDefinition.clone(exports.MolecularSurfaceParams);
}
exports.getMolecularSurfaceParams = getMolecularSurfaceParams;
function MolecularSurfaceRepresentation(ctx, getParams) {
    return representation_2.Representation.createMulti('Molecular Surface', ctx, getParams, representation_1.StructureRepresentationStateBuilder, MolecularSurfaceVisuals);
}
exports.MolecularSurfaceRepresentation = MolecularSurfaceRepresentation;
exports.MolecularSurfaceRepresentationProvider = (0, representation_1.StructureRepresentationProvider)({
    name: 'molecular-surface',
    label: 'Molecular Surface',
    description: 'Displays a molecular surface.',
    factory: MolecularSurfaceRepresentation,
    getParams: getMolecularSurfaceParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.MolecularSurfaceParams),
    defaultColorTheme: { name: 'chain-id' },
    defaultSizeTheme: { name: 'physical' },
    isApplicable: function (structure) { return structure.elementCount > 0; }
});
