// molstar/lib/commonjs/mol-repr/structure/representation/orientation.js
"use strict";
/**
 * Copyright (c) 2019 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrientationRepresentationProvider = exports.OrientationRepresentation = exports.getOrientationParams = exports.OrientationParams = void 0;
var tslib_1 = require("tslib");
var units_representation_1 = require("../units-representation");
var param_definition_1 = require("../../../mol-util/param-definition");
var representation_1 = require("../representation");
var representation_2 = require("../../../mol-repr/representation");
var orientation_ellipsoid_mesh_1 = require("../visual/orientation-ellipsoid-mesh");
var base_1 = require("../../../mol-geo/geometry/base");
var OrientationVisuals = {
    'orientation-ellipsoid-mesh': function (ctx, getParams) { return (0, units_representation_1.UnitsRepresentation)('Orientation ellipsoid mesh', ctx, getParams, orientation_ellipsoid_mesh_1.OrientationEllipsoidMeshVisual); },
};
exports.OrientationParams = (0, tslib_1.__assign)((0, tslib_1.__assign)({}, orientation_ellipsoid_mesh_1.OrientationEllipsoidMeshParams), { visuals: param_definition_1.ParamDefinition.MultiSelect(['orientation-ellipsoid-mesh'], param_definition_1.ParamDefinition.objectToOptions(OrientationVisuals)), bumpFrequency: param_definition_1.ParamDefinition.Numeric(1, { min: 0, max: 10, step: 0.1 }, base_1.BaseGeometry.ShadingCategory) });
function getOrientationParams(ctx, structure) {
    return param_definition_1.ParamDefinition.clone(exports.OrientationParams);
}
exports.getOrientationParams = getOrientationParams;
function OrientationRepresentation(ctx, getParams) {
    return representation_2.Representation.createMulti('Orientation', ctx, getParams, representation_1.StructureRepresentationStateBuilder, OrientationVisuals);
}
exports.OrientationRepresentation = OrientationRepresentation;
exports.OrientationRepresentationProvider = (0, representation_1.StructureRepresentationProvider)({
    name: 'orientation',
    label: 'Orientation',
    description: 'Displays orientation ellipsoids for polymer chains.',
    factory: OrientationRepresentation,
    getParams: getOrientationParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.OrientationParams),
    defaultColorTheme: { name: 'chain-id' },
    defaultSizeTheme: { name: 'uniform' },
    isApplicable: function (structure) { return structure.elementCount > 0; }
});
