// molstar/lib/commonjs/mol-data/int.js
"use strict";
/**
 * Copyright (c) 2017 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Iterator = exports.IntMap = exports.LinkedIndex = exports.Tuple = exports.SortedArray = exports.Segmentation = exports.OrderedSet = exports.Interval = void 0;
var interval_1 = require("./int/interval");
Object.defineProperty(exports, "Interval", { enumerable: true, get: function () { return interval_1.Interval; } });
var ordered_set_1 = require("./int/ordered-set");
Object.defineProperty(exports, "OrderedSet", { enumerable: true, get: function () { return ordered_set_1.OrderedSet; } });
var segmentation_1 = require("./int/segmentation");
Object.defineProperty(exports, "Segmentation", { enumerable: true, get: function () { return segmentation_1.Segmentation; } });
var sorted_array_1 = require("./int/sorted-array");
Object.defineProperty(exports, "SortedArray", { enumerable: true, get: function () { return sorted_array_1.SortedArray; } });
var tuple_1 = require("./int/tuple");
Object.defineProperty(exports, "Tuple", { enumerable: true, get: function () { return tuple_1.IntTuple; } });
var linked_index_1 = require("./int/linked-index");
Object.defineProperty(exports, "LinkedIndex", { enumerable: true, get: function () { return linked_index_1.LinkedIndex; } });
var map_1 = require("./int/map");
Object.defineProperty(exports, "IntMap", { enumerable: true, get: function () { return map_1.IntMap; } });
var iterator_1 = require("./iterator");
Object.defineProperty(exports, "Iterator", { enumerable: true, get: function () { return iterator_1.Iterator; } });
