// molstar/lib/commonjs/mol-model-formats/structure/gro.js
"use strict";
/**
 * Copyright (c) 2019-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.trajectoryFromGRO = exports.GroFormat = void 0;
var tslib_1 = require("tslib");
var mol_task_1 = require("../../mol-task");
var db_1 = require("../../mol-data/db");
var util_1 = require("./util");
var types_1 = require("../../mol-model/structure/model/types");
var component_1 = require("./common/component");
var util_2 = require("./common/util");
var entity_1 = require("./common/entity");
var schema_1 = require("./basic/schema");
var parser_1 = require("./basic/parser");
var trajectory_1 = require("../../mol-model/structure/trajectory");
function getBasic(atoms, modelNum) {
    var auth_atom_id = atoms.atomName;
    var auth_comp_id = atoms.residueName;
    var entityIds = new Array(atoms.count);
    var asymIds = new Array(atoms.count);
    var seqIds = new Uint32Array(atoms.count);
    var ids = new Uint32Array(atoms.count);
    var typeSymbol = new Array(atoms.count);
    var entityBuilder = new entity_1.EntityBuilder();
    var componentBuilder = new component_1.ComponentBuilder(atoms.residueNumber, atoms.atomName);
    var currentEntityId = '';
    var currentAsymIndex = 0;
    var currentAsymId = '';
    var currentSeqId = 0;
    var prevMoleculeType = 0 /* MoleculeType.Unknown */;
    var prevResidueNumber = -1;
    for (var i = 0, il = atoms.count; i < il; ++i) {
        var residueNumber = atoms.residueNumber.value(i);
        if (residueNumber !== prevResidueNumber) {
            var compId = atoms.residueName.value(i);
            var moleculeType = (0, types_1.getMoleculeType)(componentBuilder.add(compId, i).type, compId);
            if (moleculeType !== prevMoleculeType || (residueNumber !== prevResidueNumber + 1 && !(
            // gro format allows only for 5 character residueNumbers, handle overflow here
            prevResidueNumber === 99999 && residueNumber === 0))) {
                currentAsymId = (0, util_2.getChainId)(currentAsymIndex);
                currentAsymIndex += 1;
                currentSeqId = 0;
            }
            currentEntityId = entityBuilder.getEntityId(compId, moleculeType, currentAsymId);
            currentSeqId += 1;
            prevResidueNumber = residueNumber;
            prevMoleculeType = moleculeType;
        }
        entityIds[i] = currentEntityId;
        asymIds[i] = currentAsymId;
        seqIds[i] = currentSeqId;
        ids[i] = i;
        typeSymbol[i] = (0, util_1.guessElementSymbolString)(atoms.atomName.value(i), atoms.residueName.value(i));
    }
    var auth_asym_id = db_1.Column.ofStringArray(asymIds);
    var atom_site = db_1.Table.ofPartialColumns(schema_1.BasicSchema.atom_site, {
        auth_asym_id: auth_asym_id,
        auth_atom_id: auth_atom_id,
        auth_comp_id: auth_comp_id,
        auth_seq_id: atoms.residueNumber,
        Cartn_x: db_1.Column.ofFloatArray(db_1.Column.mapToArray(atoms.x, function (x) { return x * 10; }, Float32Array)),
        Cartn_y: db_1.Column.ofFloatArray(db_1.Column.mapToArray(atoms.y, function (y) { return y * 10; }, Float32Array)),
        Cartn_z: db_1.Column.ofFloatArray(db_1.Column.mapToArray(atoms.z, function (z) { return z * 10; }, Float32Array)),
        id: db_1.Column.ofIntArray(ids),
        label_asym_id: auth_asym_id,
        label_atom_id: auth_atom_id,
        label_comp_id: auth_comp_id,
        label_seq_id: db_1.Column.ofIntArray(seqIds),
        label_entity_id: db_1.Column.ofStringArray(entityIds),
        occupancy: db_1.Column.ofConst(1, atoms.count, db_1.Column.Schema.float),
        type_symbol: db_1.Column.ofStringArray(typeSymbol),
        pdbx_PDB_model_num: db_1.Column.ofConst(modelNum, atoms.count, db_1.Column.Schema.int),
    }, atoms.count);
    return (0, schema_1.createBasic)({
        entity: entityBuilder.getEntityTable(),
        chem_comp: componentBuilder.getChemCompTable(),
        atom_site: atom_site
    });
}
var GroFormat;
(function (GroFormat) {
    function is(x) {
        return (x === null || x === void 0 ? void 0 : x.kind) === 'gro';
    }
    GroFormat.is = is;
    function fromGro(gro) {
        return { kind: 'gro', name: gro.structures[0].header.title, data: gro };
    }
    GroFormat.fromGro = fromGro;
})(GroFormat || (GroFormat = {}));
exports.GroFormat = GroFormat;
// TODO reuse static model parts when hierarchy is identical
//      need to pass all gro.structures as one table into createModels
function trajectoryFromGRO(gro) {
    var _this = this;
    return mol_task_1.Task.create('Parse GRO', function (ctx) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
        var format, models, i, il, basic, m;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    format = GroFormat.fromGro(gro);
                    models = [];
                    i = 0, il = gro.structures.length;
                    _a.label = 1;
                case 1:
                    if (!(i < il)) return [3 /*break*/, 4];
                    basic = getBasic(gro.structures[i].atoms, i + 1);
                    return [4 /*yield*/, (0, parser_1.createModels)(basic, format, ctx)];
                case 2:
                    m = _a.sent();
                    if (m.frameCount === 1) {
                        models.push(m.representative);
                    }
                    _a.label = 3;
                case 3:
                    ++i;
                    return [3 /*break*/, 1];
                case 4: return [2 /*return*/, new trajectory_1.ArrayTrajectory(models)];
            }
        });
    }); });
}
exports.trajectoryFromGRO = trajectoryFromGRO;
