// molstar/lib/commonjs/mol-plugin-state/helpers/volume-representation-params.js
"use strict";
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createVolumeSizeThemeParams = exports.createVolumeColorThemeParams = exports.getVolumeThemeTypes = exports.createVolumeRepresentationParams = void 0;
var volume_1 = require("../../mol-model/volume");
var param_definition_1 = require("../../mol-util/param-definition");
function createVolumeRepresentationParams(ctx, volume, props) {
    if (props === void 0) { props = {}; }
    var p = props;
    if (typeof p.type === 'string' || typeof p.color === 'string' || typeof p.size === 'string')
        return createParamsByName(ctx, volume || volume_1.Volume.One, props);
    return createParamsProvider(ctx, volume || volume_1.Volume.One, props);
}
exports.createVolumeRepresentationParams = createVolumeRepresentationParams;
function getVolumeThemeTypes(ctx, volume) {
    var themeCtx = ctx.representation.volume.themes;
    if (!volume)
        return themeCtx.colorThemeRegistry.types;
    return themeCtx.colorThemeRegistry.getApplicableTypes({ volume: volume });
}
exports.getVolumeThemeTypes = getVolumeThemeTypes;
function createVolumeColorThemeParams(ctx, volume, typeName, themeName, params) {
    var _a = ctx.representation.volume, registry = _a.registry, themes = _a.themes;
    var repr = registry.get(typeName || registry.default.name);
    var color = themes.colorThemeRegistry.get(themeName || repr.defaultColorTheme.name);
    var colorDefaultParams = param_definition_1.ParamDefinition.getDefaultValues(color.getParams({ volume: volume || volume_1.Volume.One }));
    if (color.name === repr.defaultColorTheme.name)
        Object.assign(colorDefaultParams, repr.defaultColorTheme.props);
    return { name: color.name, params: Object.assign(colorDefaultParams, params) };
}
exports.createVolumeColorThemeParams = createVolumeColorThemeParams;
function createVolumeSizeThemeParams(ctx, volume, typeName, themeName, params) {
    var _a = ctx.representation.volume, registry = _a.registry, themes = _a.themes;
    var repr = registry.get(typeName || registry.default.name);
    var size = themes.sizeThemeRegistry.get(themeName || repr.defaultSizeTheme.name);
    var sizeDefaultParams = param_definition_1.ParamDefinition.getDefaultValues(size.getParams({ volume: volume || volume_1.Volume.One }));
    if (size.name === repr.defaultSizeTheme.name)
        Object.assign(sizeDefaultParams, repr.defaultSizeTheme.props);
    return { name: size.name, params: Object.assign(sizeDefaultParams, params) };
}
exports.createVolumeSizeThemeParams = createVolumeSizeThemeParams;
function createParamsByName(ctx, volume, props) {
    var typeProvider = (props.type && ctx.representation.volume.registry.get(props.type))
        || ctx.representation.volume.registry.default.provider;
    var colorProvider = (props.color && ctx.representation.volume.themes.colorThemeRegistry.get(props.color))
        || ctx.representation.volume.themes.colorThemeRegistry.get(typeProvider.defaultColorTheme.name);
    var sizeProvider = (props.size && ctx.representation.volume.themes.sizeThemeRegistry.get(props.size))
        || ctx.representation.volume.themes.sizeThemeRegistry.get(typeProvider.defaultSizeTheme.name);
    return createParamsProvider(ctx, volume, {
        type: typeProvider,
        typeParams: props.typeParams,
        color: colorProvider,
        colorParams: props.colorParams,
        size: sizeProvider,
        sizeParams: props.sizeParams
    });
}
function createParamsProvider(ctx, volume, props) {
    if (props === void 0) { props = {}; }
    var themeCtx = ctx.representation.volume.themes;
    var themeDataCtx = { volume: volume };
    var repr = props.type || ctx.representation.volume.registry.default.provider;
    var reprDefaultParams = param_definition_1.ParamDefinition.getDefaultValues(repr.getParams(themeCtx, volume));
    var reprParams = Object.assign(reprDefaultParams, props.typeParams);
    var color = props.color || themeCtx.colorThemeRegistry.get(repr.defaultColorTheme.name);
    var colorDefaultParams = param_definition_1.ParamDefinition.getDefaultValues(color.getParams(themeDataCtx));
    if (color.name === repr.defaultColorTheme.name)
        Object.assign(colorDefaultParams, repr.defaultColorTheme.props);
    var colorParams = Object.assign(colorDefaultParams, props.colorParams);
    var size = props.size || themeCtx.sizeThemeRegistry.get(repr.defaultSizeTheme.name);
    var sizeDefaultParams = param_definition_1.ParamDefinition.getDefaultValues(size.getParams(themeDataCtx));
    if (size.name === repr.defaultSizeTheme.name)
        Object.assign(sizeDefaultParams, repr.defaultSizeTheme.props);
    var sizeParams = Object.assign(sizeDefaultParams, props.sizeParams);
    return ({
        type: { name: repr.name, params: reprParams },
        colorTheme: { name: color.name, params: colorParams },
        sizeTheme: { name: size.name, params: sizeParams }
    });
}
