// molstar/lib/commonjs/mol-repr/volume/slice.js
"use strict";
/**
 * Copyright (c) 2020-2022 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SliceRepresentationProvider = exports.SliceParams = void 0;
exports.createImage = createImage;
exports.getSliceParams = getSliceParams;
exports.SliceVisual = SliceVisual;
exports.SliceRepresentation = SliceRepresentation;
const param_definition_1 = require("../../mol-util/param-definition");
const image_1 = require("../../mol-geo/geometry/image/image");
const volume_1 = require("../../mol-model/volume");
const representation_1 = require("./representation");
const location_iterator_1 = require("../../mol-geo/util/location-iterator");
const location_1 = require("../../mol-model/location");
const loci_1 = require("../../mol-model/loci");
const int_1 = require("../../mol-data/int");
const util_1 = require("../../mol-geo/util");
const color_1 = require("../../mol-util/color");
const color_2 = require("../../mol-theme/color");
const number_packing_1 = require("../../mol-util/number-packing");
const util_2 = require("./util");
async function createImage(ctx, volume, key, theme, props, image) {
    const { dimension: { name: dim }, isoValue } = props;
    const { space, data } = volume.grid.cells;
    const { min, max } = volume.grid.stats;
    const isoVal = volume_1.Volume.IsoValue.toAbsolute(isoValue, volume.grid.stats).absoluteValue;
    // TODO more color themes
    const color = 'color' in theme.color ? theme.color.color(location_1.NullLocation, false) : (0, color_1.Color)(0xffffff);
    const [r, g, b] = color_1.Color.toRgbNormalized(color);
    const { width, height, x, y, z, x0, y0, z0, nx, ny, nz } = getSliceInfo(volume.grid, props);
    const corners = new Float32Array(dim === 'x' ? [x, 0, 0, x, y, 0, x, 0, z, x, y, z] :
        dim === 'y' ? [0, y, 0, x, y, 0, 0, y, z, x, y, z] :
            [0, 0, z, 0, y, z, x, 0, z, x, y, z]);
    const imageArray = new Uint8Array(width * height * 4);
    const groupArray = getPackedGroupArray(volume.grid, props);
    let i = 0;
    for (let iy = y0; iy < ny; ++iy) {
        for (let ix = x0; ix < nx; ++ix) {
            for (let iz = z0; iz < nz; ++iz) {
                const val = space.get(data, ix, iy, iz);
                const normVal = (val - min) / (max - min);
                imageArray[i] = r * normVal * 2 * 255;
                imageArray[i + 1] = g * normVal * 2 * 255;
                imageArray[i + 2] = b * normVal * 2 * 255;
                imageArray[i + 3] = val >= isoVal ? 255 : 0;
                i += 4;
            }
        }
    }
    const imageTexture = { width, height, array: imageArray, flipY: true };
    const groupTexture = { width, height, array: groupArray, flipY: true };
    const transform = volume_1.Grid.getGridToCartesianTransform(volume.grid);
    (0, util_1.transformPositionArray)(transform, corners, 0, 4);
    return image_1.Image.create(imageTexture, corners, groupTexture, image);
}
function getSliceInfo(grid, props) {
    const { dimension: { name: dim, params: index } } = props;
    const { space } = grid.cells;
    let width, height;
    let x, y, z;
    let x0 = 0, y0 = 0, z0 = 0;
    let [nx, ny, nz] = space.dimensions;
    if (dim === 'x') {
        x = index, y = ny - 1, z = nz - 1;
        width = nz, height = ny;
        x0 = x, nx = x0 + 1;
    }
    else if (dim === 'y') {
        x = nx - 1, y = index, z = nz - 1;
        width = nz, height = nx;
        y0 = y, ny = y0 + 1;
    }
    else {
        x = nx - 1, y = ny - 1, z = index;
        width = nx, height = ny;
        z0 = z, nz = z0 + 1;
    }
    return {
        width, height,
        x, y, z,
        x0, y0, z0,
        nx, ny, nz
    };
}
function getPackedGroupArray(grid, props) {
    const { space } = grid.cells;
    const { width, height, x0, y0, z0, nx, ny, nz } = getSliceInfo(grid, props);
    const groupArray = new Uint8Array(width * height * 4);
    let j = 0;
    for (let iy = y0; iy < ny; ++iy) {
        for (let ix = x0; ix < nx; ++ix) {
            for (let iz = z0; iz < nz; ++iz) {
                (0, number_packing_1.packIntToRGBArray)(space.dataOffset(ix, iy, iz), groupArray, j);
                j += 4;
            }
        }
    }
    return groupArray;
}
function getGroupArray(grid, props) {
    const { space } = grid.cells;
    const { width, height, x0, y0, z0, nx, ny, nz } = getSliceInfo(grid, props);
    const groupArray = new Uint32Array(width * height);
    let j = 0;
    for (let iy = y0; iy < ny; ++iy) {
        for (let ix = x0; ix < nx; ++ix) {
            for (let iz = z0; iz < nz; ++iz) {
                groupArray[j] = space.dataOffset(ix, iy, iz);
                j += 1;
            }
        }
    }
    return groupArray;
}
function getLoci(volume, props) {
    // TODO cache somehow?
    const groupArray = getGroupArray(volume.grid, props);
    return volume_1.Volume.Cell.Loci(volume, int_1.SortedArray.ofUnsortedArray(groupArray));
}
function getSliceLoci(pickingId, volume, key, props, id) {
    const { objectId, groupId } = pickingId;
    if (id === objectId) {
        const granularity = volume_1.Volume.PickingGranularity.get(volume);
        if (granularity === 'volume') {
            return volume_1.Volume.Loci(volume);
        }
        if (granularity === 'object') {
            return getLoci(volume, props);
        }
        else {
            return volume_1.Volume.Cell.Loci(volume, int_1.Interval.ofSingleton(groupId));
        }
    }
    return loci_1.EmptyLoci;
}
function eachSlice(loci, volume, key, props, apply) {
    return (0, util_2.eachVolumeLoci)(loci, volume, undefined, apply);
}
//
exports.SliceParams = {
    ...image_1.Image.Params,
    quality: { ...image_1.Image.Params.quality, isEssential: false },
    dimension: param_definition_1.ParamDefinition.MappedStatic('x', {
        x: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 0, step: 1 }),
        y: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 0, step: 1 }),
        z: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 0, step: 1 }),
    }, { isEssential: true }),
    isoValue: volume_1.Volume.IsoValueParam,
};
function getSliceParams(ctx, volume) {
    const p = param_definition_1.ParamDefinition.clone(exports.SliceParams);
    const dim = volume.grid.cells.space.dimensions;
    p.dimension = param_definition_1.ParamDefinition.MappedStatic('x', {
        x: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: dim[0] - 1, step: 1 }),
        y: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: dim[1] - 1, step: 1 }),
        z: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: dim[2] - 1, step: 1 }),
    }, { isEssential: true });
    p.isoValue = volume_1.Volume.createIsoValueParam(volume_1.Volume.IsoValue.absolute(volume.grid.stats.min), volume.grid.stats);
    return p;
}
function SliceVisual(materialId) {
    return (0, representation_1.VolumeVisual)({
        defaultProps: param_definition_1.ParamDefinition.getDefaultValues(exports.SliceParams),
        createGeometry: createImage,
        createLocationIterator: (volume) => (0, location_iterator_1.LocationIterator)(volume.grid.cells.data.length, 1, 1, () => location_1.NullLocation),
        getLoci: getSliceLoci,
        eachLocation: eachSlice,
        setUpdateState: (state, volume, newProps, currentProps, newTheme, currentTheme) => {
            state.createGeometry = (newProps.dimension.name !== currentProps.dimension.name ||
                newProps.dimension.params !== currentProps.dimension.params ||
                !volume_1.Volume.IsoValue.areSame(newProps.isoValue, currentProps.isoValue, volume.grid.stats) ||
                !color_2.ColorTheme.areEqual(newTheme.color, currentTheme.color));
        },
        geometryUtils: {
            ...image_1.Image.Utils,
            createRenderableState: (props) => {
                const state = image_1.Image.Utils.createRenderableState(props);
                updateRenderableState(state, props);
                return state;
            },
            updateRenderableState
        }
    }, materialId);
}
function updateRenderableState(state, props) {
    image_1.Image.Utils.updateRenderableState(state, props);
    state.opaque = false;
    state.writeDepth = true;
}
function SliceRepresentation(ctx, getParams) {
    return (0, representation_1.VolumeRepresentation)('Slice', ctx, getParams, SliceVisual, getLoci);
}
exports.SliceRepresentationProvider = (0, representation_1.VolumeRepresentationProvider)({
    name: 'slice',
    label: 'Slice',
    description: 'Slice of volume rendered as image with interpolation.',
    factory: SliceRepresentation,
    getParams: getSliceParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.SliceParams),
    defaultColorTheme: { name: 'uniform' },
    defaultSizeTheme: { name: 'uniform' },
    isApplicable: (volume) => !volume_1.Volume.isEmpty(volume) && !volume_1.Volume.Segmentation.get(volume)
});
