// molstar/lib/commonjs/mol-model-props/common/custom-structure-property.js
"use strict";
/**
 * Copyright (c) 2019-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomStructureProperty = void 0;
var tslib_1 = require("tslib");
var param_definition_1 = require("../../mol-util/param-definition");
var mol_util_1 = require("../../mol-util");
var custom_property_1 = require("../../mol-model/custom-property");
var string_1 = require("../../mol-util/string");
var CustomStructureProperty;
(function (CustomStructureProperty) {
    function createProvider(builder) {
        var _this = this;
        var descriptorName = builder.descriptor.name;
        var propertyDataName = builder.type === 'root' ? 'inheritedPropertyData' : 'currentPropertyData';
        var get = function (data) {
            if (!(descriptorName in data[propertyDataName])) {
                data[propertyDataName][descriptorName] = {
                    props: tslib_1.__assign({}, param_definition_1.ParamDefinition.getDefaultValues(builder.getParams(data))),
                    data: mol_util_1.ValueBox.create(undefined)
                };
            }
            return data[propertyDataName][descriptorName];
        };
        var set = function (data, props, value) {
            var property = get(data);
            data[propertyDataName][descriptorName] = {
                props: props,
                data: mol_util_1.ValueBox.withValue(property.data, value)
            };
        };
        return {
            label: builder.label,
            descriptor: builder.descriptor,
            isHidden: builder.isHidden,
            getParams: function (data) {
                var params = param_definition_1.ParamDefinition.clone(builder.getParams(data));
                param_definition_1.ParamDefinition.setDefaultValues(params, get(data).props);
                return params;
            },
            defaultParams: builder.defaultParams,
            isApplicable: builder.isApplicable,
            attach: function (ctx, data, props, addRef) {
                if (props === void 0) { props = {}; }
                return tslib_1.__awaiter(_this, void 0, void 0, function () {
                    var rootProps, property, p, _a, value, assets;
                    return tslib_1.__generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                if (addRef)
                                    data.customPropertyDescriptors.reference(builder.descriptor, true);
                                if (builder.type === 'root')
                                    data = data.root;
                                rootProps = get(data.root).props;
                                property = get(data);
                                p = param_definition_1.ParamDefinition.merge(builder.defaultParams, rootProps, props);
                                if (property.data.value && param_definition_1.ParamDefinition.areEqual(builder.defaultParams, property.props, p))
                                    return [2 /*return*/];
                                return [4 /*yield*/, builder.obtain(ctx, data, p)];
                            case 1:
                                _a = _b.sent(), value = _a.value, assets = _a.assets;
                                data.customPropertyDescriptors.add(builder.descriptor);
                                data.customPropertyDescriptors.assets(builder.descriptor, assets);
                                set(data, p, value);
                                return [2 /*return*/];
                        }
                    });
                });
            },
            ref: function (data, add) { return data.customPropertyDescriptors.reference(builder.descriptor, add); },
            get: function (data) { return get(data).data; },
            set: function (data, props, value) {
                if (props === void 0) { props = {}; }
                if (builder.type === 'root')
                    data = data.root;
                var property = get(data);
                var p = param_definition_1.ParamDefinition.merge(builder.defaultParams, property.props, props);
                if (!param_definition_1.ParamDefinition.areEqual(builder.defaultParams, property.props, p)) {
                    // this invalidates property.value
                    set(data, p, value);
                    // dispose of assets
                    data.customPropertyDescriptors.assets(builder.descriptor);
                }
            },
            props: function (data) { return get(data).props; },
        };
    }
    CustomStructureProperty.createProvider = createProvider;
    function createSimple(name, type, defaultValue) {
        var _this = this;
        var defaultParams = { value: param_definition_1.ParamDefinition.Value(defaultValue, { isHidden: true }) };
        return createProvider({
            label: (0, string_1.stringToWords)(name),
            descriptor: (0, custom_property_1.CustomPropertyDescriptor)({ name: name }),
            isHidden: true,
            type: type,
            defaultParams: defaultParams,
            getParams: function () { return ({ value: param_definition_1.ParamDefinition.Value(defaultValue, { isHidden: true }) }); },
            isApplicable: function () { return true; },
            obtain: function (ctx, data, props) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                return tslib_1.__generator(this, function (_a) {
                    return [2 /*return*/, tslib_1.__assign(tslib_1.__assign({}, param_definition_1.ParamDefinition.getDefaultValues(defaultParams)), props)];
                });
            }); }
        });
    }
    CustomStructureProperty.createSimple = createSimple;
})(CustomStructureProperty || (CustomStructureProperty = {}));
exports.CustomStructureProperty = CustomStructureProperty;
