// molstar/lib/extensions/model-export/index.js
/**
 * Copyright (c) 2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
import { __extends } from "tslib";
import { PluginBehavior } from '../../mol-plugin/behavior/behavior';
import { ModelExportUI } from './ui';
export var ModelExport = PluginBehavior.create({
    name: 'extension-model-export',
    category: 'misc',
    display: {
        name: 'Model Export'
    },
    ctor: /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.register = function () {
            this.ctx.customStructureControls.set('model-export', ModelExportUI);
        };
        class_1.prototype.update = function () {
            return false;
        };
        class_1.prototype.unregister = function () {
            this.ctx.customStructureControls.delete('model-export');
        };
        return class_1;
    }(PluginBehavior.Handler)),
    params: function () { return ({}); }
});
