// molstar/lib/commonjs/mol-theme/color/atom-id.js
"use strict";
/**
 * Copyright (c) 2021 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AtomIdColorThemeProvider = exports.AtomIdColorTheme = exports.getAtomIdColorThemeParams = exports.AtomIdColorThemeParams = void 0;
var tslib_1 = require("tslib");
var structure_1 = require("../../mol-model/structure");
var color_1 = require("../../mol-util/color");
var param_definition_1 = require("../../mol-util/param-definition");
var palette_1 = require("../../mol-util/color/palette");
var DefaultList = 'many-distinct';
var DefaultColor = (0, color_1.Color)(0xFAFAFA);
var Description = 'Gives every atom a color based on its `label_atom_id` value.';
exports.AtomIdColorThemeParams = tslib_1.__assign({}, (0, palette_1.getPaletteParams)({ type: 'colors', colorList: DefaultList }));
function getAtomIdColorThemeParams(ctx) {
    var params = param_definition_1.ParamDefinition.clone(exports.AtomIdColorThemeParams);
    return params;
}
exports.getAtomIdColorThemeParams = getAtomIdColorThemeParams;
function getAtomIdSerialMap(structure) {
    var map = new Map();
    for (var _i = 0, _a = structure.models; _i < _a.length; _i++) {
        var m = _a[_i];
        var label_atom_id = m.atomicHierarchy.atoms.label_atom_id;
        for (var i = 0, il = label_atom_id.rowCount; i < il; ++i) {
            var id = label_atom_id.value(i);
            if (!map.has(id))
                map.set(id, map.size);
        }
    }
    return map;
}
function AtomIdColorTheme(ctx, props) {
    var color;
    var legend;
    if (ctx.structure) {
        var l_1 = structure_1.StructureElement.Location.create(ctx.structure.root);
        var atomIdSerialMap_1 = getAtomIdSerialMap(ctx.structure.root);
        var labelTable_1 = Array.from(atomIdSerialMap_1.keys());
        var valueLabel = function (i) { return labelTable_1[i]; };
        var palette_2 = (0, palette_1.getPalette)(atomIdSerialMap_1.size, props, { valueLabel: valueLabel });
        legend = palette_2.legend;
        color = function (location) {
            var serial = undefined;
            if (structure_1.StructureElement.Location.is(location)) {
                var id = structure_1.StructureProperties.atom.label_atom_id(location);
                serial = atomIdSerialMap_1.get(id);
            }
            else if (structure_1.Bond.isLocation(location)) {
                l_1.unit = location.aUnit;
                l_1.element = location.aUnit.elements[location.aIndex];
                var id = structure_1.StructureProperties.atom.label_atom_id(l_1);
                serial = atomIdSerialMap_1.get(id);
            }
            return serial === undefined ? DefaultColor : palette_2.color(serial);
        };
    }
    else {
        color = function () { return DefaultColor; };
    }
    return {
        factory: AtomIdColorTheme,
        granularity: 'group',
        preferSmoothing: true,
        color: color,
        props: props,
        description: Description,
        legend: legend
    };
}
exports.AtomIdColorTheme = AtomIdColorTheme;
exports.AtomIdColorThemeProvider = {
    name: 'atom-id',
    label: 'Atom Id',
    category: "Atom Property" /* ColorTheme.Category.Atom */,
    factory: AtomIdColorTheme,
    getParams: getAtomIdColorThemeParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.AtomIdColorThemeParams),
    isApplicable: function (ctx) { return !!ctx.structure; }
};
