// molstar/lib/commonjs/mol-theme/color/chain-id.js
"use strict";
/**
 * Copyright (c) 2018-2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainIdColorThemeProvider = exports.ChainIdColorTheme = exports.getChainIdColorThemeParams = exports.ChainIdColorThemeParams = void 0;
var tslib_1 = require("tslib");
var structure_1 = require("../../mol-model/structure");
var color_1 = require("../../mol-util/color");
var param_definition_1 = require("../../mol-util/param-definition");
var palette_1 = require("../../mol-util/color/palette");
var DefaultList = 'many-distinct';
var DefaultColor = (0, color_1.Color)(0xFAFAFA);
var Description = 'Gives every chain a color based on its `asym_id` value.';
exports.ChainIdColorThemeParams = (0, tslib_1.__assign)({ asymId: param_definition_1.ParamDefinition.Select('auth', param_definition_1.ParamDefinition.arrayToOptions(['auth', 'label'])) }, (0, palette_1.getPaletteParams)({ type: 'colors', colorList: DefaultList }));
function getChainIdColorThemeParams(ctx) {
    var _a;
    var params = param_definition_1.ParamDefinition.clone(exports.ChainIdColorThemeParams);
    if ((_a = ctx.structure) === null || _a === void 0 ? void 0 : _a.models.some(function (m) { return m.coarseHierarchy.isDefined; })) {
        params.asymId.defaultValue = 'label';
    }
    return params;
}
exports.getChainIdColorThemeParams = getChainIdColorThemeParams;
function getAsymId(unit, type) {
    switch (unit.kind) {
        case 0 /* Atomic */:
            return type === 'auth'
                ? structure_1.StructureProperties.chain.auth_asym_id
                : structure_1.StructureProperties.chain.label_asym_id;
        case 1 /* Spheres */:
        case 2 /* Gaussians */:
            return structure_1.StructureProperties.coarse.asym_id;
    }
}
function getAsymIdKey(location, type) {
    var asymId = getAsymId(location.unit, type)(location);
    return location.structure.root.models.length > 1
        ? getKey(location.unit.model, asymId)
        : asymId;
}
function getKey(model, asymId) {
    return "".concat(asymId, "|").concat((structure_1.Model.Index.get(model).value || 0) + 1);
}
function getAsymIdSerialMap(structure, type) {
    var map = new Map();
    var _loop_1 = function (m) {
        var asymIdOffset = structure_1.Model.AsymIdOffset.get(m).value;
        var offset = (type === 'auth' ? asymIdOffset === null || asymIdOffset === void 0 ? void 0 : asymIdOffset.auth : asymIdOffset === null || asymIdOffset === void 0 ? void 0 : asymIdOffset.label) || 0;
        var count = 0;
        m.properties.structAsymMap.forEach(function (_a, label_id) {
            var auth_id = _a.auth_id;
            var asymId = type === 'auth' ? auth_id : label_id;
            var k = structure.models.length > 1
                ? getKey(m, asymId)
                : asymId;
            if (!map.has(k)) {
                map.set(k, count + offset);
                ++count;
            }
        });
    };
    for (var _i = 0, _a = structure.models; _i < _a.length; _i++) {
        var m = _a[_i];
        _loop_1(m);
    }
    return map;
}
function ChainIdColorTheme(ctx, props) {
    var color;
    var legend;
    if (ctx.structure) {
        var l_1 = structure_1.StructureElement.Location.create(ctx.structure.root);
        var asymIdSerialMap_1 = getAsymIdSerialMap(ctx.structure.root, props.asymId);
        var labelTable_1 = Array.from(asymIdSerialMap_1.keys());
        var valueLabel = function (i) { return labelTable_1[i]; };
        var palette_2 = (0, palette_1.getPalette)(asymIdSerialMap_1.size, props, { valueLabel: valueLabel });
        legend = palette_2.legend;
        color = function (location) {
            var serial = undefined;
            if (structure_1.StructureElement.Location.is(location)) {
                var k = getAsymIdKey(location, props.asymId);
                serial = asymIdSerialMap_1.get(k);
            }
            else if (structure_1.Bond.isLocation(location)) {
                l_1.unit = location.aUnit;
                l_1.element = location.aUnit.elements[location.aIndex];
                var k = getAsymIdKey(l_1, props.asymId);
                serial = asymIdSerialMap_1.get(k);
            }
            return serial === undefined ? DefaultColor : palette_2.color(serial);
        };
    }
    else {
        color = function () { return DefaultColor; };
    }
    return {
        factory: ChainIdColorTheme,
        granularity: 'group',
        color: color,
        props: props,
        description: Description,
        legend: legend
    };
}
exports.ChainIdColorTheme = ChainIdColorTheme;
exports.ChainIdColorThemeProvider = {
    name: 'chain-id',
    label: 'Chain Id',
    category: "Chain Property" /* Chain */,
    factory: ChainIdColorTheme,
    getParams: getChainIdColorThemeParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.ChainIdColorThemeParams),
    isApplicable: function (ctx) { return !!ctx.structure; }
};
