// molstar/lib/commonjs/mol-theme/color/polymer-id.js
"use strict";
/**
 * Copyright (c) 2018-2019 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolymerIdColorThemeProvider = exports.PolymerIdColorTheme = exports.getPolymerIdColorThemeParams = exports.PolymerIdColorThemeParams = void 0;
var tslib_1 = require("tslib");
var structure_1 = require("../../mol-model/structure");
var color_1 = require("../../mol-util/color");
var param_definition_1 = require("../../mol-util/param-definition");
var palette_1 = require("../../mol-util/color/palette");
var int_1 = require("../../mol-data/int");
var lists_1 = require("../../mol-util/color/lists");
var DefaultList = 'dark-2';
var DefaultColor = (0, color_1.Color)(0xFAFAFA);
var Description = 'Gives every polymer chain a color based on its `asym_id` value.';
exports.PolymerIdColorThemeParams = tslib_1.__assign({}, (0, palette_1.getPaletteParams)({ type: 'colors', colorList: DefaultList }));
function getPolymerIdColorThemeParams(ctx) {
    var params = param_definition_1.ParamDefinition.clone(exports.PolymerIdColorThemeParams);
    if (ctx.structure) {
        if (getPolymerAsymIdSerialMap(ctx.structure.root).size > lists_1.ColorLists[DefaultList].list.length) {
            params.palette.defaultValue.name = 'colors';
            params.palette.defaultValue.params = tslib_1.__assign(tslib_1.__assign({}, params.palette.defaultValue.params), { list: { kind: 'interpolate', colors: (0, lists_1.getColorListFromName)(DefaultList).list } });
        }
    }
    return params;
}
exports.getPolymerIdColorThemeParams = getPolymerIdColorThemeParams;
function getAsymId(unit) {
    switch (unit.kind) {
        case 0 /* Unit.Kind.Atomic */:
            return structure_1.StructureProperties.chain.label_asym_id;
        case 1 /* Unit.Kind.Spheres */:
        case 2 /* Unit.Kind.Gaussians */:
            return structure_1.StructureProperties.coarse.asym_id;
    }
}
function getPolymerAsymIdSerialMap(structure) {
    var map = new Map();
    for (var i = 0, il = structure.unitSymmetryGroups.length; i < il; ++i) {
        var unit = structure.unitSymmetryGroups[i].units[0];
        var model = unit.model;
        if (structure_1.Unit.isAtomic(unit)) {
            var _a = model.atomicHierarchy, chainAtomSegments = _a.chainAtomSegments, chains = _a.chains;
            var chainIt = int_1.Segmentation.transientSegments(chainAtomSegments, unit.elements);
            while (chainIt.hasNext) {
                var chainIndex = chainIt.move().index;
                var entityId = chains.label_entity_id.value(chainIndex);
                var eI = model.entities.getEntityIndex(entityId);
                if (model.entities.data.type.value(eI) === 'polymer') {
                    var asymId = chains.label_asym_id.value(chainIndex);
                    if (!map.has(asymId))
                        map.set(asymId, map.size);
                }
            }
        }
        else if (structure_1.Unit.isCoarse(unit)) {
            var _b = structure_1.Unit.isSpheres(unit)
                ? model.coarseHierarchy.spheres
                : model.coarseHierarchy.gaussians, chainElementSegments = _b.chainElementSegments, asym_id = _b.asym_id, entity_id = _b.entity_id;
            var chainIt = int_1.Segmentation.transientSegments(chainElementSegments, unit.elements);
            while (chainIt.hasNext) {
                var chainIndex = chainIt.move().index;
                var elementIndex = chainElementSegments.offsets[chainIndex];
                var entityId = entity_id.value(elementIndex);
                var eI = model.entities.getEntityIndex(entityId);
                if (model.entities.data.type.value(eI) === 'polymer') {
                    var asymId = asym_id.value(elementIndex);
                    if (!map.has(asymId))
                        map.set(asymId, map.size);
                }
            }
        }
    }
    return map;
}
function PolymerIdColorTheme(ctx, props) {
    var color;
    var legend;
    if (ctx.structure) {
        var l_1 = structure_1.StructureElement.Location.create(ctx.structure);
        var polymerAsymIdSerialMap_1 = getPolymerAsymIdSerialMap(ctx.structure.root);
        var labelTable_1 = Array.from(polymerAsymIdSerialMap_1.keys());
        var valueLabel = function (i) { return labelTable_1[i]; };
        var palette_2 = (0, palette_1.getPalette)(polymerAsymIdSerialMap_1.size, props, { valueLabel: valueLabel });
        legend = palette_2.legend;
        color = function (location) {
            var serial = undefined;
            if (structure_1.StructureElement.Location.is(location)) {
                var asym_id = getAsymId(location.unit);
                serial = polymerAsymIdSerialMap_1.get(asym_id(location));
            }
            else if (structure_1.Bond.isLocation(location)) {
                var asym_id = getAsymId(location.aUnit);
                l_1.unit = location.aUnit;
                l_1.element = location.aUnit.elements[location.aIndex];
                serial = polymerAsymIdSerialMap_1.get(asym_id(l_1));
            }
            return serial === undefined ? DefaultColor : palette_2.color(serial);
        };
    }
    else {
        color = function () { return DefaultColor; };
    }
    return {
        factory: PolymerIdColorTheme,
        granularity: 'group',
        color: color,
        props: props,
        description: Description,
        legend: legend
    };
}
exports.PolymerIdColorTheme = PolymerIdColorTheme;
exports.PolymerIdColorThemeProvider = {
    name: 'polymer-id',
    label: 'Polymer Chain Id',
    category: "Chain Property" /* ColorTheme.Category.Chain */,
    factory: PolymerIdColorTheme,
    getParams: getPolymerIdColorThemeParams,
    defaultValues: param_definition_1.ParamDefinition.getDefaultValues(exports.PolymerIdColorThemeParams),
    isApplicable: function (ctx) { return !!ctx.structure; }
};
