// molstar/lib/commonjs/examples/basic-wrapper/index.js
"use strict";
/**
 * Copyright (c) 2019 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var pdbe_1 = require("../../extensions/pdbe");
var loci_1 = require("../../mol-model/loci");
var structure_1 = require("../../mol-model/structure");
var model_index_1 = require("../../mol-plugin-state/animation/built-in/model-index");
var mol_plugin_ui_1 = require("../../mol-plugin-ui");
var spec_1 = require("../../mol-plugin-ui/spec");
var commands_1 = require("../../mol-plugin/commands");
var script_1 = require("../../mol-script/script");
var assets_1 = require("../../mol-util/assets");
var color_1 = require("../../mol-util/color");
var coloring_1 = require("./coloring");
var controls_1 = require("./controls");
var custom_theme_1 = require("./custom-theme");
require("./index.html");
var superposition_1 = require("./superposition");
require('mol-plugin-ui/skin/light.scss');
var BasicWrapper = /** @class */ (function () {
    function BasicWrapper() {
        var _this = this;
        this.animate = {
            modelIndex: {
                targetFps: 8,
                onceForward: function () { _this.plugin.managers.animation.play(model_index_1.AnimateModelIndex, { duration: { name: 'computed', params: { targetFps: _this.animateModelIndexTargetFps() } }, mode: { name: 'once', params: { direction: 'forward' } } }); },
                onceBackward: function () { _this.plugin.managers.animation.play(model_index_1.AnimateModelIndex, { duration: { name: 'computed', params: { targetFps: _this.animateModelIndexTargetFps() } }, mode: { name: 'once', params: { direction: 'backward' } } }); },
                palindrome: function () { _this.plugin.managers.animation.play(model_index_1.AnimateModelIndex, { duration: { name: 'computed', params: { targetFps: _this.animateModelIndexTargetFps() } }, mode: { name: 'palindrome', params: {} } }); },
                loop: function () { _this.plugin.managers.animation.play(model_index_1.AnimateModelIndex, { duration: { name: 'computed', params: { targetFps: _this.animateModelIndexTargetFps() } }, mode: { name: 'loop', params: { direction: 'forward' } } }); },
                stop: function () { return _this.plugin.managers.animation.stop(); }
            }
        };
        this.coloring = {
            applyStripes: function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                var _this = this;
                return (0, tslib_1.__generator)(this, function (_a) {
                    this.plugin.dataTransaction(function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                        var _i, _a, s;
                        return (0, tslib_1.__generator)(this, function (_b) {
                            switch (_b.label) {
                                case 0:
                                    _i = 0, _a = this.plugin.managers.structure.hierarchy.current.structures;
                                    _b.label = 1;
                                case 1:
                                    if (!(_i < _a.length)) return [3 /*break*/, 4];
                                    s = _a[_i];
                                    return [4 /*yield*/, this.plugin.managers.structure.component.updateRepresentationsTheme(s.components, { color: coloring_1.StripedResidues.propertyProvider.descriptor.name })];
                                case 2:
                                    _b.sent();
                                    _b.label = 3;
                                case 3:
                                    _i++;
                                    return [3 /*break*/, 1];
                                case 4: return [2 /*return*/];
                            }
                        });
                    }); });
                    return [2 /*return*/];
                });
            }); },
            applyCustomTheme: function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                var _this = this;
                return (0, tslib_1.__generator)(this, function (_a) {
                    this.plugin.dataTransaction(function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                        var _i, _a, s;
                        return (0, tslib_1.__generator)(this, function (_b) {
                            switch (_b.label) {
                                case 0:
                                    _i = 0, _a = this.plugin.managers.structure.hierarchy.current.structures;
                                    _b.label = 1;
                                case 1:
                                    if (!(_i < _a.length)) return [3 /*break*/, 4];
                                    s = _a[_i];
                                    return [4 /*yield*/, this.plugin.managers.structure.component.updateRepresentationsTheme(s.components, { color: custom_theme_1.CustomColorThemeProvider.name })];
                                case 2:
                                    _b.sent();
                                    _b.label = 3;
                                case 3:
                                    _i++;
                                    return [3 /*break*/, 1];
                                case 4: return [2 /*return*/];
                            }
                        });
                    }); });
                    return [2 /*return*/];
                });
            }); },
            applyDefault: function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                var _this = this;
                return (0, tslib_1.__generator)(this, function (_a) {
                    this.plugin.dataTransaction(function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                        var _i, _a, s;
                        return (0, tslib_1.__generator)(this, function (_b) {
                            switch (_b.label) {
                                case 0:
                                    _i = 0, _a = this.plugin.managers.structure.hierarchy.current.structures;
                                    _b.label = 1;
                                case 1:
                                    if (!(_i < _a.length)) return [3 /*break*/, 4];
                                    s = _a[_i];
                                    return [4 /*yield*/, this.plugin.managers.structure.component.updateRepresentationsTheme(s.components, { color: 'default' })];
                                case 2:
                                    _b.sent();
                                    _b.label = 3;
                                case 3:
                                    _i++;
                                    return [3 /*break*/, 1];
                                case 4: return [2 /*return*/];
                            }
                        });
                    }); });
                    return [2 /*return*/];
                });
            }); }
        };
        this.interactivity = {
            highlightOn: function () {
                var _a, _b;
                var data = (_b = (_a = _this.plugin.managers.structure.hierarchy.current.structures[0]) === null || _a === void 0 ? void 0 : _a.cell.obj) === null || _b === void 0 ? void 0 : _b.data;
                if (!data)
                    return;
                var seq_id = 7;
                var sel = script_1.Script.getStructureSelection(function (Q) { return Q.struct.generator.atomGroups({
                    'residue-test': Q.core.rel.eq([Q.struct.atomProperty.macromolecular.label_seq_id(), seq_id]),
                    'group-by': Q.struct.atomProperty.macromolecular.residueKey()
                }); }, data);
                var loci = structure_1.StructureSelection.toLociWithSourceUnits(sel);
                _this.plugin.managers.interactivity.lociHighlights.highlightOnly({ loci: loci });
            },
            clearHighlight: function () {
                _this.plugin.managers.interactivity.lociHighlights.highlightOnly({ loci: loci_1.EmptyLoci });
            }
        };
        this.tests = {
            staticSuperposition: function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                return (0, tslib_1.__generator)(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.plugin.clear()];
                        case 1:
                            _a.sent();
                            return [2 /*return*/, (0, superposition_1.buildStaticSuperposition)(this.plugin, superposition_1.StaticSuperpositionTestData)];
                    }
                });
            }); },
            dynamicSuperposition: function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                return (0, tslib_1.__generator)(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, this.plugin.clear()];
                        case 1:
                            _a.sent();
                            return [2 /*return*/, (0, superposition_1.dynamicSuperpositionTest)(this.plugin, ['1tqn', '2hhb', '4hhb'], 'HEM')];
                    }
                });
            }); },
            toggleValidationTooltip: function () {
                return _this.plugin.state.updateBehavior(pdbe_1.PDBeStructureQualityReport, function (params) { params.showTooltip = !params.showTooltip; });
            },
            showToasts: function () {
                commands_1.PluginCommands.Toast.Show(_this.plugin, {
                    title: 'Toast 1',
                    message: 'This is an example text, timeout 3s',
                    key: 'toast-1',
                    timeoutMs: 3000
                });
                commands_1.PluginCommands.Toast.Show(_this.plugin, {
                    title: 'Toast 2',
                    message: controls_1.CustomToastMessage,
                    key: 'toast-2'
                });
            },
            hideToasts: function () {
                commands_1.PluginCommands.Toast.Hide(_this.plugin, { key: 'toast-1' });
                commands_1.PluginCommands.Toast.Hide(_this.plugin, { key: 'toast-2' });
            }
        };
    }
    BasicWrapper.prototype.init = function (target) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var _a;
            return (0, tslib_1.__generator)(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4 /*yield*/, (0, mol_plugin_ui_1.createPluginUI)(typeof target === 'string' ? document.getElementById(target) : target, (0, tslib_1.__assign)((0, tslib_1.__assign)({}, (0, spec_1.DefaultPluginUISpec)()), { layout: {
                                    initial: {
                                        isExpanded: false,
                                        showControls: false
                                    }
                                }, components: {
                                    remoteState: 'none'
                                } }))];
                    case 1:
                        _a.plugin = _b.sent();
                        this.plugin.representation.structure.themes.colorThemeRegistry.add(coloring_1.StripedResidues.colorThemeProvider);
                        this.plugin.representation.structure.themes.colorThemeRegistry.add(custom_theme_1.CustomColorThemeProvider);
                        this.plugin.managers.lociLabels.addProvider(coloring_1.StripedResidues.labelProvider);
                        this.plugin.customModelProperties.register(coloring_1.StripedResidues.propertyProvider, true);
                        return [2 /*return*/];
                }
            });
        });
    };
    BasicWrapper.prototype.load = function (_a) {
        var url = _a.url, _b = _a.format, format = _b === void 0 ? 'mmcif' : _b, _c = _a.isBinary, isBinary = _c === void 0 ? false : _c, _d = _a.assemblyId, assemblyId = _d === void 0 ? '' : _d;
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var data, trajectory;
            return (0, tslib_1.__generator)(this, function (_e) {
                switch (_e.label) {
                    case 0: return [4 /*yield*/, this.plugin.clear()];
                    case 1:
                        _e.sent();
                        return [4 /*yield*/, this.plugin.builders.data.download({ url: assets_1.Asset.Url(url), isBinary: isBinary }, { state: { isGhost: true } })];
                    case 2:
                        data = _e.sent();
                        return [4 /*yield*/, this.plugin.builders.structure.parseTrajectory(data, format)];
                    case 3:
                        trajectory = _e.sent();
                        return [4 /*yield*/, this.plugin.builders.structure.hierarchy.applyPreset(trajectory, 'default', {
                                structure: assemblyId ? {
                                    name: 'assembly',
                                    params: { id: assemblyId }
                                } : {
                                    name: 'model',
                                    params: {}
                                },
                                showUnitcell: false,
                                representationPreset: 'auto'
                            })];
                    case 4:
                        _e.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    BasicWrapper.prototype.setBackground = function (color) {
        commands_1.PluginCommands.Canvas3D.SetSettings(this.plugin, { settings: function (props) { props.renderer.backgroundColor = (0, color_1.Color)(color); } });
    };
    BasicWrapper.prototype.toggleSpin = function () {
        if (!this.plugin.canvas3d)
            return;
        var trackball = this.plugin.canvas3d.props.trackball;
        commands_1.PluginCommands.Canvas3D.SetSettings(this.plugin, {
            settings: {
                trackball: (0, tslib_1.__assign)((0, tslib_1.__assign)({}, trackball), { animate: trackball.animate.name === 'spin'
                        ? { name: 'off', params: {} }
                        : { name: 'spin', params: { speed: 1 } } })
            }
        });
        if (this.plugin.canvas3d.props.trackball.animate.name !== 'spin') {
            commands_1.PluginCommands.Camera.Reset(this.plugin, {});
        }
    };
    BasicWrapper.prototype.animateModelIndexTargetFps = function () {
        return Math.max(1, this.animate.modelIndex.targetFps | 0);
    };
    return BasicWrapper;
}());
window.BasicMolStarWrapper = new BasicWrapper();
