// molstar/lib/commonjs/mol-plugin/behavior.js
"use strict";
/**
 * Copyright (c) 2018 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginBehaviors = exports.BuiltInPluginBehaviors = void 0;
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./behavior/behavior"), exports);
var StaticState = tslib_1.__importStar(require("./behavior/static/state"));
var StaticRepresentation = tslib_1.__importStar(require("./behavior/static/representation"));
var StaticCamera = tslib_1.__importStar(require("./behavior/static/camera"));
var StaticMisc = tslib_1.__importStar(require("./behavior/static/misc"));
var DynamicRepresentation = tslib_1.__importStar(require("./behavior/dynamic/representation"));
var DynamicCamera = tslib_1.__importStar(require("./behavior/dynamic/camera"));
var DynamicCustomProps = tslib_1.__importStar(require("./behavior/dynamic/custom-props"));
exports.BuiltInPluginBehaviors = {
    State: StaticState,
    Representation: StaticRepresentation,
    Camera: StaticCamera,
    Misc: StaticMisc
};
exports.PluginBehaviors = {
    Representation: DynamicRepresentation,
    Camera: DynamicCamera,
    CustomProps: DynamicCustomProps
};
