// molstar/lib/commonjs/extensions/mp4-export/index.js
"use strict";
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mp4Export = void 0;
var tslib_1 = require("tslib");
var behavior_1 = require("../../mol-plugin/behavior/behavior");
var ui_1 = require("./ui");
exports.Mp4Export = behavior_1.PluginBehavior.create({
    name: 'extension-mp4-export',
    category: 'misc',
    display: {
        name: 'MP4 Animation Export'
    },
    ctor: /** @class */ (function (_super) {
        tslib_1.__extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.register = function () {
            this.ctx.customStructureControls.set('mp4-export', ui_1.Mp4EncoderUI);
        };
        class_1.prototype.update = function () {
            return false;
        };
        class_1.prototype.unregister = function () {
            this.ctx.customStructureControls.delete('mp4-export');
        };
        return class_1;
    }(behavior_1.PluginBehavior.Handler)),
    params: function () { return ({}); }
});
