// molstar/lib/extensions/mp4-export/index.js
/**
 * Copyright (c) 2020 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */
import { __extends } from "tslib";
import { PluginBehavior } from '../../mol-plugin/behavior/behavior';
import { Mp4EncoderUI } from './ui';
export var Mp4Export = PluginBehavior.create({
    name: 'extension-mp4-export',
    category: 'misc',
    display: {
        name: 'MP4 Animation Export'
    },
    ctor: /** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.register = function () {
            this.ctx.customStructureControls.set('mp4-export', Mp4EncoderUI);
        };
        class_1.prototype.update = function () {
            return false;
        };
        class_1.prototype.unregister = function () {
            this.ctx.customStructureControls.delete('mp4-export');
        };
        return class_1;
    }(PluginBehavior.Handler)),
    params: function () { return ({}); }
});
